{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "75e102ec-f290-413c-9221-22f64a023cf4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "import os\n",
    "import datetime\n",
    "\n",
    "%matplotlib inline\n",
    "\n",
    "from selene_sdk.utils import load_path\n",
    "from selene_sdk.utils import parse_configs_and_run\n",
    "from selene_sdk.utils import DeeperDeepSEA\n",
    "from selene_sdk.predict._common import predict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "137026e7-8b30-499d-968c-587621f18c7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "dff361a6-d05c-4695-b634-e301a71b027e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c682e41d-0d8e-4d94-bd51-dbd815edb29c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from selene_sdk.utils import NonStrandSpecific\n",
    "\n",
    "\n",
    "model_architecture = NonStrandSpecific(DeeperDeepSEA(2000, 1846))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4cdc04b7-7bb8-44dd-90c1-ca9ff0916f48",
   "metadata": {},
   "outputs": [],
   "source": [
    "from selene_sdk.predict import AnalyzeSequences\n",
    "from selene_sdk.utils import load_features_list\n",
    "\n",
    "\n",
    "features = load_features_list(\"/rds-d5/project/who1000/rds-who1000-wgs10k/user/tt419/Selene_data/selene_ftp_output/distinct_features.txt\")\n",
    "analysis = AnalyzeSequences(\n",
    "    model_architecture,\n",
    "    \"/rds-d5/project/who1000/rds-who1000-wgs10k/user/tt419/Selene_data/selene_ftp_output/logs/online_sampler_outputs_base/best_model.pth.tar\",\n",
    "    sequence_length=2000,\n",
    "    features=features,\n",
    "    use_cuda=False)\n",
    "\n",
    "#analysis.in_silico_mutagenesis_from_file(\"sequences.fasta\", \n",
    "#                                         save_data=[\"abs_diffs\", \"logits\", \"predictions\"], \n",
    "#                                         output_dir=\".\",\n",
    "#                                         use_sequence_name=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7b24a12a-6fa4-421a-97ab-a66ade33729c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "\n",
    "server = \"https://grch37.rest.ensembl.org\"\n",
    "\n",
    "def get_reference(chrom, start, end=None, species=\"human\", GRCh=\"GRCh37\"):\n",
    "    if not end:\n",
    "        end = start+5000\n",
    "    assert start <= end, \"Start has to be before end\"\n",
    "    assert end - start < 10000000, \"Can't fetch more then 10 Mb (10 million bases)\"\n",
    "    sub_path = f\"/sequence/region/{species}/\"\n",
    "    region = f\"{chrom}:{start}..{end}\"\n",
    "    r = requests.get(server + sub_path + region, headers={\"Content-Type\": \"text/plain\", \"coord_system_version\": str(GRCh)})\n",
    "    if r.ok:\n",
    "        return r.text\n",
    "    else:\n",
    "        r.raise_for_status()\n",
    "        return -1\n",
    "\n",
    "def find_variant_in_reference(variant_coord, reference, ref_start, length=2000, focus=200):\n",
    "    \"\"\"\n",
    "    @param variant_coord: (chrom, pos,ref, alt)\n",
    "    \"\"\"\n",
    "    index = int(variant_coord[1])-int(ref_start)\n",
    "    ref_length = len(variant_coord[2])\n",
    "    ref_ens = reference[index:(index+ref_length)]\n",
    "    # if ref_ens == variant_coord[2]:\n",
    "    #     print(f\"The reference is not the same!Ensemble: {ref_ens}, variant_ref: {variant_coord[2]}\")\n",
    "\n",
    "    #assert ref_ens == variant_coord[2], f\"The reference is not the same!Ensemble: {reference[index-1:(index+ref_length)]}, variant_ref: {variant_coord[2]}\"\n",
    "    assert len(variant_coord[3]) <= focus, f\"The alternate length ({len(variant_coord[3])}) is longer than the focus of the network: {focus} \"\n",
    "    result = []\n",
    "    start_index = int(index-(length-focus)/2-(focus-len(variant_coord[3]))/2)\n",
    "    end_index = int(index + ref_length + (length-focus)/2 + (focus-len(variant_coord[3]))/2)\n",
    "    # The following is handling all cases: Substitution, Insertion and Deletion. It takes the whole reference up to the\n",
    "    #   variant start, then appends the variant itself (in case of a deletion it is short than before, in case of an insertion longer)\n",
    "    #   then adds the last bit of the reference, from starting index+ref_length to our result. This way the length difference of\n",
    "    #   InDels is handled.\n",
    "    result.extend(reference[start_index:index])\n",
    "    result.extend(variant_coord[3])\n",
    "    result.extend(reference[index + ref_length:end_index])\n",
    "    result = \"\".join(result)\n",
    "    diff = length-len(result)\n",
    "    if diff > 0:\n",
    "        print(f\"Adding {diff} basepairs in the end.\")\n",
    "        result.append(reference[end_index:end_index+diff])\n",
    "    elif diff < 0:\n",
    "        print(f\"Erasing {diff} basepairs in the end\")\n",
    "        result = result[:length]\n",
    "    return result, reference[start_index:start_index+length]\n",
    "\n",
    "def one_hot(sequence):  \n",
    "    classes = 4    \n",
    "    encode = {\"A\":0, \"C\":1, \"G\":2, \"T\":3}\n",
    "    num_sequence = list(map(encode.get, sequence))    \n",
    "    targets = np.array([num_sequence]).reshape(-1)    \n",
    "    result = np.eye(classes)[np.array(targets).reshape(-1)]      \n",
    "    return result.reshape(list(targets.shape)+[classes])    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "29549e53-e1e4-4d26-bb9d-7231cb382224",
   "metadata": {},
   "outputs": [],
   "source": [
    "def iterate_through_mutations_in_sequence(ref_seq, ref_start, var_df):\n",
    "    for index, var in var_df.iterrows():\n",
    "        yield find_variant_in_reference((var.chromosome, var.start, var.reference, var.alternate), ref_seq, ref_start)\n",
    "\n",
    "def get_ref_for_gene(var_df, frame_length=1000): \n",
    "    chrom = var_df[\"chromosome\"][0]\n",
    "    start = var_df[\"start\"].min() - frame_length\n",
    "    end = var_df[\"end\"].max() + frame_length\n",
    "    return (chrom, start, end),get_reference(chrom, start, end)           \n",
    "        \n",
    "        \n",
    "def load_gene(fp):\n",
    "    var_df = pd.read_pickle(fp)\n",
    "    return var_df[var_df.columns[:5].tolist()+var_df.columns[-6:].tolist()+\n",
    "                  var_df.columns[5:-6].tolist()]\n",
    "\n",
    "\n",
    "\n",
    "def predict_whole_gene(analysis, var_df):\n",
    "    coord, reference = get_ref_for_gene(var_df)\n",
    "    if coord[1] <= 55267612 and coord[2] >= 55267612:\n",
    "        print(\"Found rs13266183\")#\n",
    "    if coord[1] <= 55258127 and coord[2] >= 55258127:\n",
    "        print(\"Found rs10103692\")\n",
    "\n",
    "    iter_count = 0\n",
    "    if reference == -1:\n",
    "        return np.array([-1])\n",
    "    for frame,ref_frame in iterate_through_mutations_in_sequence(reference, coord[1], var_df):\n",
    "        encoded_frame = one_hot(frame)\n",
    "        encoded_ref_frame = one_hot(ref_frame)\n",
    "        sequences = np.stack([encoded_ref_frame.astype(int), encoded_frame.astype(int)], 0)\n",
    "        yield predict(analysis.model, sequences, use_cuda=analysis.use_cuda) \n",
    "        iter_count += 1\n",
    "        if iter_count % 100 == 0:\n",
    "            print(f\"Iteration {iter_count} done.\")\n",
    "    print(f\"Done in {iter_count} steps.\")        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "5346829d-fd3b-4cc9-aba9-f08124dbe865",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def run(analysis, gene_dir=\".\", new_fp_dir=\"\"):\n",
    "    \n",
    "    print(f\"It is now {datetime.datetime.now()}\")\n",
    "    # iterate through all genes in gene_dir, grab reference, predict each variant, save in separate directory -> save in pandas df\n",
    "    print(f\"Running with gene_dir = {gene_dir}, and new_fp_dir = {new_fp_dir}\")\n",
    "    for root, dirs, files in os.walk(gene_dir):\n",
    "        for filename in files:\n",
    "            print(f\"There is a file {filename}\")\n",
    "            if \".pickle\" in filename[-8:] and \"avro37_all\" in filename[:11] and \"SOX17\" in filename:\n",
    "                try:\n",
    "                    print(f\"Loading {filename}\")\n",
    "                    gene = load_gene(os.path.join(gene_dir,filename))\n",
    "                    gene_name = \"_\".join(filename.split(\"_\")[2:])[:-7]\n",
    "                    new_fp = f\"{new_fp_dir}predicted_{gene_name}.pickle\"\n",
    "                    if os.path.isfile(new_fp):\n",
    "                        print(f\"Gene {gene_name} is already in {new_fp}\")\n",
    "                        continue\n",
    "                    print(f\"Predicting for {gene_name}\")\n",
    "                    gene[\"prediction\"] = list(predict_whole_gene(analysis, gene))\n",
    "                    gene.to_pickle(new_fp)\n",
    "                    print(f\"Done with {gene_name}\")\n",
    "                    print(f\"It is now {datetime.datetime.now()}\")\n",
    "                except Exception as e:\n",
    "                    print(f\"Something went wrong in {filename}\")\n",
    "                    \n",
    "                    raise\n",
    "\n",
    "def main():\n",
    "    gene_dir = \"/home/tt419/Projects/data/104genes_37_acan/\"\n",
    "    # Run with my model\n",
    "    run(analysis ,gene_dir=gene_dir, new_fp_dir=\"../logs/myImpl_37_98genes-gnomad/\")\n",
    "    print(\"Done with first predictions\")\n",
    "    # Run with pretrained model by zhang (troyanskaya group)\n",
    "    #run(gene_dir=gene_dir,model_fp=\"../deepsea_keras.h5\", \\\n",
    "    #new_fp_dir=\"../logs/zhangImpl_AGCT/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0576936d-e046-4a24-892c-87ce0c53a934",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "It is now 2021-04-26 17:47:20.508754\n",
      "Running with gene_dir = /home/tt419/Projects/data/104genes_37_acan/, and new_fp_dir = /rds-d5/project/who1000/rds-who1000-wgs10k/user/tt419/Selene_data/selene_ftp_output/logs/mySelene_37_104genes-gnomad/\n",
      "There is a file avro37_allACAN_HIF1A.pickle\n",
      "There is a file avro37_allACAN_SIRT3.pickle\n",
      "There is a file avro37_allACAN_RNF213.pickle\n",
      "There is a file avro37_allACAN_KLF2.pickle\n",
      "There is a file avro37_allACAN_TBX4.pickle\n",
      "There is a file avro37_allACAN_SBF1.pickle\n",
      "There is a file avro37_allACAN_BMPR1A.pickle\n",
      "There is a file avro37_allACAN_TNF.pickle\n",
      "There is a file avro37_allACAN_SDHC.pickle\n",
      "There is a file avro37_allACAN_CTSL.pickle\n",
      "There is a file avro37_allACAN_PDGFD.pickle\n",
      "There is a file avro37_allACAN_SMAD1.pickle\n",
      "There is a file avro37_allACAN_EIF2AK4.pickle\n",
      "There is a file avro37_allACAN_SOX4.pickle\n",
      "There is a file avro37_allACAN_SOX17_PROMOTER_GENECARDS_2.pickle\n",
      "Loading avro37_allACAN_SOX17_PROMOTER_GENECARDS_2.pickle\n",
      "Gene SOX17_PROMOTER_GENECARDS_2 is already in /rds-d5/project/who1000/rds-who1000-wgs10k/user/tt419/Selene_data/selene_ftp_output/logs/mySelene_37_104genes-gnomad/predicted_SOX17_PROMOTER_GENECARDS_2.pickle\n",
      "There is a file avro37_allACAN_DNMT3A.pickle\n",
      "There is a file avro37_allACAN_FOXF1.pickle\n",
      "There is a file avro37_allACAN_ACVR1.pickle\n",
      "There is a file avro37_allACAN_AGGF1.pickle\n",
      "There is a file avro37_allACAN_FBN2.pickle\n",
      "There is a file avro37_allACAN_SOX18.pickle\n",
      "There is a file avro37_allACAN_SOX17.pickle\n",
      "Loading avro37_allACAN_SOX17.pickle\n",
      "Gene SOX17 is already in /rds-d5/project/who1000/rds-who1000-wgs10k/user/tt419/Selene_data/selene_ftp_output/logs/mySelene_37_104genes-gnomad/predicted_SOX17.pickle\n",
      "There is a file avro37_allACAN_SDHD.pickle\n",
      "There is a file avro37_allACAN_ACE2.pickle\n",
      "There is a file avro37_allACAN_TNFRSF13B.pickle\n",
      "There is a file avro37_allACAN_BDKRB2.pickle\n",
      "There is a file avro37_allACAN_SMYD1.pickle\n",
      "There is a file avro37_allACAN_BMPR2.pickle\n",
      "There is a file avro37_allACAN_FOXF2.pickle\n",
      "There is a file avro37_allACAN_CAV1.pickle\n",
      "There is a file avro37_allACAN_STRA6.pickle\n",
      "There is a file avro37_allACAN_SOX7.pickle\n",
      "There is a file avro37_allACAN_SMYD2.pickle\n",
      "There is a file avro37_allACAN_NOTCH2.pickle\n",
      "There is a file avro37_allACAN_FBN3.pickle\n",
      "There is a file avro37_allACAN_GDF2.pickle\n",
      "There is a file avro37_allACAN_SRM.pickle\n",
      "There is a file avro37_allACAN_KCNK3.pickle\n",
      "There is a file avro37_allACAN_SARS2.pickle\n",
      "There is a file avro37_allACAN_FH.pickle\n",
      "There is a file avro37_allACAN_FBLN2.pickle\n",
      "There is a file avro37_allACAN_TFAP2E.pickle\n",
      "There is a file avro37_allACAN_BMPR2_PROMOTER_GENECARDS_3.pickle\n",
      "There is a file avro37_allACAN_ADIPOQ.pickle\n",
      "There is a file avro37_allACAN_THBS1.pickle\n",
      "There is a file avro37_allACAN_SOX17_PROMOTER_GENECARDS_1.pickle\n",
      "Loading avro37_allACAN_SOX17_PROMOTER_GENECARDS_1.pickle\n",
      "Predicting for SOX17_PROMOTER_GENECARDS_1\n",
      "Iteration 100 done.\n",
      "Iteration 200 done.\n",
      "Iteration 300 done.\n",
      "Iteration 400 done.\n",
      "Iteration 500 done.\n",
      "Iteration 600 done.\n",
      "Iteration 700 done.\n",
      "Iteration 800 done.\n",
      "Iteration 900 done.\n",
      "Done in 937 steps.\n",
      "Done with SOX17_PROMOTER_GENECARDS_1\n",
      "It is now 2021-04-26 17:54:44.676473\n",
      "There is a file avro37_allACAN_SDHB.pickle\n",
      "There is a file avro37_allACAN_SMAD5.pickle\n",
      "There is a file avro37_allACAN_TET2.pickle\n",
      "There is a file avro37_allACAN_KDR.pickle\n",
      "There is a file avro37_allACAN_FOXO1.pickle\n",
      "There is a file avro37_allACAN_KCNA5.pickle\n",
      "There is a file avro37_allACAN_FBLN1.pickle\n",
      "There is a file avro37_allACAN_FBN1.pickle\n",
      "There is a file avro37_allACAN_FLNA.pickle\n",
      "There is a file avro37_allACAN_SOX17_PROMOTER_GENECARDS_3.pickle\n",
      "Loading avro37_allACAN_SOX17_PROMOTER_GENECARDS_3.pickle\n",
      "Predicting for SOX17_PROMOTER_GENECARDS_3\n",
      "Iteration 100 done.\n",
      "Iteration 200 done.\n",
      "Iteration 300 done.\n",
      "Iteration 400 done.\n",
      "Iteration 500 done.\n",
      "Iteration 600 done.\n",
      "Iteration 700 done.\n",
      "Iteration 800 done.\n",
      "Iteration 900 done.\n",
      "Iteration 1000 done.\n",
      "Done in 1014 steps.\n",
      "Done with SOX17_PROMOTER_GENECARDS_3\n",
      "It is now 2021-04-26 18:02:26.611226\n",
      "There is a file avro37_allACAN_CD248.pickle\n",
      "There is a file avro37_allACAN_EVI5.pickle\n",
      "There is a file avro37_allACAN_FNDC3B.pickle\n",
      "There is a file avro37_allACAN_SOD2.pickle\n",
      "There is a file avro37_allACAN_PARP1.pickle\n",
      "There is a file avro37_allACAN_NOTCH1.pickle\n",
      "There is a file avro37_allACAN_PRSS53.pickle\n",
      "There is a file avro37_allACAN_APOA4.pickle\n",
      "There is a file avro37_allACAN_SOX17_PROMOTER_CHICP.pickle\n",
      "Loading avro37_allACAN_SOX17_PROMOTER_CHICP.pickle\n",
      "Predicting for SOX17_PROMOTER_CHICP\n",
      "Found rs13266183\n",
      "Found rs10103692\n",
      "Iteration 100 done.\n",
      "Iteration 200 done.\n",
      "Iteration 300 done.\n",
      "Iteration 400 done.\n",
      "Iteration 500 done.\n",
      "Iteration 600 done.\n",
      "Iteration 700 done.\n",
      "Iteration 800 done.\n",
      "Iteration 900 done.\n",
      "Iteration 1000 done.\n",
      "Iteration 1100 done.\n",
      "Iteration 1200 done.\n",
      "Iteration 1300 done.\n",
      "Iteration 1400 done.\n",
      "Iteration 1500 done.\n",
      "Iteration 1600 done.\n",
      "Iteration 1700 done.\n",
      "Iteration 1800 done.\n",
      "Iteration 1900 done.\n",
      "Iteration 2000 done.\n",
      "Iteration 2100 done.\n",
      "Iteration 2200 done.\n",
      "Iteration 2300 done.\n",
      "Iteration 2400 done.\n",
      "Iteration 2500 done.\n",
      "Iteration 2600 done.\n",
      "Iteration 2700 done.\n",
      "Iteration 2800 done.\n",
      "Iteration 2900 done.\n",
      "Iteration 3000 done.\n",
      "Iteration 3100 done.\n",
      "Iteration 3200 done.\n",
      "Iteration 3300 done.\n",
      "Done in 3308 steps.\n",
      "Done with SOX17_PROMOTER_CHICP\n",
      "It is now 2021-04-26 18:29:31.741878\n",
      "There is a file avro37_allACAN_KLK3.pickle\n",
      "There is a file avro37_allACAN_IDH3G.pickle\n",
      "There is a file avro37_allACAN_TMPRSS2.pickle\n",
      "There is a file avro37_allACAN_SDHA.pickle\n",
      "There is a file avro37_allACAN_IL6R.pickle\n",
      "There is a file avro37_allACAN_ACVRL1.pickle\n",
      "There is a file avro37_allACAN_ICAM1.pickle\n",
      "There is a file avro37_allACAN_BDKRB1.pickle\n",
      "There is a file avro37_allACAN_GGCX.pickle\n",
      "There is a file avro37_allACAN_TGFBR1.pickle\n",
      "There is a file avro37_allACAN_ATP13A3.pickle\n",
      "There is a file avro37_allACAN_SMAD4.pickle\n",
      "There is a file avro37_allACAN_SLC6A19.pickle\n",
      "There is a file avro37_allACAN_ADAM17.pickle\n",
      "There is a file avro37_allACAN_APLNR.pickle\n",
      "There is a file avro37_allACAN_KLK2.pickle\n",
      "There is a file avro37_allACAN_TGFBR2.pickle\n",
      "There is a file avro37_allACAN_ACE.pickle\n",
      "There is a file avro37_allACAN_VCAM1.pickle\n",
      "There is a file avro37_allACAN_DCP1A.pickle\n",
      "There is a file avro37_allACAN_BRCA1.pickle\n",
      "There is a file avro37_allACAN_ABCA1.pickle\n",
      "There is a file avro37_allACAN_ENG.pickle\n",
      "There is a file avro37_allACAN_ADAM10.pickle\n",
      "There is a file avro37_allACAN_BMP10.pickle\n",
      "There is a file avro37_allACAN_NOTCH3.pickle\n",
      "There is a file avro37_allACAN_TMPRSS11D.pickle\n",
      "There is a file avro37_allACAN_KLK1.pickle\n",
      "There is a file avro37_allACAN_EGLN1.pickle\n",
      "There is a file avro37_allACAN_PAGE3.pickle\n",
      "There is a file avro37_allACAN_IL6.pickle\n",
      "There is a file avro37_allACAN_BMPR1B.pickle\n",
      "There is a file avro37_allACAN_APLN.pickle\n",
      "There is a file avro37_allACAN_BMPR2_PROMOTER_CHICP.pickle\n",
      "There is a file avro37_allACAN_SMAD9.pickle\n",
      "There is a file avro37_allACAN_AQP1.pickle\n"
     ]
    }
   ],
   "source": [
    "gene_dir = \"/home/tt419/Projects/data/104genes_37_acan/\"\n",
    "# Run with my model\n",
    "run(analysis ,gene_dir=gene_dir, new_fp_dir=\"/rds-d5/project/who1000/rds-who1000-wgs10k/user/tt419/Selene_data/selene_ftp_output/logs/mySelene_37_104genes-gnomad/\")\n",
    "\n",
    "\n",
    "# rs13266183 = [8, 55267612  , 55267612]\n",
    "# rs10103692 = [8, 55258127  , 55258127]\n",
    "\n",
    "# start = var_df[\"start\"].min() - frame_length\n",
    "# end = var_df[\"end\"].max() + frame_length\n",
    "# (chrom, start, end),get_reference(chrom, start, end)  \n",
    "\n",
    "\n",
    "# new_fp_dir=\"/rds-d5/project/who1000/rds-who1000-wgs10k/user/tt419/Selene_data/selene_ftp_output/logs/mySelene_37_104genes-gnomad/\"\n",
    "# print(f\"Running with gene_dir = {gene_dir}, and new_fp_dir = {new_fp_dir}\")\n",
    "# for root, dirs, files in os.walk(gene_dir):\n",
    "#     for filename in files:\n",
    "#         print(f\"There is a file {filename}\")\n",
    "#         if \".pickle\" in filename[-8:] and \"avro37_all\" in filename[:11]:\n",
    "#             try:\n",
    "#                 print(f\"Loading {filename}\")\n",
    "#                 gene = load_gene(os.path.join(gene_dir,filename))\n",
    "#                 gene_name = \"_\".join(filename.split(\"_\")[2:])[:-7]\n",
    "#                 new_fp = f\"{new_fp_dir}predicted_{gene_name}.pickle\"\n",
    "#                 if os.path.isfile(new_fp):\n",
    "#                     print(f\"Gene {gene_name} is already in {new_fp}\")\n",
    "#                     continue\n",
    "#                 print(f\"Predicting for {gene_name}\")\n",
    "#                 var_df = gene\n",
    "#                 coord, reference = get_ref_for_gene(var_df)\n",
    "#                 iter_count = 0\n",
    "#                 if reference == -1:\n",
    "#                     np.array([-1])\n",
    "#                 for frame,ref_frame in iterate_through_mutations_in_sequence(reference, coord[1], var_df):\n",
    "#                     encoded_frame = one_hot(frame)\n",
    "#                     encoded_ref_frame = one_hot(ref_frame)\n",
    "#                     sequences = np.stack([encoded_ref_frame.astype(int), encoded_frame.astype(int)], 0)\n",
    "#                     tmp = predict(analysis.model, sequences, use_cuda=analysis.use_cuda) \n",
    "#                     break\n",
    "#             except Exception as e:\n",
    "#                     print(f\"Something went wrong in {filename}\")\n",
    "                    \n",
    "#                     raise\n",
    "#         break\n",
    "#     break"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Selene",
   "language": "python",
   "name": "selene"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
