

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>selene.interpret &mdash; Selene master documentation</title>
  

  
  
  
  

  

  
  
    

  

  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Raleway" type="text/css" />
  <link rel="stylesheet" href="_static/css/selene_docs_theme.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="selene.predict" href="predict.html" />
    <link rel="prev" title="selene" href="selene.html" /> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> Selene
          

          
            
            <img src="_static/selene_logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                master (0.0.0)
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          

            
            
              
            
            
              <p class="caption"><span class="caption-text">Overview</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="overview/installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="overview/installation.html#installing-with-anaconda">Installing with Anaconda</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/installation.html#installing-from-source">Installing from source</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="overview/tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="overview/extending.html">Extending Selene</a></li>
</ul>
<p class="caption"><span class="caption-text">Packages</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="selene.html">selene</a><ul>
<li class="toctree-l2"><a class="reference internal" href="selene.html#trainmodel">TrainModel</a></li>
<li class="toctree-l2"><a class="reference internal" href="selene.html#evaluatemodel">EvaluateModel</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">selene.interpret</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#sequence-logo">sequence_logo</a></li>
<li class="toctree-l2"><a class="reference internal" href="#heatmap">heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="#ismresult">ISMResult</a></li>
<li class="toctree-l2"><a class="reference internal" href="#rescale-score-matrix">rescale_score_matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="#load-variant-abs-diff-scores">load_variant_abs_diff_scores</a></li>
<li class="toctree-l2"><a class="reference internal" href="#variant-diffs-scatter-plot">variant_diffs_scatter_plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="#ordered-variants-and-indices">ordered_variants_and_indices</a></li>
<li class="toctree-l2"><a class="reference internal" href="#sort-standard-chrs">sort_standard_chrs</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="predict.html">selene.predict</a><ul>
<li class="toctree-l2"><a class="reference internal" href="predict.html#analyzesequences">AnalyzeSequences</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="samplers.html">selene.samplers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="samplers.html#sampler">Sampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="samplers.html#onlinesampler">OnlineSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="samplers.html#intervalssampler">IntervalsSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="samplers.html#randompositionssampler">RandomPositionsSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="samplers.html#matfilesampler">MatFileSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="samplers.html#bedfilesampler">BedFileSampler</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="sequences.html">selene.sequences</a><ul>
<li class="toctree-l2"><a class="reference internal" href="sequences.html#sequence">Sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="sequences.html#genome">Genome</a></li>
<li class="toctree-l2"><a class="reference internal" href="sequences.html#proteome">Proteome</a></li>
<li class="toctree-l2"><a class="reference internal" href="sequences.html#sequence-to-encoding">sequence_to_encoding</a></li>
<li class="toctree-l2"><a class="reference internal" href="sequences.html#encoding-to-sequence">encoding_to_sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="sequences.html#get-reverse-encoding">get_reverse_encoding</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="targets.html">selene.targets</a><ul>
<li class="toctree-l2"><a class="reference internal" href="targets.html#target">Target</a></li>
<li class="toctree-l2"><a class="reference internal" href="targets.html#genomicfeatures">GenomicFeatures</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="utils.html">selene.utils</a><ul>
<li class="toctree-l2"><a class="reference internal" href="utils.html#nonstrandspecific">NonStrandSpecific</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#performancemetrics">PerformanceMetrics</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#initialize-logger">initialize_logger</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#load-features-list">load_features_list</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#load-model-from-state-dict">load_model_from_state_dict</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#get-indices-and-probabilities">get_indices_and_probabilities</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#configuration">Configuration</a></li>
</ul>
</li>
</ul>

            
          

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Selene</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>selene.interpret</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/interpret.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-selene.interpret">
<span id="selene-interpret"></span><h1>selene.interpret<a class="headerlink" href="#module-selene.interpret" title="Permalink to this headline">¶</a></h1>
<p>This module provides functions and classes for interpreting modules
trained with Selene.</p>
<div class="section" id="sequence-logo">
<h2>sequence_logo<a class="headerlink" href="#sequence-logo" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="selene.interpret.sequence_logo">
<code class="descclassname">selene.interpret.</code><code class="descname">sequence_logo</code><span class="sig-paren">(</span><em>score_matrix</em>, <em>order='value'</em>, <em>width=1.0</em>, <em>ax=None</em>, <em>sequence_type=&lt;class 'selene.sequences.genome.Genome'&gt;</em>, <em>font_properties=None</em>, <em>color_scheme=None</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/interpret/vis.html#sequence_logo"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.interpret.sequence_logo" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots a sequence logo for visualizing motifs.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first">
<li><p class="first"><strong>score_matrix</strong> (<em>np.ndarray</em>) – An <span class="math notranslate nohighlight">\(L \times N\)</span> array (where <span class="math notranslate nohighlight">\(L\)</span> is the length of
the sequence, and <span class="math notranslate nohighlight">\(N\)</span> is the size of the alphabet)
containing the scores for each base occuring at each position.</p>
</li>
<li><p class="first"><strong>order</strong> (<em>{‘alpha’, ‘value’}</em>) – The manner by which to sort the bases stacked at each position
in the sequence logo plot.</p>
<blockquote>
<div><ul class="simple">
<li>‘alpha’ - Bases go in the order they are found in the                      sequence alphabet.</li>
<li>‘value’ - Bases go in the order of their value, with the                        largest at the bottom.</li>
</ul>
</div></blockquote>
</li>
<li><p class="first"><strong>width</strong> (<em>float, optional</em>) – Default is 1. The width of each character in the plotted logo.
A value of 1 will mean that there is no gap between each the
characters at each position. A value of 0 will not draw any
characters.</p>
</li>
<li><p class="first"><strong>ax</strong> (<em>matplotlib.pyplot.Axes or None, optional</em>) – Default is <cite>None</cite>. The axes to plot on. If left as <cite>None</cite>, a new
axis will be created.</p>
</li>
<li><p class="first"><strong>sequence_type</strong> (<em>class, optional</em>) – Default is <cite>selene.sequences.Genome</cite>. The type of sequence that
the <em>in silico</em> mutagenesis results are associated with. This
should generally be a subclass of <cite>selene.sequences.Sequence</cite>.</p>
</li>
<li><p class="first"><strong>font_properties</strong> (<em>matplotlib.font_manager.FontProperties or None, optional</em>) – Default is <cite>None</cite>. A <cite>matplotlib.font_manager.FontProperties</cite>
object that specifies the properties of the font to use for
plotting the motif. If <cite>None</cite>, no font will be used, and the
text will be rendered by a path. This method of rendering paths
is  preferred, as it ensures all character heights correspond to
the actual values, and that there are no extra gaps between
the tops and bottoms of characters at each position in the
sequence logo. If the user opts to use a value other
than <cite>None</cite>, then no such guarantee can be made.</p>
</li>
<li><p class="first"><strong>color_scheme</strong> (<em>list(str) or None, optional</em>) – Default is <cite>None</cite>. A list containing the hex codes or names of
colors to use, appearing in the order of the bases of the
sequence type. If left as <cite>None</cite>, a default palette will be made
with <cite>seaborn.color_palette</cite>, and will have as many
colors as there are characters in the input sequence alphabet.</p>
</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">The axes containing the sequence logo plot.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first">matplotlib.pyplot.Axes</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal notranslate"><span class="pre">ValueError</span></code></a> – If the number of columns in <cite>score_matrix</cite> does not match the
number of characters in the alphabet of <cite>sequence_type</cite>.</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal notranslate"><span class="pre">ValueError</span></code></a> – If the number of colors in <cite>color_palette</cite> does not match the
number of characters in the alphabet of <cite>sequence_type</cite>.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<p>We have included an example of the output from a`sequence_logo`
plot below:</p>
<img alt="_images/sequence_logo_example.png" src="_images/sequence_logo_example.png" />
</dd></dl>

</div>
<div class="section" id="heatmap">
<h2>heatmap<a class="headerlink" href="#heatmap" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="selene.interpret.heatmap">
<code class="descclassname">selene.interpret.</code><code class="descname">heatmap</code><span class="sig-paren">(</span><em>score_matrix</em>, <em>mask=None</em>, <em>sequence_type=&lt;class 'selene.sequences.genome.Genome'&gt;</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/interpret/vis.html#heatmap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.interpret.heatmap" title="Permalink to this definition">¶</a></dt>
<dd><p>Plots the input matrix of scores, generally those produced by an
<em>in silico</em> mutagenesis experiment, on a heatmap.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first">
<li><p class="first"><strong>score_matrix</strong> (<em>numpy.ndarray</em>) – An <span class="math notranslate nohighlight">\(L \times N\)</span> array (where <span class="math notranslate nohighlight">\(L\)</span> is the length of
the sequence, and <span class="math notranslate nohighlight">\(N\)</span> is the size of the alphabet)
containing the scores for each base change at each position.</p>
</li>
<li><p class="first"><strong>mask</strong> (<em>numpy.ndarray, dtype=bool or None, optional</em>) – Default is <cite>None</cite>. An <span class="math notranslate nohighlight">\(L \times N\)</span> array (where <span class="math notranslate nohighlight">\(L\)</span>
is the length of the sequence, and <span class="math notranslate nohighlight">\(N\)</span> is the size of the
alphabet)  containing <cite>True</cite> at positions in the heatmap to
mask. If <cite>None</cite>, no masking will occur.</p>
</li>
<li><p class="first"><strong>sequence_type</strong> (<em>class, optional</em>) – Default is <cite>selene.sequences.Genome</cite>. The class of sequence that
the <em>in silico</em> mutagenesis results are associated with. This is
generally a sub-class of <cite>selene.sequences.Sequence</cite>.</p>
</li>
<li><p class="first"><strong>**kwargs</strong> (<em>dict</em>) – Keyword arguments to pass to <cite>seaborn.heatmap</cite>. Some useful ones
to remember are:</p>
<blockquote>
<div><ul class="simple">
<li>cbar_kws - Keyword arguments to forward to the colorbar.</li>
<li>yticklabels - Manipulate the tick labels on the y axis.</li>
<li>cbar - If <cite>False</cite>, hide the color bar. If <cite>True</cite>, show                     the colorbar.</li>
<li>cmap - The color map to use for the heatmap.</li>
</ul>
</div></blockquote>
</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">The axese containing the heatmap plot.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">matplotlib.pyplot.Axes</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="ismresult">
<h2>ISMResult<a class="headerlink" href="#ismresult" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="selene.interpret.ISMResult">
<em class="property">class </em><code class="descclassname">selene.interpret.</code><code class="descname">ISMResult</code><span class="sig-paren">(</span><em>data_frame</em>, <em>sequence_type=&lt;class 'selene.sequences.genome.Genome'&gt;</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/interpret/ism_result.html#ISMResult"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.interpret.ISMResult" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>An object storing the results of an <em>in silico</em> mutagenesis
experiment.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>data_frame</strong> (<em>pandas.DataFrame</em>) – The data frame with the results from the <em>in silico</em>
mutagenesis experiments.</li>
<li><strong>sequence_type</strong> (<em>class, optional</em>) – Default is <cite>selene.sequences.Genome</cite>. The type of sequence
that the <em>in silico</em> mutagenesis results are associated
with. This should generally be a subclass of
<cite>selene.sequences.Sequence</cite></li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal notranslate"><span class="pre">ValueError</span></code></a> – If the input data frame contains a base not included in the
alphabet of <cite>sequence_type</cite>.</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#Exception" title="(in Python v3.7)"><code class="xref py py-exc docutils literal notranslate"><span class="pre">Exception</span></code></a> – If multiple reference positions are specified in the input
data frame.</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#Exception" title="(in Python v3.7)"><code class="xref py py-exc docutils literal notranslate"><span class="pre">Exception</span></code></a> – If the input data does not contain scores for every mutation
at every position.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="staticmethod">
<dt id="selene.interpret.ISMResult.from_file">
<em class="property">static </em><code class="descname">from_file</code><span class="sig-paren">(</span><em>input_path</em>, <em>sequence_type=&lt;class 'selene.sequences.genome.Genome'&gt;</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/interpret/ism_result.html#ISMResult.from_file"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.interpret.ISMResult.from_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads a <cite>selene.interpret.ISMResult</cite> from a <cite>pandas.DataFrame</cite>
stored in a file of comma separated values (CSV).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>input_path</strong> (<em>str</em>) – A path to the file of comma separated input values.</li>
<li><strong>sequence_type</strong> (<em>class, optional</em>) – Default is <cite>selene.sequences.Genome</cite>. The type of sequence
that the <em>in silico</em> mutagenesis results are associated
with. This should generally be a subclass of
<cite>selene.sequences.Sequence</cite>.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">The <em>in silico</em> mutagenesis results that were stored in the
specified input file.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="#selene.interpret.ISMResult" title="selene.interpret.ISMResult">selene.interpret.ISMResult</a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="selene.interpret.ISMResult.get_score_matrix_for">
<code class="descname">get_score_matrix_for</code><span class="sig-paren">(</span><em>feature</em>, <em>reference_mask=None</em>, <em>dtype=&lt;class 'numpy.float64'&gt;</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/interpret/ism_result.html#ISMResult.get_score_matrix_for"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.interpret.ISMResult.get_score_matrix_for" title="Permalink to this definition">¶</a></dt>
<dd><p>Extracts a feature from the <em>in silico</em> mutagenesis results
as a matrix, where the reference base positions hold the value
for the reference prediction, and alternative positions hold the
results for making a one-base change from the reference base to
the specified alternative base.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>feature</strong> (<em>str</em>) – The name of the feature to extract as a matrix.</li>
<li><strong>reference_mask</strong> (<em>float or None, optional</em>) – Default is <cite>None</cite>. A value to mask the reference entries
with. If left as <cite>None</cite>, then no masking will be performed
on the reference positions.</li>
<li><strong>dtype</strong> (<em>numpy.dtype, optional</em>) – Default is <cite>numpy.float64</cite>. The data type to use for the
returned matrix.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">A <span class="math notranslate nohighlight">\(L \times N\)</span> shaped array (where <span class="math notranslate nohighlight">\(L\)</span> is the
sequence length, and <span class="math notranslate nohighlight">\(N\)</span> is the size of the alphabet
of <cite>sequence_type</cite>) that holds the results from the
<em>in silico</em> mutagenesis experiment for the specified
feature. The elements will be of type <cite>dtype</cite>.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first"><a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a></p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><p class="first last"><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal notranslate"><span class="pre">ValueError</span></code></a> – If the input data frame contains a base not included in the
alphabet of <cite>sequence_type</cite>.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="selene.interpret.ISMResult.reference_sequence">
<code class="descname">reference_sequence</code><a class="headerlink" href="#selene.interpret.ISMResult.reference_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>The reference sequence that the <em>in silico</em> mutagenesis
experiment was performed on.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">The reference sequence (i.e. non-mutated input) as a
string of characters.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)">str</a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="selene.interpret.ISMResult.sequence_type">
<code class="descname">sequence_type</code><a class="headerlink" href="#selene.interpret.ISMResult.sequence_type" title="Permalink to this definition">¶</a></dt>
<dd><p>The type of underlying sequence. This should generally be a
subclass of <cite>selene.sequences.Sequence</cite>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">The type of sequence that the <em>in silico</em> mutagenesis was
performed on.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">class</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="rescale-score-matrix">
<h2>rescale_score_matrix<a class="headerlink" href="#rescale-score-matrix" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="selene.interpret.rescale_score_matrix">
<code class="descclassname">selene.interpret.</code><code class="descname">rescale_score_matrix</code><span class="sig-paren">(</span><em>score_matrix</em>, <em>base_scaling='identity'</em>, <em>position_scaling='identity'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/interpret/vis.html#rescale_score_matrix"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.interpret.rescale_score_matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Performs base-wise and position-wise scaling of a score matrix for a
feature, usually produced from an <em>in silico</em> mutagenesis experiment.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first">
<li><p class="first"><strong>score_matrix</strong> (<em>numpy.ndarray</em>) – An <span class="math notranslate nohighlight">\(L \times N\)</span> matrix containing the scores for each
position, where <span class="math notranslate nohighlight">\(L\)</span> is the length of the sequence, and
<span class="math notranslate nohighlight">\(N\)</span> is the number of characters in the alphabet.</p>
</li>
<li><p class="first"><strong>base_scaling</strong> (<em>{‘identity’, ‘probability’, ‘max_effect’}</em>) –</p>
<p>The type of scaling performed on each base at a given position.</p>
<blockquote>
<div><ul class="simple">
<li>‘identity’ - No transformation will be applied to the                           data.</li>
<li>‘probability’ - The relative sizes of the bases will be                              the original input probabilities.</li>
<li>‘max_effect’ - The relative sizes of the bases will be                             the max effect of the original input                             values.</li>
</ul>
</div></blockquote>
</li>
<li><p class="first"><strong>position_scaling</strong> (<em>{‘identity’, ‘probability’, ‘max_effect’}</em>) –</p>
<p>The type of scaling performed on each position.</p>
<blockquote>
<div><ul class="simple">
<li>‘identity’    - No transformation will be applied to the data.</li>
<li>‘probability’ - The sum of values at a position will be                              equal to the sum of the original input                              values at that position.</li>
<li>‘max_effect’  - The sum of values at a position will be                              equal to the sum of the max effect values                              of the original input values at that                              position.</li>
</ul>
</div></blockquote>
</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">The transformed score matrix.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first"><a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a></p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><p class="first last"><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal notranslate"><span class="pre">ValueError</span></code></a> – If an unsupported <cite>base_scaling</cite> or <cite>position_scaling</cite> is
entered.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="load-variant-abs-diff-scores">
<h2>load_variant_abs_diff_scores<a class="headerlink" href="#load-variant-abs-diff-scores" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="selene.interpret.load_variant_abs_diff_scores">
<code class="descclassname">selene.interpret.</code><code class="descname">load_variant_abs_diff_scores</code><span class="sig-paren">(</span><em>input_path</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/interpret/vis.html#load_variant_abs_diff_scores"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.interpret.load_variant_abs_diff_scores" title="Permalink to this definition">¶</a></dt>
<dd><p>Loads the variant data, labels, and feature names from a diff scores
file output from variant effect prediction.</p>
<p>TODO: should we move this out of <cite>vis.py</cite>?</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>input_path</strong> (<em>str</em>) – Path to the input file.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><ul class="simple">
<li><cite>tuple[0]</cite> is the matrix of absolute difference scores. The rows
are the variants and the columns are the features for which the
model makes predictions.</li>
<li><cite>tuple[1]</cite> is the list of variant labels. Each tuple contains
(chrom, pos, name, ref, alt).</li>
<li><cite>tuple[2]</cite> is the list of features.</li>
</ul>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.7)">tuple</a>(np.ndarray, <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.7)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.7)">tuple</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)">str</a>)), <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.7)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)">str</a>))</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="variant-diffs-scatter-plot">
<h2>variant_diffs_scatter_plot<a class="headerlink" href="#variant-diffs-scatter-plot" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="selene.interpret.variant_diffs_scatter_plot">
<code class="descclassname">selene.interpret.</code><code class="descname">variant_diffs_scatter_plot</code><span class="sig-paren">(</span><em>data</em>, <em>labels</em>, <em>features</em>, <em>output_path</em>, <em>filter_features=None</em>, <em>labels_sort_fn=&lt;function ordered_variants_and_indices&gt;</em>, <em>nth_percentile=None</em>, <em>hg_reference_version=None</em>, <em>threshold_line=None</em>, <em>auto_open=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/interpret/vis.html#variant_diffs_scatter_plot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.interpret.variant_diffs_scatter_plot" title="Permalink to this definition">¶</a></dt>
<dd><p>Displays each variant’s max probability difference across features
as a point in a scatter plot. The points in the scatter plot are
ordered by the variant chromosome and position by default. Variants can
be sorted differently by passing in a new <cite>labels_sort_fn</cite>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>data</strong> (<em>np.ndarray</em>) – Absolute difference scores for variants across all features that a
model predicts. This is the first value in the tuple returned by
<cite>load_variant_abs_diff_scores</cite>.</li>
<li><strong>labels</strong> (<em>list(tuple(str))</em>) – A list of variant labels. This is the second value in the tuple
returned by <cite>load_variant_abs_diff_scores</cite>.</li>
<li><strong>features</strong> (<em>list(str)</em>) – A list of the features the model predicts. This is the third value
in the tuple returned by <cite>load_variant_abs_diff_scores</cite>.</li>
<li><strong>output_path</strong> (<em>str</em>) – Path to output file. Must have ‘.html’ extension.</li>
<li><strong>filter_features</strong> (<em>types.FunctionType or None, optional</em>) – Default is None. A function that takes in a <cite>list(str)</cite> of features
and returns the <cite>list(int)</cite> of feature indices over which we would
compute the max(probability difference) for each variant. For example,
a user may only want to visualize the max probability difference
for TF binding features. If <cite>None</cite>, uses all the features.</li>
<li><strong>labels_sort_fn</strong> (<em>types.FunctionType, optional</em>) – Default is <cite>ordered_variants_and_indices</cite>. A function that takes
in a <cite>list(tuple(str))</cite> of labels corresponding to the rows in <cite>data</cite>
and returns a <cite>tuple(list(tuple), list(int))</cite>, where the first value
is the ordered list of variant labels and the second value is the
ordered list of indices for those variant labels. By default,
variants are sorted by chromosome and position.</li>
<li><strong>nth_percentile</strong> (<em>int [0, 100] or None, optional</em>) – Default is None. If <cite>nth_percentile</cite> is not None, only displays the
variants with a max absolute difference score within the
<cite>nth_percentile</cite> of scores.</li>
<li><strong>hg_reference_version</strong> (<em>str {“hg19”, “hg38”} or None, optional</em>) – Default is None. On hover, we can display the gene(s) closest to
each variant if <cite>hg_reference_version</cite> is not None, where closest
can be a variant within a gene interval (where genes and
their coordinates are taken from level 1 &amp; 2 protein-coding genes
in gencode v28) or near a gene. In the future, we will allow users
to specify their own genome file so that this information can
be annotated to variants from other organisms, other genome versions,
etc.</li>
<li><strong>threshold_line</strong> (<em>float or None, optional</em>) – Default is None. If <cite>threshold_line</cite> is not None, draws a horizontal
line at the specified threshold. Helps focus the visual on variants
above a certain threshold.</li>
<li><strong>auto_open</strong> (<em>bool, optional</em>) – Default is False. If <cite>auto_open</cite>, will automatically open a web
browser that displays the plotted HTML file.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">The generated Plotly figure.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">plotly.graph_objs.graph_objs.Figure</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="ordered-variants-and-indices">
<h2>ordered_variants_and_indices<a class="headerlink" href="#ordered-variants-and-indices" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="selene.interpret.ordered_variants_and_indices">
<code class="descclassname">selene.interpret.</code><code class="descname">ordered_variants_and_indices</code><span class="sig-paren">(</span><em>labels</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/interpret/vis.html#ordered_variants_and_indices"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.interpret.ordered_variants_and_indices" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the ordered variant labels, where the labels are sorted by chromosome
and position, and the indices corresponding to the sort,</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>labels</strong> (<em>list(tuple(str))</em>) – The list of variant labels. Each label is a tuple of
(chrom, pos, name, ref, alt).</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">The first value is the ordered list of labels. Each label
is a tuple of (chrom, pos, ref, alt). The second value
is the ordered list of label indices.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.7)">tuple</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.7)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.7)">tuple</a>), <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.7)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.7)">int</a>))</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="sort-standard-chrs">
<h2>sort_standard_chrs<a class="headerlink" href="#sort-standard-chrs" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="selene.interpret.sort_standard_chrs">
<code class="descclassname">selene.interpret.</code><code class="descname">sort_standard_chrs</code><span class="sig-paren">(</span><em>chrom</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/interpret/vis.html#sort_standard_chrs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.interpret.sort_standard_chrs" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the value on which the
standard chromosomes can be sorted.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>chrom</strong> (<em>str</em>) – The chromosome</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">The value on which to sort</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.7)">int</a></td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="predict.html" class="btn btn-neutral float-right" title="selene.predict" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="selene.html" class="btn btn-neutral" title="selene" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Selene Contributors.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'master',
            LANGUAGE:'None',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  <script type="text/javascript" src="_static/js/theme.js"></script>

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>
 



</body>
</html>