

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>selene_sdk.samplers &mdash; Selene master documentation</title>
  

  
  
  
  

  

  
  
    

  

  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Raleway" type="text/css" />
  <link rel="stylesheet" href="_static/css/selene_docs_theme.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="selene_sdk.sequences" href="sequences.html" />
    <link rel="prev" title="selene_sdk.predict" href="predict.html" /> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> Selene
          

          
            
            <img src="_static/selene_logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                master (0.0.0)
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          

            
            
              
            
            
              <p class="caption"><span class="caption-text">Overview</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="overview/installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="overview/installation.html#installing-with-anaconda">Installing with Anaconda</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview/installation.html#installing-from-source">Installing from source</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="overview/tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="overview/extending.html">Extending Selene</a></li>
</ul>
<p class="caption"><span class="caption-text">Packages</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="selene.html">selene_sdk</a><ul>
<li class="toctree-l2"><a class="reference internal" href="selene.html#trainmodel">TrainModel</a></li>
<li class="toctree-l2"><a class="reference internal" href="selene.html#evaluatemodel">EvaluateModel</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="interpret.html">selene_sdk.interpret</a><ul>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#sequence-logo">sequence_logo</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#heatmap">heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#ismresult">ISMResult</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#rescale-score-matrix">rescale_score_matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#load-variant-abs-diff-scores">load_variant_abs_diff_scores</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#variant-diffs-scatter-plot">variant_diffs_scatter_plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#ordered-variants-and-indices">ordered_variants_and_indices</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#sort-standard-chrs">sort_standard_chrs</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="predict.html">selene_sdk.predict</a><ul>
<li class="toctree-l2"><a class="reference internal" href="predict.html#analyzesequences">AnalyzeSequences</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">selene_sdk.samplers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#sampler">Sampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="#onlinesampler">OnlineSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="#intervalssampler">IntervalsSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="#randompositionssampler">RandomPositionsSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="#matfilesampler">MatFileSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="#bedfilesampler">BedFileSampler</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="sequences.html">selene_sdk.sequences</a><ul>
<li class="toctree-l2"><a class="reference internal" href="sequences.html#sequence">Sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="sequences.html#genome">Genome</a></li>
<li class="toctree-l2"><a class="reference internal" href="sequences.html#proteome">Proteome</a></li>
<li class="toctree-l2"><a class="reference internal" href="sequences.html#sequence-to-encoding">sequence_to_encoding</a></li>
<li class="toctree-l2"><a class="reference internal" href="sequences.html#encoding-to-sequence">encoding_to_sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="sequences.html#get-reverse-encoding">get_reverse_encoding</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="targets.html">selene_sdk.targets</a><ul>
<li class="toctree-l2"><a class="reference internal" href="targets.html#target">Target</a></li>
<li class="toctree-l2"><a class="reference internal" href="targets.html#genomicfeatures">GenomicFeatures</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="utils.html">selene_sdk.utils</a><ul>
<li class="toctree-l2"><a class="reference internal" href="utils.html#nonstrandspecific">NonStrandSpecific</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#performancemetrics">PerformanceMetrics</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#initialize-logger">initialize_logger</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#load-features-list">load_features_list</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#load-model-from-state-dict">load_model_from_state_dict</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#get-indices-and-probabilities">get_indices_and_probabilities</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#configuration">Configuration</a></li>
</ul>
</li>
</ul>

            
          

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Selene</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>selene_sdk.samplers</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/samplers.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-selene_sdk.samplers">
<span id="selene-sdk-samplers"></span><h1>selene_sdk.samplers<a class="headerlink" href="#module-selene_sdk.samplers" title="Permalink to this headline">¶</a></h1>
<p>This module provides classes and methods for sampling labeled data
examples.</p>
<div class="section" id="sampler">
<h2>Sampler<a class="headerlink" href="#sampler" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="selene_sdk.samplers.Sampler">
<em class="property">class </em><code class="descclassname">selene_sdk.samplers.</code><code class="descname">Sampler</code><span class="sig-paren">(</span><em>seed=436</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/sampler.html#Sampler"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.Sampler" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>The base class for sampler currently enforces that all samplers
have modes for drawing training and validation samples to train a
model.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Variables:</th><td class="field-body"><ul class="first simple">
<li><strong>modes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.7)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)"><em>str</em></a><em>)</em>) – A list of the names of the modes that the object may operate in.</li>
<li><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)"><em>str</em></a><em> or </em><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.7)"><em>None</em></a>) – Default is <cite>None</cite>. The current mode that the object is operating in.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first last"><strong>seed</strong> (<em>int</em>) – The value used to seed the random number generator.</p>
</td>
</tr>
</tbody>
</table>
<dl class="attribute">
<dt id="selene_sdk.samplers.Sampler.BASE_MODES">
<code class="descname">BASE_MODES</code><em class="property"> = ('train', 'validate')</em><a class="headerlink" href="#selene_sdk.samplers.Sampler.BASE_MODES" title="Permalink to this definition">¶</a></dt>
<dd><p>The types of modes that the <cite>Sampler</cite> object can run in.</p>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.Sampler.get_data_and_targets">
<code class="descname">get_data_and_targets</code><span class="sig-paren">(</span><em>mode</em>, <em>batch_size</em>, <em>n_samples</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/sampler.html#Sampler.get_data_and_targets"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.Sampler.get_data_and_targets" title="Permalink to this definition">¶</a></dt>
<dd><p>This method fetches a subset of the data from the sampler,
divided into batches. This method also allows the user to
specify what operating mode to run the sampler in when fetching
the data.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>mode</strong> (<em>str</em>) – The operating mode that the object should run in.</li>
<li><strong>batch_size</strong> (<em>int</em>) – The size of the batches to divide the data into.</li>
<li><strong>n_samples</strong> (<em>int</em>) – The total number of samples to retrieve.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.Sampler.get_validation_set">
<code class="descname">get_validation_set</code><span class="sig-paren">(</span><em>batch_size</em>, <em>n_samples=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/sampler.html#Sampler.get_validation_set"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.Sampler.get_validation_set" title="Permalink to this definition">¶</a></dt>
<dd><p>This method returns a subset of validation data from the
sampler, divided into batches.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>batch_size</strong> (<em>int</em>) – The size of the batches to divide the data into.</li>
<li><strong>n_samples</strong> (<em>int, optional</em>) – Default is None. The total number of validation examples to
retrieve. Handling for <cite>n_samples=None</cite> should be done by
all classes that subclass <cite>selene.samplers.Sampler</cite>.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.Sampler.sample">
<code class="descname">sample</code><span class="sig-paren">(</span><em>batch_size=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/sampler.html#Sampler.sample"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.Sampler.sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Fetches a mini-batch of the data from the sampler.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>batch_size</strong> (<em>int, optional</em>) – Default is 1. The size of the batch to retrieve.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.Sampler.set_mode">
<code class="descname">set_mode</code><span class="sig-paren">(</span><em>mode</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/sampler.html#Sampler.set_mode"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.Sampler.set_mode" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the sampling mode.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>mode</strong> (<em>str</em>) – The name of the mode to use. It must be one of
<cite>Sampler.BASE_MODES</cite>.</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal notranslate"><span class="pre">ValueError</span></code></a> – If <cite>mode</cite> is not a valid mode.</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="onlinesampler">
<h2>OnlineSampler<a class="headerlink" href="#onlinesampler" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="selene_sdk.samplers.OnlineSampler">
<em class="property">class </em><code class="descclassname">selene_sdk.samplers.</code><code class="descname">OnlineSampler</code><span class="sig-paren">(</span><em>reference_sequence, target_path, features, seed=436, validation_holdout=['chr6', 'chr7'], test_holdout=['chr8', 'chr9'], sequence_length=1001, center_bin_to_predict=201, feature_thresholds=0.5, mode='train', save_datasets=[], output_dir=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/online_sampler.html#OnlineSampler"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.OnlineSampler" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">selene_sdk.samplers.sampler.Sampler</span></code></p>
<p>A sampler in which training/validation/test data is constructed
from random sampling of the dataset for each batch passed to the
model. This form of sampling may alleviate the problem of loading an
extremely large dataset into memory when developing a new model.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>reference_sequence</strong> (<em>selene.sequences.Sequence</em>) – A reference sequence from which to create examples.</li>
<li><strong>target_path</strong> (<em>str</em>) – Path to tabix-indexed, compressed BED file (<cite>*.bed.gz</cite>) of genomic
coordinates mapped to the genomic features we want to predict.</li>
<li><strong>features</strong> (<em>list(str)</em>) – List of distinct features that we aim to predict.</li>
<li><strong>seed</strong> (<em>int, optional</em>) – Default is 436. Sets the random seed for sampling.</li>
<li><strong>validation_holdout</strong> (<em>list(str) or float, optional</em>) – Default is <cite>[‘chr6’, ‘chr7’]</cite>. Holdout can be regional or
proportional. If regional, expects a list (e.g. <cite>[‘X’, ‘Y’]</cite>).
Regions must match those specified in the first column of the
tabix-indexed BED file. If proportional, specify a percentage
between (0.0, 1.0). Typically 0.10 or 0.20.</li>
<li><strong>test_holdout</strong> (<em>list(str) or float, optional</em>) – Default is <cite>[‘chr8’, ‘chr9’]</cite>. See documentation for
<cite>validation_holdout</cite> for additional information.</li>
<li><strong>sequence_length</strong> (<em>int, optional</em>) – Default is 1000. Model is trained on sequences of <cite>sequence_length</cite>
where genomic features are annotated to the center regions of
these sequences.</li>
<li><strong>center_bin_to_predict</strong> (<em>int, optional</em>) – Default is 200. Query the tabix-indexed file for a region of
length <cite>center_bin_to_predict</cite>.</li>
<li><strong>feature_thresholds</strong> (<em>float [0.0, 1.0], optional</em>) – Default is 0.5. The <cite>feature_threshold</cite> to pass to the
<cite>GenomicFeatures</cite> object.</li>
<li><strong>mode</strong> (<em>{‘train’, ‘validate’, ‘test’}, optional</em>) – Default is <cite>‘train’</cite>. The mode to run the sampler in.</li>
<li><strong>save_datasets</strong> (<em>list(str), optional</em>) – Default is <cite>[]</cite> the empty list. The list of modes for which we should
save the sampled data to file (e.g. <cite>[“test”, “validate”]</cite>).</li>
<li><strong>output_dir</strong> (<em>str or None, optional</em>) – Default is None. The path to the directory where we should
save sampled examples for a mode. If <cite>save_datasets</cite> is
a non-empty list, <cite>output_dir</cite> must be specified. If
the path in <cite>output_dir</cite> does not exist it will be created
automatically.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Variables:</th><td class="field-body"><ul class="first simple">
<li><strong>reference_sequence</strong> (<em>selene.sequences.Sequence</em>) – The reference sequence that examples are created from.</li>
<li><strong>target</strong> (<em>selene.targets.Target</em>) – The <cite>selene.targets.Target</cite> object holding the features that we
would like to predict.</li>
<li><strong>validation_holdout</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.7)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)"><em>str</em></a><em>) or </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.7)"><em>float</em></a>) – The samples to hold out for validating model performance. These
can be “regional” or “proportional”. If regional, this is a list
of region names (e.g. <cite>[‘chrX’, ‘chrY’]</cite>). These regions must
match those specified in the first column of the tabix-indexed
BED file. If proportional, this is the fraction of total samples
that will be held out.</li>
<li><strong>test_holdout</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.7)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)"><em>str</em></a><em>) or </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.7)"><em>float</em></a>) – The samples to hold out for testing model performance. See the
documentation for <cite>validation_holdout</cite> for more details.</li>
<li><strong>sequence_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.7)"><em>int</em></a>) – The length of the sequences to  train the model on.</li>
<li><strong>bin_radius</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.7)"><em>int</em></a>) – From the center of the sequence, the radius in which to detect
a feature annotation in order to include it as a sample’s label.</li>
<li><strong>surrounding_sequence_radius</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.7)"><em>int</em></a>) – The length of sequence falling outside of the feature detection
bin (i.e. <cite>bin_radius</cite>) center, but still within the
<cite>sequence_length</cite>.</li>
<li><strong>modes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.7)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)"><em>str</em></a><em>)</em>) – The list of modes that the sampler can be run in.</li>
<li><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)"><em>str</em></a>) – The current mode that the sampler is running in. Must be one of
the modes listed in <cite>modes</cite>.</li>
</ul>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal notranslate"><span class="pre">ValueError</span></code></a> – If <cite>mode</cite> is not a valid mode.</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal notranslate"><span class="pre">ValueError</span></code></a> – If the parities of <cite>sequence_length</cite> and <cite>center_bin_to_predict</cite>
are not the same.</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal notranslate"><span class="pre">ValueError</span></code></a> – If <cite>sequence_length</cite> is smaller than <cite>center_bin_to_predict</cite> is.</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal notranslate"><span class="pre">ValueError</span></code></a> – If the types of <cite>validation_holdout</cite> and <cite>test_holdout</cite> are not
the same.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="attribute">
<dt id="selene_sdk.samplers.OnlineSampler.STRAND_SIDES">
<code class="descname">STRAND_SIDES</code><em class="property"> = ('+', '-')</em><a class="headerlink" href="#selene_sdk.samplers.OnlineSampler.STRAND_SIDES" title="Permalink to this definition">¶</a></dt>
<dd><p>Defines the strands that features can be sampled from.</p>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.OnlineSampler.get_data_and_targets">
<code class="descname">get_data_and_targets</code><span class="sig-paren">(</span><em>mode</em>, <em>batch_size</em>, <em>n_samples</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/online_sampler.html#OnlineSampler.get_data_and_targets"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.OnlineSampler.get_data_and_targets" title="Permalink to this definition">¶</a></dt>
<dd><p>This method fetches a subset of the data from the sampler,
divided into batches. This method also allows the user to
specify what operating mode to run the sampler in when fetching
the data.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>mode</strong> (<em>str</em>) – The mode to run the sampler in when fetching the samples.
See <cite>selene.samplers.IntervalsSampler.modes</cite> for more
information.</li>
<li><strong>batch_size</strong> (<em>int</em>) – The size of the batches to divide the data into.</li>
<li><strong>n_samples</strong> (<em>int</em>) – The total number of samples to retrieve.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>sequences_and_targets, targets_matrix</strong> – Tuple containing the list of sequence-target pairs, as well
as a single matrix with all targets in the same order.
Note that <cite>sequences_and_targets</cite>’s sequence elements are of
the shape <span class="math notranslate nohighlight">\(B \times L \times N\)</span> and its target
elements are of the shape <span class="math notranslate nohighlight">\(B \times F\)</span>, where
<span class="math notranslate nohighlight">\(B\)</span> is <cite>batch_size</cite>, <span class="math notranslate nohighlight">\(L\)</span> is the sequence length,
<span class="math notranslate nohighlight">\(N\)</span> is the size of the sequence type’s alphabet, and
<span class="math notranslate nohighlight">\(F\)</span> is the number of features. Further,
<cite>target_matrix</cite> is of the shape <span class="math notranslate nohighlight">\(S \times F\)</span>, where
<span class="math notranslate nohighlight">\(S =\)</span> <cite>n_samples</cite>.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.7)">tuple</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.7)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.7)">tuple</a>(<a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>, <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>)), <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>)</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.OnlineSampler.get_dataset_in_batches">
<code class="descname">get_dataset_in_batches</code><span class="sig-paren">(</span><em>mode</em>, <em>batch_size</em>, <em>n_samples=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/online_sampler.html#OnlineSampler.get_dataset_in_batches"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.OnlineSampler.get_dataset_in_batches" title="Permalink to this definition">¶</a></dt>
<dd><p>This method returns a subset of the data for a specified run
mode, divided into mini-batches.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>mode</strong> (<em>str</em>) – The mode to run the sampler in when fetching the samples.
See <cite>selene.samplers.IntervalsSampler.modes</cite> for more
information.</li>
<li><strong>batch_size</strong> (<em>int</em>) – The size of the batches to divide the data into.</li>
<li><strong>n_samples</strong> (<em>int or None, optional</em>) – Default is <cite>None</cite>. The total number of samples to retrieve.
If <cite>None</cite>, it will retrieve 32000 samples if <cite>mode</cite> is validate
or 640000 samples if <cite>mode</cite> is test or train.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>sequences_and_targets, targets_matrix</strong> – Tuple containing the list of sequence-target pairs, as well
as a single matrix with all targets in the same order.
The list is length <span class="math notranslate nohighlight">\(S\)</span>, where <span class="math notranslate nohighlight">\(S =\)</span> <cite>n_samples</cite>.
Note that <cite>sequences_and_targets</cite>’s sequence elements are of
the shape <span class="math notranslate nohighlight">\(B \times L \times N\)</span> and its target
elements are of the shape <span class="math notranslate nohighlight">\(B \times F\)</span>, where
<span class="math notranslate nohighlight">\(B\)</span> is <cite>batch_size</cite>, <span class="math notranslate nohighlight">\(L\)</span> is the sequence length,
<span class="math notranslate nohighlight">\(N\)</span> is the size of the sequence type’s alphabet, and
<span class="math notranslate nohighlight">\(F\)</span> is the number of features. Further,
<cite>target_matrix</cite> is of the shape <span class="math notranslate nohighlight">\(S \times F\)</span></p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.7)">tuple</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.7)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.7)">tuple</a>(<a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>, <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>)), <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>)</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.OnlineSampler.get_feature_from_index">
<code class="descname">get_feature_from_index</code><span class="sig-paren">(</span><em>index</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/online_sampler.html#OnlineSampler.get_feature_from_index"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.OnlineSampler.get_feature_from_index" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the feature corresponding to an index in the feature
vector.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>index</strong> (<em>int</em>) – The index of the feature to retrieve the name for.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">The name of the feature occurring at the specified index.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)">str</a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.OnlineSampler.get_sequence_from_encoding">
<code class="descname">get_sequence_from_encoding</code><span class="sig-paren">(</span><em>encoding</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/online_sampler.html#OnlineSampler.get_sequence_from_encoding"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.OnlineSampler.get_sequence_from_encoding" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the string sequence from the one-hot encoding
of the sequence.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>encoding</strong> (<em>numpy.ndarray</em>) – An <span class="math notranslate nohighlight">\(L \times N\)</span> array (where <span class="math notranslate nohighlight">\(L\)</span> is the length
of the sequence and <span class="math notranslate nohighlight">\(N\)</span> is the size of the sequence
type’s alphabet) containing the one-hot encoding of the
sequence.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">The sequence of <span class="math notranslate nohighlight">\(L\)</span> characters decoded from the input.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)">str</a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.OnlineSampler.get_test_set">
<code class="descname">get_test_set</code><span class="sig-paren">(</span><em>batch_size</em>, <em>n_samples=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/online_sampler.html#OnlineSampler.get_test_set"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.OnlineSampler.get_test_set" title="Permalink to this definition">¶</a></dt>
<dd><p>This method returns a subset of testing data from the
sampler, divided into batches.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>batch_size</strong> (<em>int</em>) – The size of the batches to divide the data into.</li>
<li><strong>n_samples</strong> (<em>int or None, optional</em>) – Default is <cite>None</cite>. The total number of validation examples
to retrieve. If <cite>None</cite>, 640000 examples are retrieved.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>sequences_and_targets, targets_matrix</strong> – Tuple containing the list of sequence-target pairs, as well
as a single matrix with all targets in the same order.
Note that <cite>sequences_and_targets</cite>’s sequence elements are of
the shape <span class="math notranslate nohighlight">\(B \times L \times N\)</span> and its target
elements are of the shape <span class="math notranslate nohighlight">\(B \times F\)</span>, where
<span class="math notranslate nohighlight">\(B\)</span> is <cite>batch_size</cite>, <span class="math notranslate nohighlight">\(L\)</span> is the sequence length,
<span class="math notranslate nohighlight">\(N\)</span> is the size of the sequence type’s alphabet, and
<span class="math notranslate nohighlight">\(F\)</span> is the number of features. Further,
<cite>target_matrix</cite> is of the shape <span class="math notranslate nohighlight">\(S \times F\)</span>, where
<span class="math notranslate nohighlight">\(S =\)</span> <cite>n_samples</cite>.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.7)">tuple</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.7)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.7)">tuple</a>(<a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>, <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>)), <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>)</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><p class="first last"><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal notranslate"><span class="pre">ValueError</span></code></a> – If no test partition of the data was specified during
sampler initialization.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.OnlineSampler.get_validation_set">
<code class="descname">get_validation_set</code><span class="sig-paren">(</span><em>batch_size</em>, <em>n_samples=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/online_sampler.html#OnlineSampler.get_validation_set"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.OnlineSampler.get_validation_set" title="Permalink to this definition">¶</a></dt>
<dd><p>This method returns a subset of validation data from the
sampler, divided into batches.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>batch_size</strong> (<em>int</em>) – The size of the batches to divide the data into.</li>
<li><strong>n_samples</strong> (<em>int or None, optional</em>) – Default is <cite>None</cite>. The total number of validation examples
to retrieve. If <cite>None</cite>, 32000 examples are retrieved.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>sequences_and_targets, targets_matrix</strong> – Tuple containing the list of sequence-target pairs, as well
as a single matrix with all targets in the same order.
Note that <cite>sequences_and_targets</cite>’s sequence elements are of
the shape <span class="math notranslate nohighlight">\(B \times L \times N\)</span> and its target
elements are of the shape <span class="math notranslate nohighlight">\(B \times F\)</span>, where
<span class="math notranslate nohighlight">\(B\)</span> is <cite>batch_size</cite>, <span class="math notranslate nohighlight">\(L\)</span> is the sequence length,
<span class="math notranslate nohighlight">\(N\)</span> is the size of the sequence type’s alphabet, and
<span class="math notranslate nohighlight">\(F\)</span> is the number of features. Further,
<cite>target_matrix</cite> is of the shape <span class="math notranslate nohighlight">\(S \times F\)</span>, where
<span class="math notranslate nohighlight">\(S =\)</span> <cite>n_samples</cite>.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.7)">tuple</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.7)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.7)">tuple</a>(<a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>, <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>)), <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>)</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.OnlineSampler.save_dataset_to_file">
<code class="descname">save_dataset_to_file</code><span class="sig-paren">(</span><em>mode</em>, <em>close_filehandle=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/online_sampler.html#OnlineSampler.save_dataset_to_file"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.OnlineSampler.save_dataset_to_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Save samples for each partition (i.e. train/validate/test) to
disk.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>mode</strong> (<em>str</em>) – Must be one of the modes specified in <cite>save_datasets</cite> during
sampler initialization.</li>
<li><strong>close_filehandle</strong> (<em>bool, optional</em>) – Default is False. <cite>close_filehandle=True</cite> assumes that all
data corresponding to the input <cite>mode</cite> has been saved to
file and <cite>save_dataset_to_file</cite> will not be called with
<cite>mode</cite> again.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="intervalssampler">
<h2>IntervalsSampler<a class="headerlink" href="#intervalssampler" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="selene_sdk.samplers.IntervalsSampler">
<em class="property">class </em><code class="descclassname">selene_sdk.samplers.</code><code class="descname">IntervalsSampler</code><span class="sig-paren">(</span><em>reference_sequence, target_path, features, intervals_path, sample_negative=False, seed=436, validation_holdout=['chr6', 'chr7'], test_holdout=['chr8', 'chr9'], sequence_length=1000, center_bin_to_predict=200, feature_thresholds=0.5, mode='train', save_datasets=['test'], output_dir=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/intervals_sampler.html#IntervalsSampler"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.IntervalsSampler" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">selene_sdk.samplers.online_sampler.OnlineSampler</span></code></p>
<p>Draws samples from pre-specified windows in the reference sequence.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>reference_sequence</strong> (<em>selene.sequences.Sequence</em>) – A reference sequence from which to create examples.</li>
<li><strong>target_path</strong> (<em>str</em>) – Path to tabix-indexed, compressed BED file (<cite>*.bed.gz</cite>) of genomic
coordinates mapped to the genomic features we want to predict.</li>
<li><strong>features</strong> (<em>list(str)</em>) – List of distinct features that we aim to predict.</li>
<li><strong>intervals_path</strong> (<em>str</em>) – The path to the file that contains the intervals to sample from.
In this file, each interval should occur on a separate line.</li>
<li><strong>sample_negative</strong> (<em>bool, optional</em>) – Default is <cite>False</cite>. This tells the sampler whether negative
examples (i.e. with no positive labels) should be drawn when
generating samples. If <cite>True</cite>, both negative and positive
samples will be drawn. If <cite>False</cite>, only samples with at least
one positive label will be drawn.</li>
<li><strong>seed</strong> (<em>int, optional</em>) – Default is 436. Sets the random seed for sampling.</li>
<li><strong>validation_holdout</strong> (<em>list(str) or float, optional</em>) – Default is <cite>[‘chr6’, ‘chr7’]</cite>. Holdout can be regional or
proportional. If regional, expects a list (e.g. <cite>[‘X’, ‘Y’]</cite>).
Regions must match those specified in the first column of the
tabix-indexed BED file. If proportional, specify a percentage
between (0.0, 1.0). Typically 0.10 or 0.20.</li>
<li><strong>test_holdout</strong> (<em>list(str) or float, optional</em>) – Default is <cite>[‘chr8’, ‘chr9’]</cite>. See documentation for
<cite>validation_holdout</cite> for additional information.</li>
<li><strong>sequence_length</strong> (<em>int, optional</em>) – Default is 1000. Model is trained on sequences of <cite>sequence_length</cite>
where genomic features are annotated to the center regions of
these sequences.</li>
<li><strong>center_bin_to_predict</strong> (<em>int, optional</em>) – Default is 200. Query the tabix-indexed file for a region of
length <cite>center_bin_to_predict</cite>.</li>
<li><strong>feature_thresholds</strong> (<em>float [0.0, 1.0] or None, optional</em>) –  Default is 0.5. The <cite>feature_threshold</cite> to pass to the
<cite>GenomicFeatures</cite> object.</li>
<li><strong>mode</strong> (<em>{‘train’, ‘validate’, ‘test’}</em>) – Default is <cite>‘train’</cite>. The mode to run the sampler in.</li>
<li><strong>save_datasets</strong> (<em>list of str</em>) – Default is <cite>[“test”]</cite>. The list of modes for which we should
save the sampled data to file.</li>
<li><strong>output_dir</strong> (<em>str or None, optional</em>) – Default is None. The path to the directory where we should
save sampled examples for a mode. If <cite>save_datasets</cite> is
a non-empty list, <cite>output_dir</cite> must be specified. If
the path in <cite>output_dir</cite> does not exist it will be created
automatically.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Variables:</th><td class="field-body"><ul class="first last simple">
<li><strong>reference_sequence</strong> (<em>selene.sequences.Sequence</em>) – The reference sequence that examples are created from.</li>
<li><strong>target</strong> (<em>selene.targets.Target</em>) – The <cite>selene.targets.Target</cite> object holding the features that we
would like to predict.</li>
<li><strong>sample_from_intervals</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.7)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.7)"><em>tuple</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.7)"><em>int</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.7)"><em>int</em></a><em>)</em><em>)</em>) – A list of coordinates that specify the intervals we can draw
samples from.</li>
<li><strong>interval_lengths</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.7)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.7)"><em>int</em></a><em>)</em>) – A list of the lengths of the intervals that we can draw samples
from. The probability that we will draw a sample from an
interval is a function of that interval’s length and the length
of all other intervals.</li>
<li><strong>sample_negative</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.7)"><em>bool</em></a>) – Whether negative examples (i.e. with no positive label) should
be drawn when generating samples. If <cite>True</cite>, both negative and
positive samples will be drawn. If <cite>False</cite>, only samples with at
least one positive label will be drawn.</li>
<li><strong>validation_holdout</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.7)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)"><em>str</em></a><em>) or </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.7)"><em>float</em></a>) – The samples to hold out for validating model performance. These
can be “regional” or “proportional”. If regional, this is a list
of region names (e.g. <cite>[‘chrX’, ‘chrY’]</cite>). These Regions must
match those specified in the first column of the tabix-indexed
BED file. If proportional, this is the fraction of total samples
that will be held out.</li>
<li><strong>test_holdout</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.7)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)"><em>str</em></a><em>) or </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.7)"><em>float</em></a>) – The samples to hold out for testing model performance. See the
documentation for <cite>validation_holdout</cite> for more details.</li>
<li><strong>sequence_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.7)"><em>int</em></a>) – The length of the sequences to  train the model on.</li>
<li><strong>bin_radius</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.7)"><em>int</em></a>) – From the center of the sequence, the radius in which to detect
a feature annotation in order to include it as a sample’s label.</li>
<li><strong>surrounding_sequence_radius</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.7)"><em>int</em></a>) – The length of sequence falling outside of the feature detection
bin (i.e. <cite>bin_radius</cite>) center, but still within the
<cite>sequence_length</cite>.</li>
<li><strong>modes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.7)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)"><em>str</em></a><em>)</em>) – The list of modes that the sampler can be run in.</li>
<li><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)"><em>str</em></a>) – The current mode that the sampler is running in. Must be one of
the modes listed in <cite>modes</cite>.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="selene_sdk.samplers.IntervalsSampler.sample">
<code class="descname">sample</code><span class="sig-paren">(</span><em>batch_size=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/intervals_sampler.html#IntervalsSampler.sample"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.IntervalsSampler.sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Randomly draws a mini-batch of examples and their corresponding
labels.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>batch_size</strong> (<em>int, optional</em>) – Default is 1. The number of examples to include in the
mini-batch.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><strong>sequences, targets</strong> – A tuple containing the numeric representation of the
sequence examples and their corresponding labels. The
shape of <cite>sequences</cite> will be
<span class="math notranslate nohighlight">\(B \times L \times N\)</span>, where <span class="math notranslate nohighlight">\(B\)</span> is
<cite>batch_size</cite>, <span class="math notranslate nohighlight">\(L\)</span> is the sequence length, and
<span class="math notranslate nohighlight">\(N\)</span> is the size of the sequence type’s alphabet.
The shape of <cite>targets</cite> will be <span class="math notranslate nohighlight">\(B \times F\)</span>,
where <span class="math notranslate nohighlight">\(F\)</span> is the number of features.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.7)">tuple</a>(<a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>, <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>)</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="randompositionssampler">
<h2>RandomPositionsSampler<a class="headerlink" href="#randompositionssampler" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="selene_sdk.samplers.RandomPositionsSampler">
<em class="property">class </em><code class="descclassname">selene_sdk.samplers.</code><code class="descname">RandomPositionsSampler</code><span class="sig-paren">(</span><em>reference_sequence, target_path, features, seed=436, validation_holdout=['chr6', 'chr7'], test_holdout=['chr8', 'chr9'], sequence_length=1000, center_bin_to_predict=200, feature_thresholds=0.5, mode='train', save_datasets=[], output_dir=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/random_positions_sampler.html#RandomPositionsSampler"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.RandomPositionsSampler" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">selene_sdk.samplers.online_sampler.OnlineSampler</span></code></p>
<p>This sampler randomly selects a position in the genome and queries for
a sequence centered at that position for input to the model.</p>
<p>TODO: generalize to selene.sequences.Sequence?</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>reference_sequence</strong> (<em>selene.sequences.Genome</em>) – A reference sequence from which to create examples.</li>
<li><strong>target_path</strong> (<em>str</em>) – Path to tabix-indexed, compressed BED file (<cite>*.bed.gz</cite>) of genomic
coordinates mapped to the genomic features we want to predict.</li>
<li><strong>features</strong> (<em>list(str)</em>) – List of distinct features that we aim to predict.</li>
<li><strong>seed</strong> (<em>int, optional</em>) – Default is 436. Sets the random seed for sampling.</li>
<li><strong>validation_holdout</strong> (<em>list(str) or float, optional</em>) – Default is <cite>[‘chr6’, ‘chr7’]</cite>. Holdout can be regional or
proportional. If regional, expects a list (e.g. <cite>[‘chrX’, ‘chrY’]</cite>).
Regions must match those specified in the first column of the
tabix-indexed BED file. If proportional, specify a percentage
between (0.0, 1.0). Typically 0.10 or 0.20.</li>
<li><strong>test_holdout</strong> (<em>list(str) or float, optional</em>) – Default is <cite>[‘chr8’, ‘chr9’]</cite>. See documentation for
<cite>validation_holdout</cite> for additional information.</li>
<li><strong>sequence_length</strong> (<em>int, optional</em>) – Default is 1000. Model is trained on sequences of <cite>sequence_length</cite>
where genomic features are annotated to the center regions of
these sequences.</li>
<li><strong>center_bin_to_predict</strong> (<em>int, optional</em>) – Default is 200. Query the tabix-indexed file for a region of
length <cite>center_bin_to_predict</cite>.</li>
<li><strong>feature_thresholds</strong> (<em>float [0.0, 1.0], optional</em>) – Default is 0.5. The <cite>feature_threshold</cite> to pass to the
<cite>GenomicFeatures</cite> object.</li>
<li><strong>mode</strong> (<em>{‘train’, ‘validate’, ‘test’}</em>) – Default is <cite>‘train’</cite>. The mode to run the sampler in.</li>
<li><strong>save_datasets</strong> (<em>list(str), optional</em>) – Default is <cite>[‘test’]</cite>. The list of modes for which we should
save the sampled data to file.</li>
<li><strong>output_dir</strong> (<em>str or None, optional</em>) – Default is None. The path to the directory where we should
save sampled examples for a mode. If <cite>save_datasets</cite> is
a non-empty list, <cite>output_dir</cite> must be specified. If
the path in <cite>output_dir</cite> does not exist it will be created
automatically.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Variables:</th><td class="field-body"><ul class="first last simple">
<li><strong>reference_sequence</strong> (<em>selene.sequences.Genome</em>) – The reference sequence that examples are created from.</li>
<li><strong>target</strong> (<em>selene.targets.Target</em>) – The <cite>selene.targets.Target</cite> object holding the features that we
would like to predict.</li>
<li><strong>validation_holdout</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.7)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)"><em>str</em></a><em>) or </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.7)"><em>float</em></a>) – The samples to hold out for validating model performance. These
can be “regional” or “proportional”. If regional, this is a list
of region names (e.g. <cite>[‘chrX’, ‘chrY’]</cite>). These regions must
match those specified in the first column of the tabix-indexed
BED file. If proportional, this is the fraction of total samples
that will be held out.</li>
<li><strong>test_holdout</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.7)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)"><em>str</em></a><em>) or </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.7)"><em>float</em></a>) – The samples to hold out for testing model performance. See the
documentation for <cite>validation_holdout</cite> for more details.</li>
<li><strong>sequence_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.7)"><em>int</em></a>) – The length of the sequences to  train the model on.</li>
<li><strong>bin_radius</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.7)"><em>int</em></a>) – From the center of the sequence, the radius in which to detect
a feature annotation in order to include it as a sample’s label.</li>
<li><strong>surrounding_sequence_radius</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.7)"><em>int</em></a>) – The length of sequence falling outside of the feature detection
bin (i.e. <cite>bin_radius</cite>) center, but still within the
<cite>sequence_length</cite>.</li>
<li><strong>modes</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.7)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)"><em>str</em></a><em>)</em>) – The list of modes that the sampler can be run in.</li>
<li><strong>mode</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)"><em>str</em></a>) – The current mode that the sampler is running in. Must be one of
the modes listed in <cite>modes</cite>.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="selene_sdk.samplers.RandomPositionsSampler.sample">
<code class="descname">sample</code><span class="sig-paren">(</span><em>batch_size=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/random_positions_sampler.html#RandomPositionsSampler.sample"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.RandomPositionsSampler.sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Randomly draws a mini-batch of examples and their corresponding
labels.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>batch_size</strong> (<em>int, optional</em>) – Default is 1. The number of examples to include in the
mini-batch.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><strong>sequences, targets</strong> – A tuple containing the numeric representation of the
sequence examples and their corresponding labels. The
shape of <cite>sequences</cite> will be
<span class="math notranslate nohighlight">\(B \times L \times N\)</span>, where <span class="math notranslate nohighlight">\(B\)</span> is
<cite>batch_size</cite>, <span class="math notranslate nohighlight">\(L\)</span> is the sequence length, and
<span class="math notranslate nohighlight">\(N\)</span> is the size of the sequence type’s alphabet.
The shape of <cite>targets</cite> will be <span class="math notranslate nohighlight">\(B \times F\)</span>,
where <span class="math notranslate nohighlight">\(F\)</span> is the number of features.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.7)">tuple</a>(<a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>, <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>)</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="matfilesampler">
<h2>MatFileSampler<a class="headerlink" href="#matfilesampler" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="selene_sdk.samplers.MatFileSampler">
<em class="property">class </em><code class="descclassname">selene_sdk.samplers.</code><code class="descname">MatFileSampler</code><span class="sig-paren">(</span><em>filepath</em>, <em>sequence_key</em>, <em>targets_key=None</em>, <em>random_seed=436</em>, <em>shuffle_file=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/mat_file_sampler.html#MatFileSampler"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.MatFileSampler" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>A sampler for which the dataset is loaded directly from a <cite>*.mat</cite> file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>filepath</strong> (<em>str</em>) – The path to the file to load the data from.</li>
<li><strong>sequence_key</strong> (<em>str</em>) – The key for the sequences data matrix.</li>
<li><strong>targets_key</strong> (<em>str, optional</em>) – Default is None. The key for the targets data matrix.</li>
<li><strong>random_seed</strong> (<em>int, optional</em>) – Default is 436. Sets the random seed for sampling.</li>
<li><strong>shuffle_file</strong> (<em>bool, optional</em>) – Default is True. Shuffle the data in the matrix before
sampling from it.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Variables:</th><td class="field-body"><p class="first last"><strong>n_samples</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.7)"><em>int</em></a>) – The number of samples in the data matrix.</p>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="selene_sdk.samplers.MatFileSampler.get_data">
<code class="descname">get_data</code><span class="sig-paren">(</span><em>batch_size</em>, <em>n_samples=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/mat_file_sampler.html#MatFileSampler.get_data"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.MatFileSampler.get_data" title="Permalink to this definition">¶</a></dt>
<dd><p>This method fetches a subset of the data from the sampler,
divided into batches.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>batch_size</strong> (<em>int</em>) – The size of the batches to divide the data into.</li>
<li><strong>n_samples</strong> (<em>int, optional</em>) – Default is None. The total number of samples to retrieve.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>sequences</strong> – The list of sequences grouped into batches.
An element in the <cite>sequences</cite> list is of
the shape <span class="math notranslate nohighlight">\(B \times L \times N\)</span>, where <span class="math notranslate nohighlight">\(B\)</span>
is <cite>batch_size</cite>, <span class="math notranslate nohighlight">\(L\)</span> is the sequence length,
and <span class="math notranslate nohighlight">\(N\)</span> is the size of the sequence type’s alphabet.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.7)">list</a>(np.ndarray)</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.MatFileSampler.get_data_and_targets">
<code class="descname">get_data_and_targets</code><span class="sig-paren">(</span><em>batch_size</em>, <em>n_samples=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/mat_file_sampler.html#MatFileSampler.get_data_and_targets"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.MatFileSampler.get_data_and_targets" title="Permalink to this definition">¶</a></dt>
<dd><p>This method fetches a subset of the sequence data and
targets from the sampler, divided into batches.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>batch_size</strong> (<em>int</em>) – The size of the batches to divide the data into.</li>
<li><strong>n_samples</strong> (<em>int, optional</em>) – Default is None. The total number of samples to retrieve.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>sequences_and_targets, targets_matrix</strong> – Tuple containing the list of sequence-target pairs, as well
as a single matrix with all targets in the same order.
Note that <cite>sequences_and_targets</cite>’s sequence elements are of
the shape <span class="math notranslate nohighlight">\(B \times L \times N\)</span> and its target
elements are of the shape <span class="math notranslate nohighlight">\(B \times F\)</span>, where
<span class="math notranslate nohighlight">\(B\)</span> is <cite>batch_size</cite>, <span class="math notranslate nohighlight">\(L\)</span> is the sequence length,
<span class="math notranslate nohighlight">\(N\)</span> is the size of the sequence type’s alphabet, and
<span class="math notranslate nohighlight">\(F\)</span> is the number of features. Further,
<cite>target_matrix</cite> is of the shape <span class="math notranslate nohighlight">\(S \times F\)</span>, where
<span class="math notranslate nohighlight">\(S =\)</span> <cite>n_samples</cite>.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.7)">tuple</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.7)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.7)">tuple</a>(<a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>, <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>)), <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>)</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.MatFileSampler.sample">
<code class="descname">sample</code><span class="sig-paren">(</span><em>batch_size=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/mat_file_sampler.html#MatFileSampler.sample"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.MatFileSampler.sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Draws a mini-batch of examples and their corresponding
labels.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>batch_size</strong> (<em>int, optional</em>) – Default is 1. The number of examples to include in the
mini-batch.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><strong>sequences, targets</strong> – A tuple containing the numeric representation of the
sequence examples and their corresponding labels. The
shape of <cite>sequences</cite> will be
<span class="math notranslate nohighlight">\(B \times L \times N\)</span>, where <span class="math notranslate nohighlight">\(B\)</span> is
<cite>batch_size</cite>, <span class="math notranslate nohighlight">\(L\)</span> is the sequence length, and
<span class="math notranslate nohighlight">\(N\)</span> is the size of the sequence type’s alphabet.
The shape of <cite>targets</cite> will be <span class="math notranslate nohighlight">\(B \times F\)</span>,
where <span class="math notranslate nohighlight">\(F\)</span> is the number of features.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.7)">tuple</a>(<a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>, <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>)</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="bedfilesampler">
<h2>BedFileSampler<a class="headerlink" href="#bedfilesampler" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="selene_sdk.samplers.BedFileSampler">
<em class="property">class </em><code class="descclassname">selene_sdk.samplers.</code><code class="descname">BedFileSampler</code><span class="sig-paren">(</span><em>filepath</em>, <em>reference_sequence</em>, <em>n_samples</em>, <em>sequence_length=None</em>, <em>targets_avail=False</em>, <em>n_features=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/bed_file_sampler.html#BedFileSampler"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.BedFileSampler" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>A sampler for which the dataset is loaded directly from a <cite>*.bed</cite> file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>filepath</strong> (<em>str</em>) – The path to the file to load the data from.</li>
<li><strong>reference_sequence</strong> (<em>selene.sequences.Sequence</em>) – A reference sequence from which to create examples.</li>
<li><strong>n_samples</strong> (<em>int</em>) – Number of lines in the file. (<cite>wc -l &lt;filepath&gt;</cite>)</li>
<li><strong>sequence_length</strong> (<em>int or None, optional</em>) – Default is None. If the coordinates of each sample in the BED file
already account for the full sequence (that is,
<cite>end - start = sequence_length</cite>), there is no need to specify
the sequence length. If <cite>sequence_length</cite> is not None, the length
of each sample will be checked to determine whether the sample
coordinates need to be truncated or expanded to reach the
sequence length specified in the model architecture.</li>
<li><strong>targets_avail</strong> (<em>bool, optional</em>) – Default is False. If <cite>targets_avail</cite>, assumes that it is the
last column of the <cite>*.bed</cite> file. The last column should contain
the indices, separated by semicolons, of features (classes) found
within a given sample’s coordinates (e.g. 0;1;45;60). This assumes
that we are only looking for the absence/presence of each feature
within the interval.</li>
<li><strong>n_features</strong> (<em>int or None, optional</em>) – Default is None. If <cite>targets_avail</cite> is True, must specify
<cite>n_features</cite>, the total number of features (classes).</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Variables:</th><td class="field-body"><ul class="first last simple">
<li><strong>filepath</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)"><em>str</em></a>) – The path to the file to load the data from.</li>
<li><strong>reference_sequence</strong> (<em>selene.sequences.Sequence</em>) – A reference sequence from which to create examples.</li>
<li><strong>n_samples</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.7)"><em>int</em></a>) – Number of lines in the file. (<cite>wc -l &lt;filepath&gt;</cite>)</li>
<li><strong>sequence_length</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.7)"><em>int</em></a><em> or </em><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.7)"><em>None</em></a><em>, </em><em>optional</em>) – Default is None. If the coordinates of each sample in the BED file
already account for the full sequence (that is,
<cite>end - start = sequence_length</cite>), there is no need to specify
the sequence length. If <cite>sequence_length</cite> is not None, the length
of each sample will be checked to determine whether the sample
coordinates need to be truncated or expanded to reach the
sequence length specified in the model architecture.</li>
<li><strong>targets_avail</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.7)"><em>bool</em></a>) – If <cite>targets_avail</cite>, assumes that it is the last column of the <cite>*.bed</cite>
file. The last column should contain the indices, separated by
semicolons, of features (classes) found within a given sample’s
coordinates (e.g. 0;1;45;60). This assumes that we are only looking
or the absence/presence of each feature within the interval.</li>
<li><strong>n_features</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.7)"><em>int</em></a><em> or </em><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.7)"><em>None</em></a>) – If <cite>targets_avail</cite> is True, must specify
<cite>n_features</cite>, the total number of features (classes).</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="selene_sdk.samplers.BedFileSampler.get_data">
<code class="descname">get_data</code><span class="sig-paren">(</span><em>batch_size</em>, <em>n_samples=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/bed_file_sampler.html#BedFileSampler.get_data"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.BedFileSampler.get_data" title="Permalink to this definition">¶</a></dt>
<dd><p>This method fetches a subset of the data from the sampler,
divided into batches.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>batch_size</strong> (<em>int</em>) – The size of the batches to divide the data into.</li>
<li><strong>n_samples</strong> (<em>int, optional</em>) – Default is None. The total number of samples to retrieve.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>sequences</strong> – The list of sequences grouped into batches.
An element in the <cite>sequences</cite> list is of
the shape <span class="math notranslate nohighlight">\(B \times L \times N\)</span>, where <span class="math notranslate nohighlight">\(B\)</span>
is <cite>batch_size</cite>, <span class="math notranslate nohighlight">\(L\)</span> is the sequence length,
and <span class="math notranslate nohighlight">\(N\)</span> is the size of the sequence type’s alphabet.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.7)">list</a>(np.ndarray)</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.BedFileSampler.get_data_and_targets">
<code class="descname">get_data_and_targets</code><span class="sig-paren">(</span><em>batch_size</em>, <em>n_samples=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/bed_file_sampler.html#BedFileSampler.get_data_and_targets"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.BedFileSampler.get_data_and_targets" title="Permalink to this definition">¶</a></dt>
<dd><p>This method fetches a subset of the sequence data and
targets from the sampler, divided into batches.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>batch_size</strong> (<em>int</em>) – The size of the batches to divide the data into.</li>
<li><strong>n_samples</strong> (<em>int, optional</em>) – Default is None. The total number of samples to retrieve.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>sequences_and_targets, targets_matrix</strong> – Tuple containing the list of sequence-target pairs, as well
as a single matrix with all targets in the same order.
Note that <cite>sequences_and_targets</cite>’s sequence elements are of
the shape <span class="math notranslate nohighlight">\(B \times L \times N\)</span> and its target
elements are of the shape <span class="math notranslate nohighlight">\(B \times F\)</span>, where
<span class="math notranslate nohighlight">\(B\)</span> is <cite>batch_size</cite>, <span class="math notranslate nohighlight">\(L\)</span> is the sequence length,
<span class="math notranslate nohighlight">\(N\)</span> is the size of the sequence type’s alphabet, and
<span class="math notranslate nohighlight">\(F\)</span> is the number of features. Further,
<cite>target_matrix</cite> is of the shape <span class="math notranslate nohighlight">\(S \times F\)</span>, where
<span class="math notranslate nohighlight">\(S =\)</span> <cite>n_samples</cite>.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.7)">tuple</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.7)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.7)">tuple</a>(<a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>, <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>)), <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>)</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="selene_sdk.samplers.BedFileSampler.sample">
<code class="descname">sample</code><span class="sig-paren">(</span><em>batch_size=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene_sdk/samplers/bed_file_sampler.html#BedFileSampler.sample"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene_sdk.samplers.BedFileSampler.sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Draws a mini-batch of examples and their corresponding
labels.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>batch_size</strong> (<em>int, optional</em>) – Default is 1. The number of examples to include in the
mini-batch.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><strong>sequences, targets</strong> – A tuple containing the numeric representation of the
sequence examples and their corresponding labels. The
shape of <cite>sequences</cite> will be
<span class="math notranslate nohighlight">\(B \times L \times N\)</span>, where <span class="math notranslate nohighlight">\(B\)</span> is
<cite>batch_size</cite>, <span class="math notranslate nohighlight">\(L\)</span> is the sequence length, and
<span class="math notranslate nohighlight">\(N\)</span> is the size of the sequence type’s alphabet.
The shape of <cite>targets</cite> will be <span class="math notranslate nohighlight">\(B \times F\)</span>,
where <span class="math notranslate nohighlight">\(F\)</span> is the number of features.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.7)">tuple</a>(<a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>, <a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>)</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="sequences.html" class="btn btn-neutral float-right" title="selene_sdk.sequences" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="predict.html" class="btn btn-neutral" title="selene_sdk.predict" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Selene Contributors.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'master',
            LANGUAGE:'None',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  <script type="text/javascript" src="_static/js/theme.js"></script>

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>
 



</body>
</html>