

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>selene.sequences &mdash; Selene master documentation</title>
  

  
  
  
  

  

  
  
    

  

  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Raleway" type="text/css" />
  <link rel="stylesheet" href="_static/css/selene_docs_theme.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="selene.targets" href="targets.html" />
    <link rel="prev" title="selene.samplers" href="samplers.html" /> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> Selene
          

          
            
            <img src="_static/selene_logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                master (0.0.0)
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          

            
            
              
            
            
              <p class="caption"><span class="caption-text">Overview</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="overview/installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="overview/tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="overview/faq.html">Frequently Asked Questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="overview/extending.html">Extending Selene</a></li>
</ul>
<p class="caption"><span class="caption-text">Packages</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="selene.html">selene</a><ul>
<li class="toctree-l2"><a class="reference internal" href="selene.html#trainmodel">TrainModel</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="interpret.html">selene.interpret</a><ul>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#sequence-logo">sequence_logo</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#heatmap">heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#ismresult">ISMResult</a></li>
<li class="toctree-l2"><a class="reference internal" href="interpret.html#rescale-score-matrix">rescale_score_matrix</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="predict.html">selene.predict</a><ul>
<li class="toctree-l2"><a class="reference internal" href="predict.html#analyzesequences">AnalyzeSequences</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="predict.predict_handlers.html">selene.predict.predict_handlers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="predict.predict_handlers.html#predictionshandler">PredictionsHandler</a></li>
<li class="toctree-l2"><a class="reference internal" href="predict.predict_handlers.html#diffscorehandler">DiffScoreHandler</a></li>
<li class="toctree-l2"><a class="reference internal" href="predict.predict_handlers.html#logitscorehandler">LogitScoreHandler</a></li>
<li class="toctree-l2"><a class="reference internal" href="predict.predict_handlers.html#writepredictionshandler">WritePredictionsHandler</a></li>
<li class="toctree-l2"><a class="reference internal" href="predict.predict_handlers.html#writerefalthandler">WriteRefAltHandler</a></li>
<li class="toctree-l2"><a class="reference internal" href="predict.predict_handlers.html#write-to-file">write_to_file</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="samplers.html">selene.samplers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="samplers.html#sampler">Sampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="samplers.html#onlinesampler">OnlineSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="samplers.html#intervalssampler">IntervalsSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="samplers.html#matfilesampler">MatFileSampler</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">selene.sequences</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#sequence">Sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="#genome">Genome</a></li>
<li class="toctree-l2"><a class="reference internal" href="#proteome">Proteome</a></li>
<li class="toctree-l2"><a class="reference internal" href="#sequence-to-encoding">sequence_to_encoding</a></li>
<li class="toctree-l2"><a class="reference internal" href="#encoding-to-sequence">encoding_to_sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="#get-reverse-encoding">get_reverse_encoding</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="targets.html">selene.targets</a><ul>
<li class="toctree-l2"><a class="reference internal" href="targets.html#target">Target</a></li>
<li class="toctree-l2"><a class="reference internal" href="targets.html#genomicfeatures">GenomicFeatures</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="utils.html">selene.utils</a><ul>
<li class="toctree-l2"><a class="reference internal" href="utils.html#performancemetrics">PerformanceMetrics</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#initialize-logger">initialize_logger</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#load-features-list">load_features_list</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils.html#configuration">Configuration</a></li>
</ul>
</li>
</ul>

            
          

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Selene</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>selene.sequences</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/sequences.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-selene.sequences">
<span id="selene-sequences"></span><h1>selene.sequences<a class="headerlink" href="#module-selene.sequences" title="Permalink to this headline">¶</a></h1>
<p>This module provides the types for representing biological sequences.</p>
<div class="section" id="sequence">
<h2>Sequence<a class="headerlink" href="#sequence" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="selene.sequences.Sequence">
<em class="property">class </em><code class="descclassname">selene.sequences.</code><code class="descname">Sequence</code><a class="reference internal" href="_modules/selene/sequences/sequence.html#Sequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.sequences.Sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.7)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>The abstract base class for biological sequence classes.</p>
<dl class="attribute">
<dt id="selene.sequences.Sequence.BASES_ARR">
<code class="descname">BASES_ARR</code><a class="headerlink" href="#selene.sequences.Sequence.BASES_ARR" title="Permalink to this definition">¶</a></dt>
<dd><p>This is an array with the alphabet (i.e. all possible symbols
that may occur in a sequence). We expect that
<cite>INDEX_TO_BASE[i]==BASES_ARR[i]</cite> is <cite>True</cite> for all valid <cite>i</cite>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">The array of all members of the alphabet.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>, dtype=str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="selene.sequences.Sequence.BASE_TO_INDEX">
<code class="descname">BASE_TO_INDEX</code><a class="headerlink" href="#selene.sequences.Sequence.BASE_TO_INDEX" title="Permalink to this definition">¶</a></dt>
<dd><p>A dictionary mapping members of the alphabet (i.e. all
possible symbols that can occur in a sequence) to integers.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">The dictionary mapping the alphabet to integers.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.7)">dict</a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="selene.sequences.Sequence.INDEX_TO_BASE">
<code class="descname">INDEX_TO_BASE</code><a class="headerlink" href="#selene.sequences.Sequence.INDEX_TO_BASE" title="Permalink to this definition">¶</a></dt>
<dd><p>A dictionary mapping integers to members of the alphabet (i.e.
all possible symbols that can occur in a sequence). We expect
that <cite>INDEX_TO_BASE[i]==BASES_ARR[i]</cite> is <cite>True</cite> for all
valid <cite>i</cite>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">The dictionary mapping integers to the alphabet.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.7)">dict</a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="selene.sequences.Sequence.UNK_BASE">
<code class="descname">UNK_BASE</code><a class="headerlink" href="#selene.sequences.Sequence.UNK_BASE" title="Permalink to this definition">¶</a></dt>
<dd><p>This is a base used to represent unknown positions. This is not
the same as a character from outside the sequence’s alphabet. A
character from outside the alphabet is an error. A position with
an unknown base signifies that the position is one of the bases
from the alphabet, but we are uncertain which.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">The character representing an unknown base.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)">str</a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="selene.sequences.Sequence.coords_in_bounds">
<code class="descname">coords_in_bounds</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/sequences/sequence.html#Sequence.coords_in_bounds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.sequences.Sequence.coords_in_bounds" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks if queried coordinates are valid.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><cite>True</cite> if the coordinates are in bounds, otherwise <cite>False</cite>.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.7)">bool</a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="classmethod">
<dt id="selene.sequences.Sequence.encoding_to_sequence">
<em class="property">classmethod </em><code class="descname">encoding_to_sequence</code><span class="sig-paren">(</span><em>encoding</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/sequences/sequence.html#Sequence.encoding_to_sequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.sequences.Sequence.encoding_to_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Transforms the input numerical representation of a sequence
into a string representation.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>encoding</strong> (<em>numpy.ndarray, dtype=numpy.float32</em>) – The <span class="math notranslate nohighlight">\(L \times N\)</span> encoding of the sequence, where
<span class="math notranslate nohighlight">\(L\)</span> is the length of the sequence, and <span class="math notranslate nohighlight">\(N\)</span> is
the size of the sequence type’s alphabet.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">The sequence of bases decoded from the input array. This
sequence will be of length <span class="math notranslate nohighlight">\(L\)</span>.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)">str</a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="selene.sequences.Sequence.get_encoding_from_coords">
<code class="descname">get_encoding_from_coords</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/sequences/sequence.html#Sequence.get_encoding_from_coords"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.sequences.Sequence.get_encoding_from_coords" title="Permalink to this definition">¶</a></dt>
<dd><p>Extracts the numerical encoding for a sequence occurring at
the given coordinates.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">The <span class="math notranslate nohighlight">\(L \times N\)</span> encoding of the sequence occuring
at queried coordinates, where <span class="math notranslate nohighlight">\(L\)</span> is the length of the
sequence, and <span class="math notranslate nohighlight">\(N\)</span> is the size of the sequence type’s
alphabet. Behavior is undefined for invalid coordinates.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>, dtype=numpy.float32</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="selene.sequences.Sequence.get_sequence_from_coords">
<code class="descname">get_sequence_from_coords</code><span class="sig-paren">(</span><em>*args</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/sequences/sequence.html#Sequence.get_sequence_from_coords"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.sequences.Sequence.get_sequence_from_coords" title="Permalink to this definition">¶</a></dt>
<dd><p>Extracts a string representation of a sequence at the
given coordinates.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">The sequence of bases occuring at the queried coordinates.
This sequence will be of length <span class="math notranslate nohighlight">\(L\)</span> normally, but
only if the coordinates are valid. Behavior is undefined for
invalid coordinates.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)">str</a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="classmethod">
<dt id="selene.sequences.Sequence.sequence_to_encoding">
<em class="property">classmethod </em><code class="descname">sequence_to_encoding</code><span class="sig-paren">(</span><em>sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/sequences/sequence.html#Sequence.sequence_to_encoding"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.sequences.Sequence.sequence_to_encoding" title="Permalink to this definition">¶</a></dt>
<dd><p>Transforms a biological sequence into a numerical
representation.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>sequence</strong> (<em>str</em>) – The input sequence of characters.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">The <span class="math notranslate nohighlight">\(L \times N\)</span> encoding of the sequence, where
<span class="math notranslate nohighlight">\(L\)</span> is the length of the sequence, and <span class="math notranslate nohighlight">\(N\)</span> is
the size of the sequence type’s alphabet.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>, dtype=numpy.float32</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="genome">
<h2>Genome<a class="headerlink" href="#genome" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="selene.sequences.Genome">
<em class="property">class </em><code class="descclassname">selene.sequences.</code><code class="descname">Genome</code><span class="sig-paren">(</span><em>input_path</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/sequences/genome.html#Genome"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.sequences.Genome" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">selene.sequences.sequence.Sequence</span></code></p>
<p>This class provides access to an organism’s genomic sequence.</p>
<p>This class supports retrieving parts of the sequence and converting
these parts into their one-hot encodings. It is essentially a
wrapper class around the <cite>pyfaidx.Fasta</cite> class.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>input_path</strong> (<em>str</em>) – Path to an indexed FASTA file, that is, a <cite>*.fasta</cite> file with
a corresponding <cite>*.fai</cite> file in the same directory. This file
should contain the target organism’s genome sequence.</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Variables:</th><td class="field-body"><ul class="first last simple">
<li><strong>genome</strong> (<em>pyfaidx.Fasta</em>) – The FASTA file containing the genome sequence.</li>
<li><strong>chrs</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.7)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)"><em>str</em></a><em>)</em>) – The list of chromosome names.</li>
<li><strong>len_chrs</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.7)"><em>dict</em></a>) – A dictionary mapping the names of each chromosome in the file to
the length of said chromosome.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="attribute">
<dt id="selene.sequences.Genome.BASES_ARR">
<code class="descname">BASES_ARR</code><em class="property"> = array(['A', 'C', 'G', 'T'],       dtype='&lt;U1')</em><a class="headerlink" href="#selene.sequences.Genome.BASES_ARR" title="Permalink to this definition">¶</a></dt>
<dd><p>This is an array with the alphabet (i.e. all possible symbols
that may occur in a sequence). We expect that
<cite>INDEX_TO_BASE[i]==BASES_ARR[i]</cite> is <cite>True</cite> for all valid <cite>i</cite>.</p>
</dd></dl>

<dl class="attribute">
<dt id="selene.sequences.Genome.BASE_TO_INDEX">
<code class="descname">BASE_TO_INDEX</code><em class="property"> = {'A': 0, 'C': 1, 'G': 2, 'T': 3, 'a': 0, 'c': 1, 'g': 2, 't': 3}</em><a class="headerlink" href="#selene.sequences.Genome.BASE_TO_INDEX" title="Permalink to this definition">¶</a></dt>
<dd><p>A dictionary mapping members of the alphabet (i.e. all
possible symbols that can occur in a sequence) to integers.</p>
</dd></dl>

<dl class="attribute">
<dt id="selene.sequences.Genome.COMPLEMENTARY_BASE_DICT">
<code class="descname">COMPLEMENTARY_BASE_DICT</code><em class="property"> = {'A': 'T', 'C': 'G', 'G': 'C', 'N': 'N', 'T': 'A', 'a': 'T', 'c': 'G', 'g': 'C', 'n': 'N', 't': 'A'}</em><a class="headerlink" href="#selene.sequences.Genome.COMPLEMENTARY_BASE_DICT" title="Permalink to this definition">¶</a></dt>
<dd><p>A dictionary mapping each base to its complementary base.</p>
</dd></dl>

<dl class="attribute">
<dt id="selene.sequences.Genome.INDEX_TO_BASE">
<code class="descname">INDEX_TO_BASE</code><em class="property"> = {0: 'A', 1: 'C', 2: 'G', 3: 'T'}</em><a class="headerlink" href="#selene.sequences.Genome.INDEX_TO_BASE" title="Permalink to this definition">¶</a></dt>
<dd><p>A dictionary mapping integers to members of the alphabet (i.e.
all possible symbols that can occur in a sequence). We expect
that <cite>INDEX_TO_BASE[i]==BASES_ARR[i]</cite> is <cite>True</cite> for all
valid <cite>i</cite>.</p>
</dd></dl>

<dl class="attribute">
<dt id="selene.sequences.Genome.UNK_BASE">
<code class="descname">UNK_BASE</code><em class="property"> = 'N'</em><a class="headerlink" href="#selene.sequences.Genome.UNK_BASE" title="Permalink to this definition">¶</a></dt>
<dd><p>This is a base used to represent unknown positions. This is not
the same as a character from outside the sequence’s alphabet. A
character from outside the alphabet is an error. A position with
an unknown base signifies that the position is one of the bases
from the alphabet, but we are uncertain which.</p>
</dd></dl>

<dl class="method">
<dt id="selene.sequences.Genome.coords_in_bounds">
<code class="descname">coords_in_bounds</code><span class="sig-paren">(</span><em>chrom</em>, <em>start</em>, <em>end</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/sequences/genome.html#Genome.coords_in_bounds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.sequences.Genome.coords_in_bounds" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if the region we want to query is within the bounds of the
queried chromosome.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>chrom</strong> (<em>str</em>) – The name of the chromosomes, e.g. “chr1”.</li>
<li><strong>start</strong> (<em>int</em>) – The 0-based start coordinate of the sequence.</li>
<li><strong>end</strong> (<em>int</em>) – One past the last coordinate of the sequence.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">Whether we can retrieve a sequence from the bounds specified
in the input.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.7)">bool</a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="classmethod">
<dt id="selene.sequences.Genome.encoding_to_sequence">
<em class="property">classmethod </em><code class="descname">encoding_to_sequence</code><span class="sig-paren">(</span><em>encoding</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/sequences/genome.html#Genome.encoding_to_sequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.sequences.Genome.encoding_to_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts an input one-hot encoding to its DNA sequence.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>encoding</strong> (<em>numpy.ndarray, dtype=numpy.float32</em>) – An <span class="math notranslate nohighlight">\(L \times 4\)</span> one-hot encoding of the sequence,
where <span class="math notranslate nohighlight">\(L\)</span> is the length of the output sequence.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">The sequence of <span class="math notranslate nohighlight">\(L\)</span> nucleotides decoded from the
input array.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)">str</a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="selene.sequences.Genome.get_chr_lens">
<code class="descname">get_chr_lens</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/sequences/genome.html#Genome.get_chr_lens"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.sequences.Genome.get_chr_lens" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the name and length of each chromosome sequence in the file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">A list of tuples of the chromosome names and lengths.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.7)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.7)">tuple</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.7)">int</a>))</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="selene.sequences.Genome.get_chrs">
<code class="descname">get_chrs</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/sequences/genome.html#Genome.get_chrs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.sequences.Genome.get_chrs" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the list of chromosome names.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">A list of the chromosome names.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.7)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)">str</a>)</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="selene.sequences.Genome.get_encoding_from_coords">
<code class="descname">get_encoding_from_coords</code><span class="sig-paren">(</span><em>chrom</em>, <em>start</em>, <em>end</em>, <em>strand='+'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/sequences/genome.html#Genome.get_encoding_from_coords"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.sequences.Genome.get_encoding_from_coords" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the one-hot encoding of the genomic sequence at the
queried coordinates.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>chrom</strong> (<em>str</em>) – The name of the chromosome or region, e.g. “chr1”.</li>
<li><strong>start</strong> (<em>int</em>) – The 0-based start coordinate of the first position in the
sequence.</li>
<li><strong>end</strong> (<em>int</em>) – One past the 0-based last position in the sequence.</li>
<li><strong>strand</strong> (<em>{‘+’, ‘-‘}, optional</em>) – Default is ‘+’. The strand the sequence is located on.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">The <span class="math notranslate nohighlight">\(L \times 4\)</span> encoding of the sequence, where
<span class="math notranslate nohighlight">\(L = end - start\)</span>.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first"><a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>, dtype=bool</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><p class="first last"><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal notranslate"><span class="pre">ValueError</span></code></a> – If the input char to <cite>strand</cite> is not one of the specified
choices.
(Raised in the call to <cite>self.get_sequence_from_coords</cite>)</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="selene.sequences.Genome.get_sequence_from_coords">
<code class="descname">get_sequence_from_coords</code><span class="sig-paren">(</span><em>chrom</em>, <em>start</em>, <em>end</em>, <em>strand='+'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/sequences/genome.html#Genome.get_sequence_from_coords"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.sequences.Genome.get_sequence_from_coords" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the queried chromosome’s sequence at the input coordinates.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>chrom</strong> (<em>str</em>) – The name of the chromosomes, e.g. “chr1”.</li>
<li><strong>start</strong> (<em>int</em>) – The 0-based start coordinate of the sequence.</li>
<li><strong>end</strong> (<em>int</em>) – One past the 0-based last position in the sequence.</li>
<li><strong>strand</strong> (<em>{‘+’, ‘-‘}, optional</em>) – Default is ‘+’. The strand the sequence is located on.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">The genomic sequence of length <span class="math notranslate nohighlight">\(L\)</span> where
<span class="math notranslate nohighlight">\(L = end - start\)</span>.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)">str</a></p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><p class="first last"><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal notranslate"><span class="pre">ValueError</span></code></a> – If the input char to <cite>strand</cite> is not one of the specified
choices.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="classmethod">
<dt id="selene.sequences.Genome.sequence_to_encoding">
<em class="property">classmethod </em><code class="descname">sequence_to_encoding</code><span class="sig-paren">(</span><em>sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/sequences/genome.html#Genome.sequence_to_encoding"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.sequences.Genome.sequence_to_encoding" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts an input sequence to its one-hot encoding.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>sequence</strong> (<em>str</em>) – A nucleotide sequence of length <span class="math notranslate nohighlight">\(L\)</span></td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">The <span class="math notranslate nohighlight">\(L \times 4\)</span> one-hot encoding of the sequence.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>, dtype=numpy.float32</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="proteome">
<h2>Proteome<a class="headerlink" href="#proteome" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="selene.sequences.Proteome">
<em class="property">class </em><code class="descclassname">selene.sequences.</code><code class="descname">Proteome</code><span class="sig-paren">(</span><em>input_path</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/sequences/proteome.html#Proteome"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.sequences.Proteome" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">selene.sequences.sequence.Sequence</span></code></p>
<p>Provides access to an organism’s proteomic sequence.</p>
<p>It supports retrieving parts of the sequence and converting these
parts into their one-hot encodings. It is essentially a wrapper
class around the <cite>pyfaidx.Fasta</cite> class.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>input_path</strong> (<em>str</em>) – Path to an indexed FASTA file containing amino acid
sequences, that is, a <cite>*.fasta</cite> file with a corresponding
<cite>*.fai</cite> file in the same directory. File should contain the
sequences from which training examples will be created.</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Variables:</th><td class="field-body"><ul class="first last simple">
<li><strong>proteome</strong> (<em>pyfaidx.Fasta</em>) – The FASTA or FAA file containing the protein sequences.</li>
<li><strong>prots</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.7)"><em>list</em></a><em>(</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)"><em>str</em></a><em>)</em>) – The list of protein names.</li>
<li><strong>len_prots</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.7)"><em>dict</em></a>) – A dictionary that maps protein names to the lengths, and does so
for all protein sequences in the proteome.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="attribute">
<dt id="selene.sequences.Proteome.BASES_ARR">
<code class="descname">BASES_ARR</code><em class="property"> = array(['A', 'R', 'N', 'D', 'C', 'E', 'Q', 'G', 'H', 'I', 'L', 'K', 'M',        'F', 'P', 'S', 'T', 'W', 'Y', 'V'],       dtype='&lt;U1')</em><a class="headerlink" href="#selene.sequences.Proteome.BASES_ARR" title="Permalink to this definition">¶</a></dt>
<dd><p>This is an array with the alphabet (i.e. all possible symbols
that may occur in a sequence). We expect that
<cite>INDEX_TO_BASE[i]==BASES_ARR[i]</cite> is <cite>True</cite> for all valid <cite>i</cite>.</p>
</dd></dl>

<dl class="attribute">
<dt id="selene.sequences.Proteome.BASE_TO_INDEX">
<code class="descname">BASE_TO_INDEX</code><em class="property"> = {'A': 0, 'C': 4, 'D': 3, 'E': 5, 'F': 13, 'G': 7, 'H': 8, 'I': 9, 'K': 11, 'L': 10, 'M': 12, 'N': 2, 'P': 14, 'Q': 6, 'R': 1, 'S': 15, 'T': 16, 'V': 19, 'W': 17, 'Y': 18}</em><a class="headerlink" href="#selene.sequences.Proteome.BASE_TO_INDEX" title="Permalink to this definition">¶</a></dt>
<dd><p>A dictionary mapping members of the alphabet (i.e. all
possible symbols that can occur in a sequence) to integers.</p>
</dd></dl>

<dl class="attribute">
<dt id="selene.sequences.Proteome.INDEX_TO_BASE">
<code class="descname">INDEX_TO_BASE</code><em class="property"> = {0: 'A', 1: 'R', 2: 'N', 3: 'D', 4: 'C', 5: 'E', 6: 'Q', 7: 'G', 8: 'H', 9: 'I', 10: 'L', 11: 'K', 12: 'M', 13: 'F', 14: 'P', 15: 'S', 16: 'T', 17: 'W', 18: 'Y', 19: 'V'}</em><a class="headerlink" href="#selene.sequences.Proteome.INDEX_TO_BASE" title="Permalink to this definition">¶</a></dt>
<dd><p>A dictionary mapping integers to members of the alphabet (i.e.
all possible symbols that can occur in a sequence). We expect
that <cite>INDEX_TO_BASE[i]==BASES_ARR[i]</cite> is <cite>True</cite> for all
valid <cite>i</cite>.</p>
</dd></dl>

<dl class="attribute">
<dt id="selene.sequences.Proteome.UNK_BASE">
<code class="descname">UNK_BASE</code><em class="property"> = 'X'</em><a class="headerlink" href="#selene.sequences.Proteome.UNK_BASE" title="Permalink to this definition">¶</a></dt>
<dd><p>This is a base used to represent unknown positions. This is not
the same as a character from outside the sequence’s alphabet. A
character from outside the alphabet is an error. A position with
an unknown base signifies that the position is one of the bases
from the alphabet, but we are uncertain which.</p>
</dd></dl>

<dl class="method">
<dt id="selene.sequences.Proteome.coords_in_bounds">
<code class="descname">coords_in_bounds</code><span class="sig-paren">(</span><em>prot</em>, <em>start</em>, <em>end</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/sequences/proteome.html#Proteome.coords_in_bounds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.sequences.Proteome.coords_in_bounds" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if the coordinates we want to query is valid.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>prot</strong> (<em>str</em>) – The name of the protein, e.g. “YFP”.</li>
<li><strong>start</strong> (<em>int</em>) – The 0-based start coordinate of the first position in the
sequence.</li>
<li><strong>end</strong> (<em>int</em>) – One past the 0-based last position in the sequence.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">A boolean indicating whether we can retrieve a sequence from
the queried coordinates.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.7)">bool</a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="classmethod">
<dt id="selene.sequences.Proteome.encoding_to_sequence">
<em class="property">classmethod </em><code class="descname">encoding_to_sequence</code><span class="sig-paren">(</span><em>encoding</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/sequences/proteome.html#Proteome.encoding_to_sequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.sequences.Proteome.encoding_to_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts an input one-hot encoding to its amino acid sequence.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>encoding</strong> (<em>numpy.ndarray, dtype=numpy.float32</em>) – The <span class="math notranslate nohighlight">\(L \times 20\)</span> encoding of the sequence, where
<span class="math notranslate nohighlight">\(L\)</span> is the length of the output amino acid sequence.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">The sequence of <span class="math notranslate nohighlight">\(L\)</span> amino acids decoded from the
input array.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)">str</a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="selene.sequences.Proteome.get_encoding_from_coords">
<code class="descname">get_encoding_from_coords</code><span class="sig-paren">(</span><em>prot</em>, <em>start</em>, <em>end</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/sequences/proteome.html#Proteome.get_encoding_from_coords"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.sequences.Proteome.get_encoding_from_coords" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the one-hot encoding of the protein’s sequence at the
input coordinates.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>prot</strong> (<em>str</em>) – The name of the protein, e.g. “YFP”.</li>
<li><strong>start</strong> (<em>int</em>) – The 0-based start coordinate of the first position in the
sequence.</li>
<li><strong>end</strong> (<em>int</em>) – One past the 0-based last position in the sequence.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">The <span class="math notranslate nohighlight">\(L \times 20\)</span> encoding of the sequence, where
<span class="math notranslate nohighlight">\(L = end - start\)</span>.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>, dtype=numpy.float32</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="selene.sequences.Proteome.get_prot_lens">
<code class="descname">get_prot_lens</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/sequences/proteome.html#Proteome.get_prot_lens"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.sequences.Proteome.get_prot_lens" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the name and length of each protein sequence in the file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">A list of tuples of protein names and protein lengths.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.7)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.7)">tuple</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)">str</a>, <a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.7)">int</a>))</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="selene.sequences.Proteome.get_prots">
<code class="descname">get_prots</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/sequences/proteome.html#Proteome.get_prots"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.sequences.Proteome.get_prots" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the list of protein names.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">A list of the protein names.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.7)">list</a>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)">str</a>)</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="selene.sequences.Proteome.get_sequence_from_coords">
<code class="descname">get_sequence_from_coords</code><span class="sig-paren">(</span><em>prot</em>, <em>start</em>, <em>end</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/sequences/proteome.html#Proteome.get_sequence_from_coords"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.sequences.Proteome.get_sequence_from_coords" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the queried protein sequence at the input coordinates.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>prot</strong> (<em>str</em>) – The protein name, e.g. “YFP”.</li>
<li><strong>start</strong> (<em>int</em>) – The 0-based start coordinate of the first position in the
sequence.</li>
<li><strong>end</strong> (<em>int</em>) – One past the 0-based last position in the sequence.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">The sequence of <span class="math notranslate nohighlight">\(L\)</span> amino acids at the specified
coordinates, where <span class="math notranslate nohighlight">\(L = end - start\)</span>.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)">str</a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="classmethod">
<dt id="selene.sequences.Proteome.sequence_to_encoding">
<em class="property">classmethod </em><code class="descname">sequence_to_encoding</code><span class="sig-paren">(</span><em>sequence</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/sequences/proteome.html#Proteome.sequence_to_encoding"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.sequences.Proteome.sequence_to_encoding" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts an input sequence to its one-hot encoding.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>sequence</strong> (<em>str</em>) – The input sequence of amino acids of length <span class="math notranslate nohighlight">\(L\)</span>.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">The <span class="math notranslate nohighlight">\(L \times 20\)</span> array, where <cite>L</cite> was the length of
the input sequence.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>, dtype=numpy.float32</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="sequence-to-encoding">
<h2>sequence_to_encoding<a class="headerlink" href="#sequence-to-encoding" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="selene.sequences.sequence_to_encoding">
<code class="descclassname">selene.sequences.</code><code class="descname">sequence_to_encoding</code><span class="sig-paren">(</span><em>sequence</em>, <em>base_to_index</em>, <em>bases_arr</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/sequences/sequence.html#sequence_to_encoding"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.sequences.sequence_to_encoding" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts an input sequence to its one-hot encoding.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>sequence</strong> (<em>str</em>) – The input sequence of length <span class="math notranslate nohighlight">\(L\)</span>.</li>
<li><strong>base_to_index</strong> (<em>dict</em>) – A dict that maps input characters to indices, where the indices
specify the column to assign as 1 when a base exists at the
current position in the input. If a base does not exist at the
current position in the input, it’s corresponding column in the
encoding is set as zero. Note that the rows correspond directly
to the positions in the input sequence. For instance, with a
a genome you would have each of <cite>[‘A’, ‘C’, ‘G’, ‘T’]</cite> as keys,
mapping to values of <cite>[0, 1, 2, 3]</cite>.</li>
<li><strong>bases_arr</strong> (<em>list(str)</em>) – The characters in the sequence’s alphabet.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">The <span class="math notranslate nohighlight">\(L \times N\)</span> encoding of the sequence, where
<span class="math notranslate nohighlight">\(L\)</span> is the length of the input sequence and <span class="math notranslate nohighlight">\(N\)</span> is
the size of the sequence alphabet.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.html#numpy.ndarray" title="(in NumPy v1.14)">numpy.ndarray</a>, dtype=numpy.float32</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="encoding-to-sequence">
<h2>encoding_to_sequence<a class="headerlink" href="#encoding-to-sequence" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="selene.sequences.encoding_to_sequence">
<code class="descclassname">selene.sequences.</code><code class="descname">encoding_to_sequence</code><span class="sig-paren">(</span><em>encoding</em>, <em>bases_arr</em>, <em>unk_base</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/sequences/sequence.html#encoding_to_sequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.sequences.encoding_to_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Converts a sequence one-hot encoding to its string sequence.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>encoding</strong> (<em>numpy.ndarray, dtype=numpy.float32</em>) – The <span class="math notranslate nohighlight">\(L \times N\)</span> encoding of the sequence, where
<span class="math notranslate nohighlight">\(L\)</span> is the length of the sequence, and <span class="math notranslate nohighlight">\(N\)</span> is the
size of the sequence alphabet.</li>
<li><strong>bases_arr</strong> (<em>list(str)</em>) – A list of the bases in the sequence’s alphabet that corresponds
to the correct columns for those bases in the encoding.</li>
<li><strong>unk_base</strong> (<em>str</em>) – The base corresponding to the “unknown” character in this
encoding. See <cite>selene.sequences.Sequence.UNK_BASE</cite> for more
information.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">The sequence of <span class="math notranslate nohighlight">\(L\)</span> characters decoded from the
input array.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.7)">str</a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="get-reverse-encoding">
<h2>get_reverse_encoding<a class="headerlink" href="#get-reverse-encoding" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="selene.sequences.get_reverse_encoding">
<code class="descclassname">selene.sequences.</code><code class="descname">get_reverse_encoding</code><span class="sig-paren">(</span><em>encoding</em>, <em>bases_arr</em>, <em>base_to_index</em>, <em>complementary_base_dict</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/selene/sequences/sequence.html#get_reverse_encoding"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#selene.sequences.get_reverse_encoding" title="Permalink to this definition">¶</a></dt>
<dd><p># TODO(DOCUMENTATION): What is the documentation for this?
:Parameters: * <strong>encoding</strong></p>
<blockquote>
<div><ul class="simple">
<li><strong>bases_arr</strong></li>
<li><strong>base_to_index</strong></li>
<li><strong>complementary_base_dict</strong> (<em>dict</em>)</li>
</ul>
</div></blockquote>
</dd></dl>

</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="targets.html" class="btn btn-neutral float-right" title="selene.targets" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="samplers.html" class="btn btn-neutral" title="selene.samplers" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Selene Contributors.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'master',
            LANGUAGE:'None',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  <script type="text/javascript" src="_static/js/theme.js"></script>

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>
 



</body>
</html>