

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>selene_sdk.predict.model_predict &mdash; Selene master documentation</title>
  

  
  
    <link rel="shortcut icon" href="../../../_static/favicon.ico"/>
  
  
  

  

  
  
    

  

  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Raleway" type="text/css" />
  <link rel="stylesheet" href="../../../_static/css/selene_docs_theme.css" type="text/css" />
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 

  
  <script src="../../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../../index.html" class="icon icon-home"> Selene
          

          
            
            <img src="../../../_static/selene_logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                master (0.2.0)
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          

            
            
              
            
            
              <p class="caption"><span class="caption-text">Overview</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/overview.html">Functional overview of the SDK</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#sampling">Sampling</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/overview.html#sequences-submodule-api"><em>sequences</em> submodule (API)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/overview.html#targets-submodule-api"><em>targets</em> submodule (API)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/overview.html#samplers-submodule-api"><em>samplers</em> submodule (API)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#training-a-model-api">Training a model (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#evaluating-a-model-api">Evaluating a model (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#using-a-model-to-make-predictions-api">Using a model to make predictions (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#visualizing-model-predictions-api">Visualizing model predictions (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/overview.html#the-utilities-submodule-api">The utilities submodule (API)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/installation.html#installing-with-anaconda">Installing with Anaconda</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/installation.html#installing-with-pip">Installing with pip</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/installation.html#installing-from-source">Installing from source</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/cli.html">Selene CLI operations and outputs</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#operations">Operations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#general-configurations">General configurations</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#model-architecture">Model architecture</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#expected-input-class-and-methods">Expected input class and methods</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#model-architecture-configuration">Model architecture configuration</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#a-note-for-the-following-sections">A note for the following sections</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#train">Train</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#required-parameters">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#optional-parameters">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#additional-notes">Additional notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#expected-outputs-for-training">Expected outputs for training</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#evaluate">Evaluate</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#required-parameters">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#optional-parameters">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#additional-notes">Additional notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#expected-outputs-for-evaluation">Expected outputs for evaluation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#analyze-sequences">Analyze sequences</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#required-parameters">Required parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#optional-parameters">Optional parameters</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#prediction-on-sequences">Prediction on sequences</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#parameters">Parameters</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#variant-effect-prediction">Variant effect prediction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#parameters">Parameters</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#in-silico-mutagenesis"><em>In silico</em> mutagenesis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#parameters-for-a-single-sequence-input">Parameters for a single sequence input</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#parameters-for-fasta-file-input">Parameters for FASTA file input:</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#sampler-configurations">Sampler configurations</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#samplers-used-for-training-and-evaluation-optionally">Samplers used for training (and evaluation, optionally)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#random-positions-sampler">Random positions sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#intervals-sampler">Intervals sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#multiple-file-sampler">Multiple-file sampler</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#important-note">Important note</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#samplers-used-for-evaluation">Samplers used for evaluation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#bed-file-sampler">BED file sampler</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#matrix-file-sampler">Matrix file sampler</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/cli.html#examples-of-full-configuration-files">Examples of full configuration files</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#training-using-intervals-sampler">Training (using intervals sampler)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#some-notes">Some notes</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#evaluate-using-matrix-file-sampler">Evaluate (using matrix file sampler)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#some-notes">Some notes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../overview/cli.html#analyze-sequences-variant-effect-prediction">Analyze sequences (variant effect prediction)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../overview/cli.html#some-notes">Some notes</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/faq.html">FAQ and additional resources</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/faq.html#extending-selene">Extending Selene</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../overview/faq.html#hyperparameter-optimization">Hyperparameter optimization</a></li>
</ul>
</li>
</ul>
<p class="caption"><span class="caption-text">Packages</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../selene.html">selene_sdk</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../selene.html#trainmodel">TrainModel</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../selene.html#evaluatemodel">EvaluateModel</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../interpret.html">selene_sdk.interpret</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#sequence-logo">sequence_logo</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#heatmap">heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#ismresult">ISMResult</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#rescale-score-matrix">rescale_score_matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#load-variant-abs-diff-scores">load_variant_abs_diff_scores</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#variant-diffs-scatter-plot">variant_diffs_scatter_plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#ordered-variants-and-indices">ordered_variants_and_indices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#sort-standard-chrs">sort_standard_chrs</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../predict.html">selene_sdk.predict</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../predict.html#analyzesequences">AnalyzeSequences</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../samplers.html">selene_sdk.samplers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#sampler">Sampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#onlinesampler">OnlineSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#intervalssampler">IntervalsSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#randompositionssampler">RandomPositionsSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#multifilesampler">MultiFileSampler</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../samplers.file_samplers.html">selene_sdk.samplers.file_samplers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.file_samplers.html#bedfilesampler">BedFileSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.file_samplers.html#matfilesampler">MatFileSampler</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../sequences.html">selene_sdk.sequences</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#sequence">Sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#genome">Genome</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#proteome">Proteome</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#sequence-to-encoding">sequence_to_encoding</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#encoding-to-sequence">encoding_to_sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#get-reverse-encoding">get_reverse_encoding</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../targets.html">selene_sdk.targets</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../targets.html#target">Target</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../targets.html#genomicfeatures">GenomicFeatures</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../utils.html">selene_sdk.utils</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#nonstrandspecific">NonStrandSpecific</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#performancemetrics">PerformanceMetrics</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#initialize-logger">initialize_logger</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#load-features-list">load_features_list</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#load-model-from-state-dict">load_model_from_state_dict</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#get-indices-and-probabilities">get_indices_and_probabilities</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#load-path-for-config-yml">load_path (for config.yml)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#instantiate-configuration-object">instantiate (configuration object)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#initialize-model">initialize_model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#parse-configs-and-run">parse_configs_and_run</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#execute-run-selene-operations">execute (run Selene operations)</a></li>
</ul>
</li>
</ul>

            
          

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">Selene</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>selene_sdk.predict.model_predict</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for selene_sdk.predict.model_predict</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">This module provides the `AnalyzeSequences` class and supporting</span>
<span class="sd">methods.</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">itertools</span>
<span class="kn">import</span> <span class="nn">os</span>

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pyfaidx</span>
<span class="kn">import</span> <span class="nn">torch</span>
<span class="kn">from</span> <span class="nn">torch.autograd</span> <span class="k">import</span> <span class="n">Variable</span>

<span class="kn">from</span> <span class="nn">.predict_handlers</span> <span class="k">import</span> <span class="n">AbsDiffScoreHandler</span>
<span class="kn">from</span> <span class="nn">.predict_handlers</span> <span class="k">import</span> <span class="n">DiffScoreHandler</span>
<span class="kn">from</span> <span class="nn">.predict_handlers</span> <span class="k">import</span> <span class="n">LogitScoreHandler</span>
<span class="kn">from</span> <span class="nn">.predict_handlers</span> <span class="k">import</span> <span class="n">WritePredictionsHandler</span>
<span class="kn">from</span> <span class="nn">.predict_handlers</span> <span class="k">import</span> <span class="n">WriteRefAltHandler</span>
<span class="kn">from</span> <span class="nn">..sequences</span> <span class="k">import</span> <span class="n">Genome</span>
<span class="kn">from</span> <span class="nn">..utils</span> <span class="k">import</span> <span class="n">load_model_from_state_dict</span>


<span class="c1"># TODO: MAKE THESE GENERIC:</span>
<span class="n">ISM_COLS</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;pos&quot;</span><span class="p">,</span> <span class="s2">&quot;ref&quot;</span><span class="p">,</span> <span class="s2">&quot;alt&quot;</span><span class="p">]</span>
<span class="n">VCF_REQUIRED_COLS</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;#CHROM&quot;</span><span class="p">,</span> <span class="s2">&quot;POS&quot;</span><span class="p">,</span> <span class="s2">&quot;ID&quot;</span><span class="p">,</span> <span class="s2">&quot;REF&quot;</span><span class="p">,</span> <span class="s2">&quot;ALT&quot;</span><span class="p">]</span>
<span class="n">VARIANTEFFECT_COLS</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">,</span> <span class="s2">&quot;pos&quot;</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="s2">&quot;ref&quot;</span><span class="p">,</span> <span class="s2">&quot;alt&quot;</span><span class="p">]</span>


<span class="k">def</span> <span class="nf">in_silico_mutagenesis_sequences</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span>
                                    <span class="n">mutate_n_bases</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                                    <span class="n">reference_sequence</span><span class="o">=</span><span class="n">Genome</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates a list containing each mutation that occurs from an</span>
<span class="sd">    *in silico* mutagenesis across the whole sequence.</span>

<span class="sd">    Please note that we have not parallelized this function yet, so</span>
<span class="sd">    runtime increases exponentially when you increase `mutate_n_bases`.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    sequence : str</span>
<span class="sd">        A string containing the sequence we would like to mutate.</span>
<span class="sd">    mutate_n_bases : int, optional</span>
<span class="sd">        Default is 1. The number of base changes to make with each set of</span>
<span class="sd">        mutations evaluated, e.g. `mutate_n_bases = 2` considers all</span>
<span class="sd">        pairs of SNPs.</span>
<span class="sd">    reference_sequence : class, optional</span>
<span class="sd">        Default is `selene_sdk.sequences.Genome`. The type of sequence</span>
<span class="sd">        that has been passed in.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    list(list(tuple))</span>
<span class="sd">        A list of all possible mutations. Each element in the list is</span>
<span class="sd">        itself a list of tuples, e.g. element = [(0, &#39;T&#39;)] when only mutating</span>
<span class="sd">        1 base at a time. Each tuple is the position to mutate and the base</span>
<span class="sd">        with which we are replacing the reference base.</span>

<span class="sd">        For a sequence of length 1000, mutating 1 base at a time means that</span>
<span class="sd">        we return a list with length of 3000-4000, depending on the number of</span>
<span class="sd">        unknown bases in the input sequences.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sequence_alts</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">ref</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">sequence</span><span class="p">):</span>
        <span class="n">alts</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">base</span> <span class="ow">in</span> <span class="n">reference_sequence</span><span class="o">.</span><span class="n">BASES_ARR</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">base</span> <span class="o">==</span> <span class="n">ref</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">alts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">base</span><span class="p">)</span>
        <span class="n">sequence_alts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alts</span><span class="p">)</span>
    <span class="n">all_mutated_sequences</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">indices</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">combinations</span><span class="p">(</span>
            <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)),</span> <span class="n">mutate_n_bases</span><span class="p">):</span>
        <span class="n">pos_mutations</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">indices</span><span class="p">:</span>
            <span class="n">pos_mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sequence_alts</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
        <span class="k">for</span> <span class="n">mutations</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">product</span><span class="p">(</span><span class="o">*</span><span class="n">pos_mutations</span><span class="p">):</span>
            <span class="n">all_mutated_sequences</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">indices</span><span class="p">,</span> <span class="n">mutations</span><span class="p">)))</span>
    <span class="k">return</span> <span class="n">all_mutated_sequences</span>


<span class="k">def</span> <span class="nf">_ism_sample_id</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">mutation_information</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    TODO</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    sequence : str</span>
<span class="sd">        The input sequence to mutate.</span>
<span class="sd">    mutation_information : list(tuple)</span>
<span class="sd">        TODO</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    TODO</span>
<span class="sd">        TODO</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">positions</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">refs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">alts</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">position</span><span class="p">,</span> <span class="n">alt</span><span class="p">)</span> <span class="ow">in</span> <span class="n">mutation_information</span><span class="p">:</span>
        <span class="n">positions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">position</span><span class="p">))</span>
        <span class="n">refs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sequence</span><span class="p">[</span><span class="n">position</span><span class="p">])</span>
        <span class="n">alts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alt</span><span class="p">)</span>
    <span class="k">return</span> <span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">positions</span><span class="p">),</span> <span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">refs</span><span class="p">),</span> <span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">alts</span><span class="p">))</span>


<span class="k">def</span> <span class="nf">mutate_sequence</span><span class="p">(</span><span class="n">encoding</span><span class="p">,</span>
                    <span class="n">mutation_information</span><span class="p">,</span>
                    <span class="n">reference_sequence</span><span class="o">=</span><span class="n">Genome</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Transforms a sequence with a set of mutations.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    encoding : numpy.ndarray</span>
<span class="sd">        An :math:`L \\times N` array (where :math:`L` is the sequence&#39;s</span>
<span class="sd">        length and :math:`N` is the size of the sequence type&#39;s</span>
<span class="sd">        alphabet) holding the one-hot encoding of the</span>
<span class="sd">        reference sequence.</span>
<span class="sd">    mutation_information : list(tuple)</span>
<span class="sd">        List of tuples of (`int`, `str`). Each tuple is the position to</span>
<span class="sd">        mutate and the base to which to mutate that position in the</span>
<span class="sd">        sequence.</span>
<span class="sd">    reference_sequence : class, optional</span>
<span class="sd">        Default is `selene_sdk.sequences.Genome`. A reference sequence</span>
<span class="sd">        from which to retrieve smaller sequences..</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    numpy.ndarray</span>
<span class="sd">        An :math:`L \\times N` array holding the one-hot encoding of</span>
<span class="sd">        the mutated sequence.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">mutated_seq</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">encoding</span><span class="p">)</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">position</span><span class="p">,</span> <span class="n">alt</span><span class="p">)</span> <span class="ow">in</span> <span class="n">mutation_information</span><span class="p">:</span>
        <span class="n">replace_base</span> <span class="o">=</span> <span class="n">reference_sequence</span><span class="o">.</span><span class="n">BASE_TO_INDEX</span><span class="p">[</span><span class="n">alt</span><span class="p">]</span>
        <span class="n">mutated_seq</span><span class="p">[</span><span class="n">position</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">mutated_seq</span><span class="p">[</span><span class="n">position</span><span class="p">,</span> <span class="n">replace_base</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="k">return</span> <span class="n">mutated_seq</span>


<span class="c1"># TODO: Is this a general method that might belong in utils?</span>
<span class="k">def</span> <span class="nf">read_vcf_file</span><span class="p">(</span><span class="n">input_path</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Read the relevant columns for a variant call format (VCF) file to</span>
<span class="sd">    collect variants for variant effect prediction.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    input_path : str</span>
<span class="sd">        Path to the VCF file.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    list(tuple)</span>
<span class="sd">        List of variants. Tuple = (chrom, position, id, ref, alt)</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">variants</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">input_path</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file_handle</span><span class="p">:</span>
        <span class="n">lines</span> <span class="o">=</span> <span class="n">file_handle</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
        <span class="n">index</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">lines</span><span class="p">):</span>
            <span class="k">if</span> <span class="s1">&#39;#&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">line</span><span class="p">:</span>
                <span class="k">break</span>
            <span class="k">if</span> <span class="s2">&quot;#CHROM&quot;</span> <span class="ow">in</span> <span class="n">line</span><span class="p">:</span>
                <span class="n">cols</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">cols</span><span class="p">[:</span><span class="mi">5</span><span class="p">]</span> <span class="o">!=</span> <span class="n">VCF_REQUIRED_COLS</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="s2">&quot;First 5 columns in file </span><span class="si">{0}</span><span class="s2"> were </span><span class="si">{1}</span><span class="s2">. &quot;</span>
                        <span class="s2">&quot;Expected columns: </span><span class="si">{2}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                            <span class="n">input_path</span><span class="p">,</span> <span class="n">cols</span><span class="p">[:</span><span class="mi">5</span><span class="p">],</span> <span class="n">VCF_REQUIRED_COLS</span><span class="p">))</span>
                <span class="n">index</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">break</span>

        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">lines</span><span class="p">[</span><span class="n">index</span><span class="p">:]:</span>
            <span class="n">cols</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cols</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">5</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">chrom</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">cols</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="n">pos</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cols</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
            <span class="n">name</span> <span class="o">=</span> <span class="n">cols</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
            <span class="n">ref</span> <span class="o">=</span> <span class="n">cols</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
            <span class="n">alt</span> <span class="o">=</span> <span class="n">cols</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span>
            <span class="n">variants</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">chrom</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">alt</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">variants</span>


<span class="k">def</span> <span class="nf">_add_sequence_surrounding_alt</span><span class="p">(</span><span class="n">alt_sequence</span><span class="p">,</span>
                                  <span class="n">sequence_length</span><span class="p">,</span>
                                  <span class="n">chrom</span><span class="p">,</span>
                                  <span class="n">ref_start</span><span class="p">,</span>
                                  <span class="n">ref_end</span><span class="p">,</span>
                                  <span class="n">reference_sequence</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    TODO</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    alt_sequence : str</span>
<span class="sd">    sequence_length : int</span>
<span class="sd">    chrom : str</span>
<span class="sd">    ref_start : int</span>
<span class="sd">    ref_end : int</span>
<span class="sd">    reference_sequence : selene_sdk.sequences.Sequence</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    str</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">alt_len</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">alt_sequence</span><span class="p">)</span>
    <span class="n">add_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">((</span><span class="n">sequence_length</span> <span class="o">-</span> <span class="n">alt_len</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
    <span class="n">add_end</span> <span class="o">=</span> <span class="n">add_start</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">sequence_length</span> <span class="o">-</span> <span class="n">alt_len</span><span class="p">)</span> <span class="o">%</span> <span class="mi">2</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">add_end</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="n">lhs_end</span> <span class="o">=</span> <span class="n">ref_start</span>
    <span class="n">lhs_start</span> <span class="o">=</span> <span class="n">ref_start</span> <span class="o">-</span> <span class="n">add_start</span>

    <span class="n">rhs_start</span> <span class="o">=</span> <span class="n">ref_end</span>
    <span class="n">rhs_end</span> <span class="o">=</span> <span class="n">ref_end</span> <span class="o">+</span> <span class="n">add_end</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">reference_sequence</span><span class="o">.</span><span class="n">coords_in_bounds</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">rhs_start</span><span class="p">,</span> <span class="n">rhs_end</span><span class="p">):</span>
        <span class="c1"># add everything to the LHS</span>
        <span class="n">lhs_start</span> <span class="o">=</span> <span class="n">ref_start</span> <span class="o">-</span> <span class="n">sequence_length</span> <span class="o">+</span> <span class="n">alt_len</span>
        <span class="n">alt_sequence</span> <span class="o">=</span> <span class="n">reference_sequence</span><span class="o">.</span><span class="n">get_sequence_from_coords</span><span class="p">(</span>
            <span class="n">chrom</span><span class="p">,</span> <span class="n">lhs_start</span><span class="p">,</span> <span class="n">lhs_end</span><span class="p">)</span> <span class="o">+</span> <span class="n">alt_sequence</span>
    <span class="k">elif</span> <span class="ow">not</span> <span class="n">reference_sequence</span><span class="o">.</span><span class="n">coords_in_bounds</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">lhs_start</span><span class="p">,</span> <span class="n">lhs_end</span><span class="p">):</span>
        <span class="c1"># add everything to RHS</span>
        <span class="n">rhs_end</span> <span class="o">=</span> <span class="n">ref_end</span> <span class="o">+</span> <span class="n">sequence_length</span> <span class="o">-</span> <span class="n">alt_len</span>
        <span class="n">alt_sequence</span> <span class="o">+=</span> <span class="n">reference_sequence</span><span class="o">.</span><span class="n">get_sequence_from_coords</span><span class="p">(</span>
            <span class="n">chrom</span><span class="p">,</span> <span class="n">rhs_start</span><span class="p">,</span> <span class="n">rhs_end</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">lhs_start</span> <span class="o">&gt;=</span> <span class="n">lhs_end</span><span class="p">:</span>
            <span class="n">lhs_sequence</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">lhs_sequence</span> <span class="o">=</span> <span class="n">reference_sequence</span><span class="o">.</span><span class="n">get_sequence_from_coords</span><span class="p">(</span>
                <span class="n">chrom</span><span class="p">,</span> <span class="n">lhs_start</span><span class="p">,</span> <span class="n">lhs_end</span><span class="p">)</span>
        <span class="n">rhs_sequence</span> <span class="o">=</span> <span class="n">reference_sequence</span><span class="o">.</span><span class="n">get_sequence_from_coords</span><span class="p">(</span>
            <span class="n">chrom</span><span class="p">,</span> <span class="n">rhs_start</span><span class="p">,</span> <span class="n">rhs_end</span><span class="p">)</span>
        <span class="n">alt_sequence</span> <span class="o">=</span> <span class="n">lhs_sequence</span> <span class="o">+</span> <span class="n">alt_sequence</span> <span class="o">+</span> <span class="n">rhs_sequence</span>
    <span class="k">return</span> <span class="n">alt_sequence</span>


<div class="viewcode-block" id="AnalyzeSequences"><a class="viewcode-back" href="../../../predict.html#selene_sdk.predict.AnalyzeSequences">[docs]</a><span class="k">class</span> <span class="nc">AnalyzeSequences</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Score sequences and their variants using the predictions made</span>
<span class="sd">    by a trained model.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    model : torch.nn.Module</span>
<span class="sd">        A sequence-based model architecture.</span>
<span class="sd">    trained_model_path : str</span>
<span class="sd">        The path to the weights file for a trained sequence-based model.</span>
<span class="sd">        Architecture must match `model`.</span>
<span class="sd">    sequence_length : int</span>
<span class="sd">        The length of sequences that the model is expecting.</span>
<span class="sd">    features : list(str)</span>
<span class="sd">        The names of the features that the model is predicting.</span>
<span class="sd">    batch_size : int, optional</span>
<span class="sd">        Default is 64. The size of the mini-batches to use.</span>
<span class="sd">    use_cuda : bool, optional</span>
<span class="sd">        Default is `False`. Specifies whether CUDA-enabled GPUs are available</span>
<span class="sd">        for torch to use.</span>
<span class="sd">    reference_sequence : class, optional</span>
<span class="sd">        Default is `selene_sdk.sequences.Genome`. The type of sequence on</span>
<span class="sd">        which this analysis will be performed. Please note that if you need</span>
<span class="sd">        to use variant effect prediction, you cannot only pass in the</span>
<span class="sd">        class--you must pass in the constructed `selene_sdk.sequences.Sequence`</span>
<span class="sd">        object with a particular sequence version (e.g. `Genome(&quot;hg19.fa&quot;)`).</span>
<span class="sd">        This version does NOT have to be the same sequence version that the</span>
<span class="sd">        model was trained on. That is, if the sequences in your variants file</span>
<span class="sd">        are hg19 but your model was trained on hg38 sequences, you should pass</span>
<span class="sd">        in hg19.</span>

<span class="sd">    Attributes</span>
<span class="sd">    ----------</span>
<span class="sd">    model : torch.nn.Module</span>
<span class="sd">        A sequence-based model that has already been trained.</span>
<span class="sd">    sequence_length : int</span>
<span class="sd">        The length of sequences that the model is expecting.</span>
<span class="sd">    batch_size : int</span>
<span class="sd">        The size of the mini-batches to use.</span>
<span class="sd">    features : list(str)</span>
<span class="sd">        The names of the features that the model is predicting.</span>
<span class="sd">    use_cuda : bool</span>
<span class="sd">        Specifies whether to use CUDA or not.</span>
<span class="sd">    reference_sequence : class</span>
<span class="sd">        The type of sequence on which this analysis will be performed.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                 <span class="n">model</span><span class="p">,</span>
                 <span class="n">trained_model_path</span><span class="p">,</span>
                 <span class="n">sequence_length</span><span class="p">,</span>
                 <span class="n">features</span><span class="p">,</span>
                 <span class="n">batch_size</span><span class="o">=</span><span class="mi">64</span><span class="p">,</span>
                 <span class="n">use_cuda</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                 <span class="n">reference_sequence</span><span class="o">=</span><span class="n">Genome</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Constructs a new `AnalyzeSequences` object.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">trained_model</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">load</span><span class="p">(</span>
            <span class="n">trained_model_path</span><span class="p">,</span>
            <span class="n">map_location</span><span class="o">=</span><span class="k">lambda</span> <span class="n">storage</span><span class="p">,</span> <span class="n">location</span><span class="p">:</span> <span class="n">storage</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">model</span> <span class="o">=</span> <span class="n">load_model_from_state_dict</span><span class="p">(</span>
            <span class="n">trained_model</span><span class="p">[</span><span class="s2">&quot;state_dict&quot;</span><span class="p">],</span> <span class="n">model</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">eval</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">use_cuda</span> <span class="o">=</span> <span class="n">use_cuda</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">use_cuda</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">cuda</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span> <span class="o">=</span> <span class="n">sequence_length</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_start_radius</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sequence_length</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_end_radius</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_start_radius</span>
        <span class="k">if</span> <span class="n">sequence_length</span> <span class="o">%</span> <span class="mi">2</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_end_radius</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">batch_size</span> <span class="o">=</span> <span class="n">batch_size</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="n">features</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span> <span class="o">=</span> <span class="n">reference_sequence</span>

<div class="viewcode-block" id="AnalyzeSequences.predict"><a class="viewcode-back" href="../../../predict.html#selene_sdk.predict.AnalyzeSequences.predict">[docs]</a>    <span class="k">def</span> <span class="nf">predict</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">batch_sequences</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return model predictions for a batch of sequences.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        batch_sequences : numpy.ndarray</span>
<span class="sd">            `batch_sequences` has the shape :math:`B \\times L \\times N`,</span>
<span class="sd">            where :math:`B` is `batch_size`, :math:`L` is the sequence length,</span>
<span class="sd">            :math:`N` is the size of the sequence type&#39;s alphabet.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        numpy.ndarray</span>
<span class="sd">            The model predictions of shape :math:`B \\times F`, where :math:`F`</span>
<span class="sd">            is the number of features (classes) the model predicts.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">inputs</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">Tensor</span><span class="p">(</span><span class="n">batch_sequences</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">use_cuda</span><span class="p">:</span>
            <span class="n">inputs</span> <span class="o">=</span> <span class="n">inputs</span><span class="o">.</span><span class="n">cuda</span><span class="p">()</span>
        <span class="k">with</span> <span class="n">torch</span><span class="o">.</span><span class="n">no_grad</span><span class="p">():</span>
            <span class="n">inputs</span> <span class="o">=</span> <span class="n">Variable</span><span class="p">(</span><span class="n">inputs</span><span class="p">)</span>
            <span class="n">outputs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">forward</span><span class="p">(</span><span class="n">inputs</span><span class="o">.</span><span class="n">transpose</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
            <span class="k">return</span> <span class="n">outputs</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">cpu</span><span class="p">()</span><span class="o">.</span><span class="n">numpy</span><span class="p">()</span></div>

    <span class="k">def</span> <span class="nf">_initialize_reporters</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                              <span class="n">save_data</span><span class="p">,</span>
                              <span class="n">output_path_prefix</span><span class="p">,</span>
                              <span class="n">nonfeature_cols</span><span class="p">,</span>
                              <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;ism&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        TODO</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        save_data : list(str)</span>
<span class="sd">            A list of the data files to output. Must input 1 or more of the</span>
<span class="sd">            following options: [&quot;abs_diffs&quot;, &quot;diffs&quot;, &quot;logits&quot;, &quot;predictions&quot;].</span>
<span class="sd">        output_path_prefix : str</span>
<span class="sd">            TODO</span>
<span class="sd">        nonfeature_cols : list(str)</span>
<span class="sd">            TODO</span>
<span class="sd">        mode : str</span>
<span class="sd">            TODO</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        TODO</span>
<span class="sd">            TODO</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">reporters</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="s2">&quot;diffs&quot;</span> <span class="ow">in</span> <span class="n">save_data</span><span class="p">:</span>
            <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{0}</span><span class="s2">_diffs.tsv&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">output_path_prefix</span><span class="p">)</span>
            <span class="n">diff_handler</span> <span class="o">=</span> <span class="n">DiffScoreHandler</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">features</span><span class="p">,</span> <span class="n">nonfeature_cols</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="n">reporters</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">diff_handler</span><span class="p">)</span>
        <span class="k">if</span> <span class="s2">&quot;abs_diffs&quot;</span> <span class="ow">in</span> <span class="n">save_data</span><span class="p">:</span>
            <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{0}</span><span class="s2">_abs_diffs.tsv&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">output_path_prefix</span><span class="p">)</span>
            <span class="n">abs_diff_handler</span> <span class="o">=</span> <span class="n">AbsDiffScoreHandler</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">features</span><span class="p">,</span> <span class="n">nonfeature_cols</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="n">reporters</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">abs_diff_handler</span><span class="p">)</span>
        <span class="k">if</span> <span class="s2">&quot;logits&quot;</span> <span class="ow">in</span> <span class="n">save_data</span><span class="p">:</span>
            <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{0}</span><span class="s2">_logits.tsv&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">output_path_prefix</span><span class="p">)</span>
            <span class="n">logit_handler</span> <span class="o">=</span> <span class="n">LogitScoreHandler</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">features</span><span class="p">,</span> <span class="n">nonfeature_cols</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="n">reporters</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">logit_handler</span><span class="p">)</span>
        <span class="k">if</span> <span class="s2">&quot;predictions&quot;</span> <span class="ow">in</span> <span class="n">save_data</span> <span class="ow">and</span> <span class="n">mode</span> <span class="o">!=</span> <span class="s2">&quot;varianteffect&quot;</span><span class="p">:</span>
            <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{0}</span><span class="s2">_predictions.tsv&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">output_path_prefix</span><span class="p">)</span>
            <span class="n">preds_handler</span> <span class="o">=</span> <span class="n">WritePredictionsHandler</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">features</span><span class="p">,</span> <span class="n">nonfeature_cols</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="n">reporters</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">preds_handler</span><span class="p">)</span>
        <span class="k">elif</span> <span class="s2">&quot;predictions&quot;</span> <span class="ow">in</span> <span class="n">save_data</span> <span class="ow">and</span> <span class="n">mode</span> <span class="o">==</span> <span class="s2">&quot;varianteffect&quot;</span><span class="p">:</span>
            <span class="n">filename</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{0}</span><span class="s2">_predictions&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">output_path_prefix</span><span class="p">)</span>
            <span class="n">preds_handler</span> <span class="o">=</span> <span class="n">WriteRefAltHandler</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">features</span><span class="p">,</span> <span class="n">nonfeature_cols</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span>
            <span class="n">reporters</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">preds_handler</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">reporters</span>

    <span class="k">def</span> <span class="nf">_pad_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sequence</span><span class="p">):</span>
        <span class="n">diff</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">))</span> <span class="o">/</span> <span class="mi">2</span>
        <span class="n">pad_l</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">floor</span><span class="p">(</span><span class="n">diff</span><span class="p">))</span>
        <span class="n">pad_r</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="n">diff</span><span class="p">))</span>
        <span class="n">sequence</span> <span class="o">=</span> <span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">UNK_BASE</span> <span class="o">*</span> <span class="n">pad_l</span><span class="p">)</span> <span class="o">+</span>
                    <span class="n">sequence</span> <span class="o">+</span>
                    <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">UNK_BASE</span> <span class="o">*</span> <span class="n">pad_r</span><span class="p">))</span>
        <span class="k">return</span> <span class="nb">str</span><span class="o">.</span><span class="n">upper</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_truncate_sequence</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sequence</span><span class="p">):</span>
        <span class="n">start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">)</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span>
        <span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">start</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">)</span>
        <span class="k">return</span> <span class="nb">str</span><span class="o">.</span><span class="n">upper</span><span class="p">(</span><span class="n">sequence</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">])</span>

<div class="viewcode-block" id="AnalyzeSequences.get_predictions_for_fasta_file"><a class="viewcode-back" href="../../../predict.html#selene_sdk.predict.AnalyzeSequences.get_predictions_for_fasta_file">[docs]</a>    <span class="k">def</span> <span class="nf">get_predictions_for_fasta_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">input_path</span><span class="p">,</span> <span class="n">output_dir</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get model predictions for sequences in a FASTA file.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        input_path : str</span>
<span class="sd">            Input path to the FASTA file.</span>
<span class="sd">        output_dir : str</span>
<span class="sd">            Output directory to write the model predictions.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>
<span class="sd">            Writes the output to a file in `output_dir`.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="n">_</span><span class="p">,</span> <span class="n">filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">input_path</span><span class="p">)</span>
        <span class="n">output_prefix</span> <span class="o">=</span> <span class="s1">&#39;.&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">filename</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)[:</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>

        <span class="n">reporter</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_initialize_reporters</span><span class="p">(</span>
            <span class="p">[</span><span class="s2">&quot;predictions&quot;</span><span class="p">],</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">output_prefix</span><span class="p">),</span>
            <span class="p">[</span><span class="s2">&quot;index&quot;</span><span class="p">,</span> <span class="s2">&quot;name&quot;</span><span class="p">],</span>
            <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;prediction&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">fasta_file</span> <span class="o">=</span> <span class="n">pyfaidx</span><span class="o">.</span><span class="n">Fasta</span><span class="p">(</span><span class="n">input_path</span><span class="p">)</span>
        <span class="n">sequences</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">batch_size</span><span class="p">,</span>
                              <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">,</span>
                              <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">BASES_ARR</span><span class="p">)))</span>
        <span class="n">batch_ids</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">fasta_record</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">fasta_file</span><span class="p">):</span>
            <span class="n">cur_sequence</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">fasta_record</span><span class="p">)</span>

            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cur_sequence</span><span class="p">)</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">:</span>
                <span class="n">cur_sequence</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_pad_sequence</span><span class="p">(</span><span class="n">cur_sequence</span><span class="p">)</span>
            <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">cur_sequence</span><span class="p">)</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">:</span>
                <span class="n">cur_sequence</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_truncate_sequence</span><span class="p">(</span><span class="n">cur_sequence</span><span class="p">)</span>

            <span class="n">cur_sequence_encoding</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">sequence_to_encoding</span><span class="p">(</span>
                <span class="n">cur_sequence</span><span class="p">)</span>
            <span class="n">batch_ids</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">i</span><span class="p">,</span> <span class="n">fasta_record</span><span class="o">.</span><span class="n">name</span><span class="p">])</span>

            <span class="k">if</span> <span class="n">i</span> <span class="ow">and</span> <span class="n">i</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">batch_size</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">preds</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">sequences</span><span class="p">)</span>
                <span class="n">sequences</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">batch_size</span><span class="p">,</span> <span class="o">*</span><span class="n">cur_sequence_encoding</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
                <span class="n">reporter</span><span class="o">.</span><span class="n">handle_batch_predictions</span><span class="p">(</span><span class="n">preds</span><span class="p">,</span> <span class="n">batch_ids</span><span class="p">)</span>

            <span class="n">sequences</span><span class="p">[</span><span class="n">i</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">batch_size</span><span class="p">,</span> <span class="p">:,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">cur_sequence_encoding</span>

        <span class="k">if</span> <span class="n">i</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">batch_size</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">sequences</span> <span class="o">=</span> <span class="n">sequences</span><span class="p">[:</span><span class="n">i</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">batch_size</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="p">:,</span> <span class="p">:]</span>
            <span class="n">preds</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">sequences</span><span class="p">)</span>
            <span class="n">reporter</span><span class="o">.</span><span class="n">handle_batch_predictions</span><span class="p">(</span><span class="n">preds</span><span class="p">,</span> <span class="n">batch_ids</span><span class="p">)</span>

        <span class="n">fasta_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="n">reporter</span><span class="o">.</span><span class="n">write_to_file</span><span class="p">(</span><span class="n">close</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></div>


<div class="viewcode-block" id="AnalyzeSequences.in_silico_mutagenesis_predict"><a class="viewcode-back" href="../../../predict.html#selene_sdk.predict.AnalyzeSequences.in_silico_mutagenesis_predict">[docs]</a>    <span class="k">def</span> <span class="nf">in_silico_mutagenesis_predict</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                                      <span class="n">sequence</span><span class="p">,</span>
                                      <span class="n">base_preds</span><span class="p">,</span>
                                      <span class="n">mutations_list</span><span class="p">,</span>
                                      <span class="n">reporters</span><span class="o">=</span><span class="p">[]):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get the predictions for all specified mutations applied</span>
<span class="sd">        to a given sequence and, if applicable, compute the scores</span>
<span class="sd">        (&quot;abs_diffs&quot;, &quot;diffs&quot;, &quot;logits&quot;) for these mutations.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        sequence : str</span>
<span class="sd">            The sequence to mutate.</span>
<span class="sd">        base_preds : numpy.ndarray</span>
<span class="sd">            The model&#39;s prediction for `sequence`.</span>
<span class="sd">        mutations_list : list(list(tuple))</span>
<span class="sd">            The mutations to apply to the sequence. Each element in</span>
<span class="sd">            `mutations_list` is a list of tuples, where each tuple</span>
<span class="sd">            specifies the `int` position in the sequence to mutate and what</span>
<span class="sd">            `str` base to which the position is mutated (e.g. (1, &#39;A&#39;)).</span>
<span class="sd">        reporters : list(PredictionsHandler)</span>
<span class="sd">            The list of reporters, where each reporter handles the predictions</span>
<span class="sd">            made for each mutated sequence. Will collect, compute scores</span>
<span class="sd">            (e.g. `AbsDiffScoreHandler` computes the absolute difference</span>
<span class="sd">            between `base_preds` and the predictions for the mutated</span>
<span class="sd">            sequence), and output these as a file at the end.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>
<span class="sd">            Writes results to files corresponding to each reporter in</span>
<span class="sd">            `reporters`.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">current_sequence_encoding</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">sequence_to_encoding</span><span class="p">(</span>
            <span class="n">sequence</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">mutations_list</span><span class="p">),</span> <span class="bp">self</span><span class="o">.</span><span class="n">batch_size</span><span class="p">):</span>
            <span class="n">start</span> <span class="o">=</span> <span class="n">i</span>
            <span class="n">end</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">batch_size</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">mutations_list</span><span class="p">))</span>

            <span class="n">mutated_sequences</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span>
                <span class="p">(</span><span class="n">end</span> <span class="o">-</span> <span class="n">start</span><span class="p">,</span> <span class="o">*</span><span class="n">current_sequence_encoding</span><span class="o">.</span><span class="n">shape</span><span class="p">))</span>

            <span class="n">batch_ids</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">ix</span><span class="p">,</span> <span class="n">mutation_info</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">mutations_list</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">]):</span>
                <span class="n">mutated_seq</span> <span class="o">=</span> <span class="n">mutate_sequence</span><span class="p">(</span>
                    <span class="n">current_sequence_encoding</span><span class="p">,</span> <span class="n">mutation_info</span><span class="p">,</span>
                    <span class="n">reference_sequence</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="p">)</span>
                <span class="n">mutated_sequences</span><span class="p">[</span><span class="n">ix</span><span class="p">,</span> <span class="p">:,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">mutated_seq</span>
                <span class="n">batch_ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">_ism_sample_id</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">mutation_info</span><span class="p">))</span>
            <span class="n">outputs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">mutated_sequences</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">reporters</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">r</span><span class="o">.</span><span class="n">needs_base_pred</span><span class="p">:</span>
                    <span class="n">r</span><span class="o">.</span><span class="n">handle_batch_predictions</span><span class="p">(</span><span class="n">outputs</span><span class="p">,</span> <span class="n">batch_ids</span><span class="p">,</span> <span class="n">base_preds</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">r</span><span class="o">.</span><span class="n">handle_batch_predictions</span><span class="p">(</span><span class="n">outputs</span><span class="p">,</span> <span class="n">batch_ids</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">reporters</span><span class="p">:</span>
            <span class="n">r</span><span class="o">.</span><span class="n">write_to_file</span><span class="p">(</span><span class="n">close</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></div>

<div class="viewcode-block" id="AnalyzeSequences.in_silico_mutagenesis"><a class="viewcode-back" href="../../../predict.html#selene_sdk.predict.AnalyzeSequences.in_silico_mutagenesis">[docs]</a>    <span class="k">def</span> <span class="nf">in_silico_mutagenesis</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                              <span class="n">sequence</span><span class="p">,</span>
                              <span class="n">save_data</span><span class="p">,</span>
                              <span class="n">output_path_prefix</span><span class="o">=</span><span class="s2">&quot;ism&quot;</span><span class="p">,</span>
                              <span class="n">mutate_n_bases</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Applies *in silico* mutagenesis to a sequence.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        sequence : str</span>
<span class="sd">            The sequence to mutate.</span>
<span class="sd">        save_data : list(str)</span>
<span class="sd">            A list of the data files to output. Must input 1 or more of the</span>
<span class="sd">            following options: [&quot;abs_diffs&quot;, &quot;diffs&quot;, &quot;logits&quot;, &quot;predictions&quot;].</span>
<span class="sd">        output_path_prefix : str, optional</span>
<span class="sd">            The path to which the data files are written. If directories in</span>
<span class="sd">            the path do not yet exist they will be automatically created.</span>
<span class="sd">        mutate_n_bases : int, optional</span>
<span class="sd">            The number of bases to mutate at one time. We recommend leaving</span>
<span class="sd">            this parameter set to `1` at this time, as we have not yet</span>
<span class="sd">            optimized operations for double and triple mutations.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>
<span class="sd">            Writes results to files corresponding to each reporter in</span>
<span class="sd">            `reporters`.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">path_dirs</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">output_path_prefix</span><span class="p">)</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">path_dirs</span><span class="p">,</span> <span class="n">exists_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="n">n</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">n</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">:</span> <span class="c1"># Pad string length as necessary.</span>
             <span class="n">diff</span> <span class="o">=</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span> <span class="o">-</span> <span class="n">n</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span>
             <span class="n">pad_l</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">floor</span><span class="p">(</span><span class="n">diff</span><span class="p">))</span>
             <span class="n">pad_r</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="n">diff</span><span class="p">))</span>
             <span class="n">sequence</span> <span class="o">=</span> <span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">UNK_BASE</span> <span class="o">*</span> <span class="n">pad_l</span><span class="p">)</span> <span class="o">+</span>
                         <span class="n">sequence</span> <span class="o">+</span>
                         <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">UNK_BASE</span> <span class="o">*</span> <span class="n">pad_r</span><span class="p">))</span>
        <span class="k">elif</span> <span class="n">n</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">:</span>  <span class="c1"># Extract center substring of proper length.</span>
            <span class="n">start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">((</span><span class="n">n</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">)</span> <span class="o">//</span> <span class="mi">2</span><span class="p">)</span>
            <span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">start</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">)</span>
            <span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">]</span>

        <span class="n">sequence</span> <span class="o">=</span> <span class="nb">str</span><span class="o">.</span><span class="n">upper</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
        <span class="n">mutated_sequences</span> <span class="o">=</span> <span class="n">in_silico_mutagenesis_sequences</span><span class="p">(</span>
            <span class="n">sequence</span><span class="p">,</span> <span class="n">mutate_n_bases</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
            <span class="n">reference_sequence</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="p">)</span>

        <span class="n">reporters</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_initialize_reporters</span><span class="p">(</span>
            <span class="n">save_data</span><span class="p">,</span> <span class="n">output_path_prefix</span><span class="p">,</span> <span class="n">ISM_COLS</span><span class="p">)</span>

        <span class="n">current_sequence_encoding</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">sequence_to_encoding</span><span class="p">(</span>
            <span class="n">sequence</span><span class="p">)</span>

        <span class="n">base_encoding</span> <span class="o">=</span> <span class="n">current_sequence_encoding</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span>
            <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="o">*</span><span class="n">current_sequence_encoding</span><span class="o">.</span><span class="n">shape</span><span class="p">))</span>
        <span class="n">base_preds</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">base_encoding</span><span class="p">)</span>

        <span class="k">if</span> <span class="s2">&quot;predictions&quot;</span> <span class="ow">in</span> <span class="n">save_data</span><span class="p">:</span>
            <span class="n">predictions_reporter</span> <span class="o">=</span> <span class="n">reporters</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">predictions_reporter</span><span class="o">.</span><span class="n">handle_batch_predictions</span><span class="p">(</span>
                <span class="n">base_preds</span><span class="p">,</span> <span class="p">[[</span><span class="s2">&quot;NA&quot;</span><span class="p">,</span> <span class="s2">&quot;NA&quot;</span><span class="p">,</span> <span class="s2">&quot;NA&quot;</span><span class="p">]])</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">in_silico_mutagenesis_predict</span><span class="p">(</span>
            <span class="n">sequence</span><span class="p">,</span> <span class="n">base_preds</span><span class="p">,</span> <span class="n">mutated_sequences</span><span class="p">,</span> <span class="n">reporters</span><span class="o">=</span><span class="n">reporters</span><span class="p">)</span></div>

<div class="viewcode-block" id="AnalyzeSequences.in_silico_mutagenesis_from_file"><a class="viewcode-back" href="../../../predict.html#selene_sdk.predict.AnalyzeSequences.in_silico_mutagenesis_from_file">[docs]</a>    <span class="k">def</span> <span class="nf">in_silico_mutagenesis_from_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                                        <span class="n">input_path</span><span class="p">,</span>
                                        <span class="n">save_data</span><span class="p">,</span>
                                        <span class="n">output_dir</span><span class="p">,</span>
                                        <span class="n">mutate_n_bases</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
                                        <span class="n">use_sequence_name</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Apply *in silico* mutagenesis to all sequences in a FASTA file.</span>

<span class="sd">        Please note that we have not parallelized this function yet, so runtime</span>
<span class="sd">        increases exponentially when you increase `mutate_n_bases`.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        input_path: str</span>
<span class="sd">            The path to the FASTA file of sequences.</span>
<span class="sd">        save_data : list(str)</span>
<span class="sd">            A list of the data files to output. Must input 1 or more of the</span>
<span class="sd">            following options: [&quot;abs_diffs&quot;, &quot;diffs&quot;, &quot;logits&quot;, &quot;predictions&quot;].</span>
<span class="sd">        output_dir : str</span>
<span class="sd">            The path to the output directory. Directories in the path will be</span>
<span class="sd">            created if they do not currently exist.</span>
<span class="sd">        mutate_n_bases : int, optional</span>
<span class="sd">            Default is 1. The number of bases to mutate at one time in</span>
<span class="sd">            *in silico* mutagenesis.</span>
<span class="sd">        use_sequence_name : bool, optional.</span>
<span class="sd">            Default is True. If `use_sequence_name`, output files are prefixed</span>
<span class="sd">            by the sequence name/description corresponding to each sequence</span>
<span class="sd">            in the FASTA file. Spaces in the sequence name are replaced with</span>
<span class="sd">            underscores &#39;_&#39;. If not `use_sequence_name`, output files are</span>
<span class="sd">            prefixed with an index :math:`i` (starting with 0) corresponding</span>
<span class="sd">            to the :math:`i`th sequence in the FASTA file.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>
<span class="sd">            Outputs data files from *in silico* mutagenesis to `output_dir`.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="n">fasta_file</span> <span class="o">=</span> <span class="n">pyfaidx</span><span class="o">.</span><span class="n">Fasta</span><span class="p">(</span><span class="n">input_path</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">fasta_record</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">fasta_file</span><span class="p">):</span>
            <span class="n">cur_sequence</span> <span class="o">=</span> <span class="nb">str</span><span class="o">.</span><span class="n">upper</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">fasta_record</span><span class="p">))</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cur_sequence</span><span class="p">)</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">:</span>
                <span class="n">cur_sequence</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_pad_sequence</span><span class="p">(</span><span class="n">cur_sequence</span><span class="p">)</span>
            <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">cur_sequence</span><span class="p">)</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">:</span>
                <span class="n">cur_sequence</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_truncate_sequence</span><span class="p">(</span><span class="n">cur_sequence</span><span class="p">)</span>

            <span class="c1"># Generate mut sequences and base preds.</span>
            <span class="n">mutated_sequences</span> <span class="o">=</span> <span class="n">in_silico_mutagenesis_sequences</span><span class="p">(</span>
                <span class="n">cur_sequence</span><span class="p">,</span> <span class="n">mutate_n_bases</span><span class="o">=</span><span class="n">mutate_n_bases</span><span class="p">,</span>
                <span class="n">reference_sequence</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="p">)</span>
            <span class="n">cur_sequence_encoding</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">sequence_to_encoding</span><span class="p">(</span>
                <span class="n">cur_sequence</span><span class="p">)</span>
            <span class="n">base_encoding</span> <span class="o">=</span> <span class="n">cur_sequence_encoding</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span>
                <span class="mi">1</span><span class="p">,</span> <span class="o">*</span><span class="n">cur_sequence_encoding</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
            <span class="n">base_preds</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">base_encoding</span><span class="p">)</span>

            <span class="n">file_prefix</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="k">if</span> <span class="n">use_sequence_name</span><span class="p">:</span>
                <span class="n">file_prefix</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                    <span class="n">output_dir</span><span class="p">,</span> <span class="n">fasta_record</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="s1">&#39;_&#39;</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">file_prefix</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                    <span class="n">output_dir</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>
            <span class="c1"># Write base to file, and make mut preds.</span>
            <span class="n">reporters</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_initialize_reporters</span><span class="p">(</span>
                <span class="n">save_data</span><span class="p">,</span> <span class="n">file_prefix</span><span class="p">,</span> <span class="n">ISM_COLS</span><span class="p">)</span>

            <span class="k">if</span> <span class="s2">&quot;predictions&quot;</span> <span class="ow">in</span> <span class="n">save_data</span><span class="p">:</span>
                <span class="n">predictions_reporter</span> <span class="o">=</span> <span class="n">reporters</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
                <span class="n">predictions_reporter</span><span class="o">.</span><span class="n">handle_batch_predictions</span><span class="p">(</span>
                    <span class="n">base_preds</span><span class="p">,</span> <span class="p">[[</span><span class="s2">&quot;NA&quot;</span><span class="p">,</span> <span class="s2">&quot;NA&quot;</span><span class="p">,</span> <span class="s2">&quot;NA&quot;</span><span class="p">]])</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">in_silico_mutagenesis_predict</span><span class="p">(</span>
                <span class="n">cur_sequence</span><span class="p">,</span> <span class="n">base_preds</span><span class="p">,</span> <span class="n">mutated_sequences</span><span class="p">,</span>
                <span class="n">reporters</span><span class="o">=</span><span class="n">reporters</span><span class="p">)</span>
        <span class="n">fasta_file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>

    <span class="k">def</span> <span class="nf">_handle_ref_alt_predictions</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                                   <span class="n">batch_ref_seqs</span><span class="p">,</span>
                                   <span class="n">batch_alt_seqs</span><span class="p">,</span>
                                   <span class="n">batch_ids</span><span class="p">,</span>
                                   <span class="n">reporters</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        TODO</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        batch_ref_seqs : list(np.ndarray)</span>
<span class="sd">            TODO</span>
<span class="sd">        batch_alt_seqs : list(np.ndarray)</span>
<span class="sd">            TODO</span>
<span class="sd">        reporters : list(PredictionsHandler)</span>
<span class="sd">            TODO</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">batch_ref_seqs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">batch_ref_seqs</span><span class="p">)</span>
        <span class="n">batch_alt_seqs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">batch_alt_seqs</span><span class="p">)</span>

        <span class="n">ref_outputs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">batch_ref_seqs</span><span class="p">)</span>
        <span class="n">alt_outputs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">batch_alt_seqs</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">reporters</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">r</span><span class="o">.</span><span class="n">needs_base_pred</span><span class="p">:</span>
                <span class="n">r</span><span class="o">.</span><span class="n">handle_batch_predictions</span><span class="p">(</span>
                    <span class="n">alt_outputs</span><span class="p">,</span> <span class="n">batch_ids</span><span class="p">,</span> <span class="n">ref_outputs</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">r</span><span class="o">.</span><span class="n">handle_batch_predictions</span><span class="p">(</span><span class="n">alt_outputs</span><span class="p">,</span> <span class="n">batch_ids</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_process_alts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                      <span class="n">all_alts</span><span class="p">,</span>
                      <span class="n">ref</span><span class="p">,</span>
                      <span class="n">chrom</span><span class="p">,</span>
                      <span class="n">start</span><span class="p">,</span>
                      <span class="n">end</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        TODO</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        all_alts : TODO</span>
<span class="sd">            TODO</span>
<span class="sd">        ref : TODO</span>
<span class="sd">            TODO</span>
<span class="sd">        chrom : str</span>
<span class="sd">            TODO</span>
<span class="sd">        start : TODO</span>
<span class="sd">            TODO</span>
<span class="sd">        end : TODO</span>
<span class="sd">            TODO</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        TODO</span>
<span class="sd">            TODO</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">sequence</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">get_sequence_from_coords</span><span class="p">(</span>
            <span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span>

        <span class="n">start_pos</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_start_radius</span> <span class="o">-</span> <span class="nb">int</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ref</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>

        <span class="n">alt_encodings</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">all_alts</span><span class="p">:</span>
            <span class="n">prefix</span> <span class="o">=</span> <span class="n">sequence</span><span class="p">[:</span><span class="n">start_pos</span><span class="p">]</span>
            <span class="n">suffix</span> <span class="o">=</span> <span class="n">sequence</span><span class="p">[</span><span class="n">start_pos</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">ref</span><span class="p">):]</span>
            <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">prefix</span><span class="p">)</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">suffix</span><span class="p">)</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">ref</span><span class="p">)</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span>
            <span class="k">if</span> <span class="n">a</span> <span class="o">==</span> <span class="s1">&#39;*&#39;</span><span class="p">:</span>  <span class="c1"># indicates a deletion</span>
                <span class="n">alt_sequence</span> <span class="o">=</span> <span class="n">prefix</span> <span class="o">+</span> <span class="n">suffix</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">alt_sequence</span> <span class="o">=</span> <span class="n">prefix</span> <span class="o">+</span> <span class="n">a</span> <span class="o">+</span> <span class="n">suffix</span>

            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">alt_sequence</span><span class="p">)</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">:</span>
                <span class="c1"># truncate on both sides equally</span>
                <span class="n">midpoint</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">alt_sequence</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
                <span class="n">start</span> <span class="o">=</span> <span class="n">midpoint</span> <span class="o">-</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
                <span class="n">end</span> <span class="o">=</span> <span class="n">midpoint</span> <span class="o">+</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span> <span class="o">%</span> <span class="mi">2</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">end</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">alt_sequence</span> <span class="o">=</span> <span class="n">alt_sequence</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">]</span>
            <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">alt_sequence</span><span class="p">)</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">:</span>
                <span class="n">alt_sequence</span> <span class="o">=</span> <span class="n">_add_sequence_surrounding_alt</span><span class="p">(</span>
                    <span class="n">alt_sequence</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span><span class="p">,</span>
                    <span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="p">)</span>
            <span class="n">alt_encoding</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">sequence_to_encoding</span><span class="p">(</span>
                <span class="n">alt_sequence</span><span class="p">)</span>
            <span class="n">alt_encodings</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">alt_encoding</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">alt_encodings</span>

<div class="viewcode-block" id="AnalyzeSequences.variant_effect_prediction"><a class="viewcode-back" href="../../../predict.html#selene_sdk.predict.AnalyzeSequences.variant_effect_prediction">[docs]</a>    <span class="k">def</span> <span class="nf">variant_effect_prediction</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                                  <span class="n">vcf_file</span><span class="p">,</span>
                                  <span class="n">save_data</span><span class="p">,</span>
                                  <span class="n">output_dir</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get model predictions and scores for a list of variants.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        vcf_file : str</span>
<span class="sd">            Path to a VCF file. Must contain the columns</span>
<span class="sd">            [#CHROM, POS, ID, REF, ALT], in order. Column header does not need</span>
<span class="sd">            to be present.</span>
<span class="sd">        save_data : list(str)</span>
<span class="sd">            A list of the data files to output. Must input 1 or more of the</span>
<span class="sd">            following options: [&quot;abs_diffs&quot;, &quot;diffs&quot;, &quot;logits&quot;, &quot;predictions&quot;].</span>
<span class="sd">        output_dir : str or None, optional</span>
<span class="sd">            Path to the output directory. If no path is specified, will save</span>
<span class="sd">            files corresponding to the options in `save_data` to the</span>
<span class="sd">            current working directory.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        None</span>
<span class="sd">            Saves all files to `output_dir`.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">variants</span> <span class="o">=</span> <span class="n">read_vcf_file</span><span class="p">(</span><span class="n">vcf_file</span><span class="p">)</span>

        <span class="c1"># TODO: GIVE USER MORE CONTROL OVER PREFIX.</span>
        <span class="n">path</span><span class="p">,</span> <span class="n">filename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">vcf_file</span><span class="p">)</span>
        <span class="n">output_path_prefix</span> <span class="o">=</span> <span class="s1">&#39;.&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">filename</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)[:</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
        <span class="k">if</span> <span class="n">output_dir</span><span class="p">:</span>
            <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">output_dir</span> <span class="o">=</span> <span class="n">path</span>
        <span class="n">output_path_prefix</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="n">output_path_prefix</span><span class="p">)</span>
        <span class="n">reporters</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_initialize_reporters</span><span class="p">(</span>
            <span class="n">save_data</span><span class="p">,</span> <span class="n">output_path_prefix</span><span class="p">,</span> <span class="n">VARIANTEFFECT_COLS</span><span class="p">,</span>
            <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;varianteffect&quot;</span><span class="p">)</span>

        <span class="n">batch_ref_seqs</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">batch_alt_seqs</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">batch_ids</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">alt</span><span class="p">)</span> <span class="ow">in</span> <span class="n">variants</span><span class="p">:</span>
            <span class="n">center</span> <span class="o">=</span> <span class="n">pos</span> <span class="o">+</span> <span class="nb">int</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ref</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span>
            <span class="n">start</span> <span class="o">=</span> <span class="n">center</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">_start_radius</span>
            <span class="n">end</span> <span class="o">=</span> <span class="n">center</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">_end_radius</span>
            <span class="k">assert</span> <span class="n">end</span> <span class="o">-</span> <span class="n">start</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="p">,</span> <span class="n">Genome</span><span class="p">):</span>
                <span class="k">if</span> <span class="s2">&quot;chr&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">chrom</span><span class="p">:</span>
                    <span class="n">chrom</span> <span class="o">=</span> <span class="s2">&quot;chr&quot;</span> <span class="o">+</span> <span class="n">chrom</span>
                <span class="k">if</span> <span class="s2">&quot;MT&quot;</span> <span class="ow">in</span> <span class="n">chrom</span><span class="p">:</span>
                    <span class="n">chrom</span> <span class="o">=</span> <span class="n">chrom</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">coords_in_bounds</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">):</span>
                <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">reporters</span><span class="p">:</span>
                    <span class="n">r</span><span class="o">.</span><span class="n">handle_NA</span><span class="p">((</span><span class="n">chrom</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">alt</span><span class="p">))</span>
                <span class="k">continue</span>
            <span class="n">seq_encoding</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">get_encoding_from_coords</span><span class="p">(</span>
                <span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span>
            <span class="n">ref_encoding</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">sequence_to_encoding</span><span class="p">(</span><span class="n">ref</span><span class="p">)</span>
            <span class="n">start_pos</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_start_radius</span> <span class="o">-</span> <span class="nb">int</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ref</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
            <span class="n">seq_encoding</span><span class="p">[</span><span class="n">start_pos</span><span class="p">:</span><span class="n">start_pos</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">ref</span><span class="p">),</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">ref_encoding</span>

            <span class="n">all_alts</span> <span class="o">=</span> <span class="n">alt</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)</span>
            <span class="n">alt_encodings</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_process_alts</span><span class="p">(</span>
                <span class="n">all_alts</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">all_alts</span><span class="p">:</span>
                <span class="n">batch_ref_seqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">seq_encoding</span><span class="p">)</span>
                <span class="n">batch_ids</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">chrom</span><span class="p">,</span> <span class="n">pos</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">ref</span><span class="p">,</span> <span class="n">a</span><span class="p">))</span>
            <span class="n">batch_alt_seqs</span> <span class="o">+=</span> <span class="n">alt_encodings</span>

            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">batch_ref_seqs</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">batch_size</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_handle_ref_alt_predictions</span><span class="p">(</span>
                    <span class="n">batch_ref_seqs</span><span class="p">,</span> <span class="n">batch_alt_seqs</span><span class="p">,</span> <span class="n">batch_ids</span><span class="p">,</span> <span class="n">reporters</span><span class="p">)</span>
                <span class="n">batch_ref_seqs</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="n">batch_alt_seqs</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="n">batch_ids</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">if</span> <span class="n">batch_ref_seqs</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_handle_ref_alt_predictions</span><span class="p">(</span>
                <span class="n">batch_ref_seqs</span><span class="p">,</span> <span class="n">batch_alt_seqs</span><span class="p">,</span> <span class="n">batch_ids</span><span class="p">,</span> <span class="n">reporters</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">reporters</span><span class="p">:</span>
            <span class="n">r</span><span class="o">.</span><span class="n">write_to_file</span><span class="p">(</span><span class="n">close</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></div></div>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Selene Contributors

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../../../_static/jquery.js"></script>
        <script type="text/javascript" src="../../../_static/underscore.js"></script>
        <script type="text/javascript" src="../../../_static/doctools.js"></script>
        <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    

  

  <script type="text/javascript" src="../../../_static/js/theme.js"></script>

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>
 



</body>
</html>