

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>selene.samplers.online_sampler &mdash; Selene master documentation</title>
  

  
  
  
  

  

  
  
    

  

  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Raleway" type="text/css" />
  <link rel="stylesheet" href="../../../_static/css/selene_docs_theme.css" type="text/css" />
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 

  
  <script src="../../../_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../../../index.html" class="icon icon-home"> Selene
          

          
            
            <img src="../../../_static/selene_logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                master (0.0.0)
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          

            
            
              
            
            
              <p class="caption"><span class="caption-text">Overview</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/faq.html">Frequently Asked Questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../overview/extending.html">Extending Selene</a></li>
</ul>
<p class="caption"><span class="caption-text">Packages</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../selene.html">selene</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../selene.html#trainmodel">TrainModel</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../interpret.html">selene.interpret</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#sequence-logo">sequence_logo</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#heatmap">heatmap</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#ismresult">ISMResult</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../interpret.html#rescale-score-matrix">rescale_score_matrix</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../predict.html">selene.predict</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../predict.html#analyzesequences">AnalyzeSequences</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../predict.predict_handlers.html">selene.predict.predict_handlers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../predict.predict_handlers.html#predictionshandler">PredictionsHandler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../predict.predict_handlers.html#diffscorehandler">DiffScoreHandler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../predict.predict_handlers.html#logitscorehandler">LogitScoreHandler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../predict.predict_handlers.html#writepredictionshandler">WritePredictionsHandler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../predict.predict_handlers.html#writerefalthandler">WriteRefAltHandler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../predict.predict_handlers.html#write-to-file">write_to_file</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../samplers.html">selene.samplers</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#sampler">Sampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#onlinesampler">OnlineSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#intervalssampler">IntervalsSampler</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../samplers.html#matfilesampler">MatFileSampler</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../sequences.html">selene.sequences</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#sequence">Sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#genome">Genome</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#proteome">Proteome</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#sequence-to-encoding">sequence_to_encoding</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#encoding-to-sequence">encoding_to_sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../sequences.html#get-reverse-encoding">get_reverse_encoding</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../targets.html">selene.targets</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../targets.html#target">Target</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../targets.html#genomicfeatures">GenomicFeatures</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../utils.html">selene.utils</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#performancemetrics">PerformanceMetrics</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#initialize-logger">initialize_logger</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#load-features-list">load_features_list</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../utils.html#configuration">Configuration</a></li>
</ul>
</li>
</ul>

            
          

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">Selene</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>selene.samplers.online_sampler</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for selene.samplers.online_sampler</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">This module provides the `OnlineSampler` class and supporting methods.</span>
<span class="sd">Objects of the class `OnlineSampler`, are samplers which load examples</span>
<span class="sd">&quot;on the fly&quot; rather than storing them all persistently in memory.</span>

<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">from</span> <span class="nn">abc</span> <span class="k">import</span> <span class="n">ABCMeta</span>
<span class="kn">import</span> <span class="nn">os</span>

<span class="kn">from</span> <span class="nn">.sampler</span> <span class="k">import</span> <span class="n">Sampler</span>
<span class="kn">from</span> <span class="nn">..targets</span> <span class="k">import</span> <span class="n">GenomicFeatures</span>


<div class="viewcode-block" id="OnlineSampler"><a class="viewcode-back" href="../../../samplers.html#selene.samplers.OnlineSampler">[docs]</a><span class="k">class</span> <span class="nc">OnlineSampler</span><span class="p">(</span><span class="n">Sampler</span><span class="p">,</span> <span class="n">metaclass</span><span class="o">=</span><span class="n">ABCMeta</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A sampler in which training/validation/test data is constructed</span>
<span class="sd">    from random sampling of the dataset for each batch passed to the</span>
<span class="sd">    model. This form of sampling may alleviate the problem of loading an</span>
<span class="sd">    extremely large dataset into memory when developing a new model.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    reference_sequence : selene.sequences.Sequence</span>
<span class="sd">        A reference sequence from which to create examples.</span>
<span class="sd">    target_path : str</span>
<span class="sd">        Path to tabix-indexed, compressed BED file (`*.bed.gz`) of genomic</span>
<span class="sd">        coordinates mapped to the genomic features we want to predict.</span>
<span class="sd">    features : list(str)</span>
<span class="sd">        List of distinct features that we aim to predict.</span>
<span class="sd">    seed : int, optional</span>
<span class="sd">        Default is 436. Sets the random seed for sampling.</span>
<span class="sd">    validation_holdout : list(str) or float, optional</span>
<span class="sd">        Default is `[&#39;chr6&#39;, &#39;chr7&#39;]`. Holdout can be regional or</span>
<span class="sd">        proportional. If regional, expects a list (e.g. `[&#39;X&#39;, &#39;Y&#39;]`).</span>
<span class="sd">        Regions must match those specified in the first column of the</span>
<span class="sd">        tabix-indexed BED file. If proportional, specify a percentage</span>
<span class="sd">        between (0.0, 1.0). Typically 0.10 or 0.20.</span>
<span class="sd">    test_holdout : list(str) or float, optional</span>
<span class="sd">        Default is `[&#39;chr8&#39;, &#39;chr9&#39;]`. See documentation for</span>
<span class="sd">        `validation_holdout` for additional information.</span>
<span class="sd">    sequence_length : int, optional</span>
<span class="sd">        Default is 1001. Model is trained on sequences of `sequence_length`</span>
<span class="sd">        where genomic features are annotated to the center regions of</span>
<span class="sd">        these sequences.</span>
<span class="sd">    center_bin_to_predict : int, optional</span>
<span class="sd">        Default is 201. Query the tabix-indexed file for a region of</span>
<span class="sd">        length `center_bin_to_predict`.</span>
<span class="sd">    feature_thresholds : float [0.0, 1.0], optional</span>
<span class="sd">        Default is 0.5. The `feature_threshold` to pass to the</span>
<span class="sd">        `GenomicFeatures` object.</span>
<span class="sd">    mode : {&#39;train&#39;, &#39;validate&#39;, &#39;test&#39;}</span>
<span class="sd">        Default is `&#39;train&#39;`. The mode to run the sampler in.</span>
<span class="sd">    save_datasets : list(str)</span>
<span class="sd">        Default is `[&quot;test&quot;]`. The list of modes for which we should</span>
<span class="sd">        save the sampled data to file.</span>

<span class="sd">    Attributes</span>
<span class="sd">    ----------</span>
<span class="sd">    reference_sequence : selene.sequences.Sequence</span>
<span class="sd">        The reference sequence that examples are created from.</span>
<span class="sd">    target : selene.targets.Target</span>
<span class="sd">        The `selene.targets.Target` object holding the features that we</span>
<span class="sd">        would like to predict.</span>
<span class="sd">    validation_holdout : list(str) or float</span>
<span class="sd">        The samples to hold out for validating model performance. These</span>
<span class="sd">        can be &quot;regional&quot; or &quot;proportional&quot;. If regional, this is a list</span>
<span class="sd">        of region names (e.g. `[&#39;chrX&#39;, &#39;chrY&#39;]`). These Regions must</span>
<span class="sd">        match those specified in the first column of the tabix-indexed</span>
<span class="sd">        BED file. If proportional, this is the fraction of total samples</span>
<span class="sd">        that will be held out.</span>
<span class="sd">    test_holdout : list(str) or float</span>
<span class="sd">        The samples to hold out for testing model performance. See the</span>
<span class="sd">        documentation for `validation_holdout` for more details.</span>
<span class="sd">    sequence_length : int</span>
<span class="sd">        The length of the sequences to  train the model on.</span>
<span class="sd">    bin_radius : int</span>
<span class="sd">        From the center of the sequence, the radius in which to detect</span>
<span class="sd">        a feature annotation in order to include it as a sample&#39;s label.</span>
<span class="sd">    surrounding_sequence_radius : int</span>
<span class="sd">        The length of sequence falling outside of the feature detection</span>
<span class="sd">        bin (i.e. `bin_radius`) center, but still within the</span>
<span class="sd">        `sequence_length`.</span>
<span class="sd">    modes : list(str)</span>
<span class="sd">        The list of modes that the sampler can be run in.</span>
<span class="sd">    mode : str</span>
<span class="sd">        The current mode that the sampler is running in. Must be one of</span>
<span class="sd">        the modes listed in `modes`.</span>
<span class="sd">    save_datasets : list(str)</span>
<span class="sd">        A list of modes for which we should write the sampled data to a</span>
<span class="sd">        file.</span>

<span class="sd">    Raises</span>
<span class="sd">    ------</span>
<span class="sd">    ValueError</span>
<span class="sd">            If `mode` is not a valid mode.</span>
<span class="sd">    ValueError</span>
<span class="sd">        If the parities of `sequence_length` and `center_bin_to_predict`</span>
<span class="sd">        are not the same.</span>
<span class="sd">    ValueError</span>
<span class="sd">        If `sequence_length` is smaller than `center_bin_to_predict` is.</span>
<span class="sd">    ValueError</span>
<span class="sd">        If the types of `validation_holdout` and `test_holdout` are not</span>
<span class="sd">        the same.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># @TODO: Is there a compelling reason why STRAND_SIDES should be here?</span>
    <span class="n">STRAND_SIDES</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;-&#39;</span><span class="p">)</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Defines the strands that features can be sampled from.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                 <span class="n">reference_sequence</span><span class="p">,</span>
                 <span class="n">target_path</span><span class="p">,</span>
                 <span class="n">features</span><span class="p">,</span>
                 <span class="n">seed</span><span class="o">=</span><span class="mi">436</span><span class="p">,</span>
                 <span class="n">validation_holdout</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;chr6&#39;</span><span class="p">,</span> <span class="s1">&#39;chr7&#39;</span><span class="p">],</span>
                 <span class="n">test_holdout</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;chr8&#39;</span><span class="p">,</span> <span class="s1">&#39;chr9&#39;</span><span class="p">],</span>
                 <span class="n">sequence_length</span><span class="o">=</span><span class="mi">1001</span><span class="p">,</span>
                 <span class="n">center_bin_to_predict</span><span class="o">=</span><span class="mi">201</span><span class="p">,</span>
                 <span class="n">feature_thresholds</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
                 <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;train&quot;</span><span class="p">,</span>
                 <span class="n">save_datasets</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;test&quot;</span><span class="p">]):</span>

        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Creates a new `OnlineSampler` object.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="nb">super</span><span class="p">(</span><span class="n">OnlineSampler</span><span class="p">,</span> <span class="bp">self</span><span class="p">)</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">seed</span><span class="o">=</span><span class="n">seed</span><span class="p">)</span>

        <span class="k">if</span> <span class="p">(</span><span class="n">sequence_length</span> <span class="o">+</span> <span class="n">center_bin_to_predict</span><span class="p">)</span> <span class="o">%</span> <span class="mi">2</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Sequence length of </span><span class="si">{0}</span><span class="s2"> with a center bin length of </span><span class="si">{1}</span><span class="s2"> &quot;</span>
                <span class="s2">&quot;is invalid. These 2 inputs should both be odd or both be &quot;</span>
                <span class="s2">&quot;even.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">sequence_length</span><span class="p">,</span> <span class="n">center_bin_to_predict</span><span class="p">))</span>

        <span class="n">surrounding_sequence_length</span> <span class="o">=</span> <span class="n">sequence_length</span> <span class="o">-</span> <span class="n">center_bin_to_predict</span>
        <span class="k">if</span> <span class="n">surrounding_sequence_length</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Sequence length of </span><span class="si">{0}</span><span class="s2"> is less than the center bin &quot;</span>
                <span class="s2">&quot;length of </span><span class="si">{1}</span><span class="s2">.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">sequence_length</span><span class="p">,</span> <span class="n">center_bin_to_predict</span><span class="p">))</span>

        <span class="c1"># specifying a test holdout partition is optional</span>
        <span class="k">if</span> <span class="n">test_holdout</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">modes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;test&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">validation_holdout</span><span class="p">,</span> <span class="p">(</span><span class="nb">list</span><span class="p">,))</span> <span class="ow">and</span> \
                    <span class="nb">isinstance</span><span class="p">(</span><span class="n">test_holdout</span><span class="p">,</span> <span class="p">(</span><span class="nb">list</span><span class="p">,)):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">validation_holdout</span> <span class="o">=</span> <span class="p">[</span>
                    <span class="nb">str</span><span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">validation_holdout</span><span class="p">]</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">test_holdout</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">test_holdout</span><span class="p">]</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_holdout_type</span> <span class="o">=</span> <span class="s2">&quot;chromosome&quot;</span>
            <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">validation_holdout</span><span class="p">,</span> <span class="nb">float</span><span class="p">)</span> <span class="ow">and</span> \
                    <span class="nb">isinstance</span><span class="p">(</span><span class="n">test_holdout</span><span class="p">,</span> <span class="nb">float</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">validation_holdout</span> <span class="o">=</span> <span class="n">validation_holdout</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">test_holdout</span> <span class="o">=</span> <span class="n">test_holdout</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_holdout_type</span> <span class="o">=</span> <span class="s2">&quot;proportion&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                    <span class="s2">&quot;Validation holdout and test holdout must have the &quot;</span>
                    <span class="s2">&quot;same type (list or float) but validation was &quot;</span>
                    <span class="s2">&quot;type </span><span class="si">{0}</span><span class="s2"> and test was type </span><span class="si">{1}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="nb">type</span><span class="p">(</span><span class="n">validation_holdout</span><span class="p">),</span> <span class="nb">type</span><span class="p">(</span><span class="n">test_holdout</span><span class="p">)))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">test_holdout</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">validation_holdout</span><span class="p">,</span> <span class="p">(</span><span class="nb">list</span><span class="p">,)):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">validation_holdout</span> <span class="o">=</span> <span class="p">[</span>
                    <span class="nb">str</span><span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">validation_holdout</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">validation_holdout</span> <span class="o">=</span> <span class="n">validation_holdout</span>

        <span class="k">if</span> <span class="n">mode</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">modes</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Mode must be one of </span><span class="si">{0}</span><span class="s2">. Input was &#39;</span><span class="si">{1}</span><span class="s2">&#39;.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">modes</span><span class="p">,</span> <span class="n">mode</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mode</span> <span class="o">=</span> <span class="n">mode</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">surrounding_sequence_radius</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span>
            <span class="n">surrounding_sequence_length</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sequence_length</span> <span class="o">=</span> <span class="n">sequence_length</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">bin_radius</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">center_bin_to_predict</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_start_radius</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">bin_radius</span>
        <span class="k">if</span> <span class="n">center_bin_to_predict</span> <span class="o">%</span> <span class="mi">2</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_end_radius</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">bin_radius</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_end_radius</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">bin_radius</span> <span class="o">+</span> <span class="mi">1</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span> <span class="o">=</span> <span class="n">reference_sequence</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_features</span> <span class="o">=</span> <span class="n">features</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">n_features</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_features</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">target</span> <span class="o">=</span> <span class="n">GenomicFeatures</span><span class="p">(</span>
            <span class="n">target_path</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_features</span><span class="p">,</span>
            <span class="n">feature_thresholds</span><span class="o">=</span><span class="n">feature_thresholds</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">save_datasets</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">mode</span> <span class="ow">in</span> <span class="n">save_datasets</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">save_datasets</span><span class="p">[</span><span class="n">mode</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

<div class="viewcode-block" id="OnlineSampler.get_feature_from_index"><a class="viewcode-back" href="../../../samplers.html#selene.samplers.OnlineSampler.get_feature_from_index">[docs]</a>    <span class="k">def</span> <span class="nf">get_feature_from_index</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns the feature corresponding to an index in the feature</span>
<span class="sd">        vector.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        index : int</span>
<span class="sd">            The index of the feature to retrieve the name for.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        str</span>
<span class="sd">            The name of the feature occurring at the specified index.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="o">.</span><span class="n">index_feature_dict</span><span class="p">[</span><span class="n">index</span><span class="p">]</span></div>

<div class="viewcode-block" id="OnlineSampler.get_sequence_from_encoding"><a class="viewcode-back" href="../../../samplers.html#selene.samplers.OnlineSampler.get_sequence_from_encoding">[docs]</a>    <span class="k">def</span> <span class="nf">get_sequence_from_encoding</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">encoding</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Gets the string sequence from the one-hot encoding</span>
<span class="sd">        of the sequence.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        encoding : numpy.ndarray</span>
<span class="sd">            An :math:`L \\times N` array (where :math:`L` is the length</span>
<span class="sd">            of the sequence and :math:`N` is the size of the sequence</span>
<span class="sd">            type&#39;s alphabet) containing the one-hot encoding of the</span>
<span class="sd">            sequence.</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        str</span>
<span class="sd">            The sequence of :math:`L` characters decoded from the input.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference_sequence</span><span class="o">.</span><span class="n">encoding_to_sequence</span><span class="p">(</span><span class="n">encoding</span><span class="p">)</span></div>

    <span class="c1"># @TODO: Enable saving of training data coordinates intermittently.</span>
<div class="viewcode-block" id="OnlineSampler.save_datasets_to_file"><a class="viewcode-back" href="../../../samplers.html#selene.samplers.OnlineSampler.save_datasets_to_file">[docs]</a>    <span class="k">def</span> <span class="nf">save_datasets_to_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">output_dir</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Save samples for each partition (i.e. train/validate/test) to</span>
<span class="sd">        disk.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        output_dir : str</span>
<span class="sd">            Path to the output directory where we should save each of</span>
<span class="sd">            the datasets.</span>

<span class="sd">        Notes</span>
<span class="sd">        -----</span>
<span class="sd">        This likely only works for validation and test right now.</span>
<span class="sd">        Training data may be too big to store in a list in memory, and</span>
<span class="sd">        cannot yet be written to file intermittently.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">mode</span><span class="p">,</span> <span class="n">samples</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">save_datasets</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">filepath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_dir</span><span class="p">,</span> <span class="s2">&quot;</span><span class="si">{0}</span><span class="s2">_data.bed&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">mode</span><span class="p">))</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="s1">&#39;w+&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file_handle</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">cols</span> <span class="ow">in</span> <span class="n">samples</span><span class="p">:</span>
                    <span class="n">line</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">cols</span><span class="p">])</span>
                    <span class="n">file_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{0}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">line</span><span class="p">))</span></div></div>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Selene Contributors.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../../',
            VERSION:'master',
            LANGUAGE:'None',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  <script type="text/javascript" src="../../../_static/js/theme.js"></script>

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>
 



</body>
</html>